// @matter/node/dist/esm/endpoint/storage/DatasourceStore.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function DatasourceStore(endpointStore, behaviorId, initialValues) {
  return {
    initialValues,
    async set(transaction, values) {
      if (behaviorId === "__proto__") {
        return;
      }
      const participant = participantFor(transaction, endpointStore);
      if (!participant.mutations) {
        participant.mutations = {};
      }
      const behaviorMutations = participant.mutations[behaviorId];
      if (behaviorMutations) {
        for (const key in values) {
          behaviorMutations[key] = values[key];
        }
      } else {
        participant.mutations[behaviorId] = { ...values };
      }
    }
  };
}
function participantFor(transaction, endpointStore) {
  let participant = transaction.getParticipant(endpointStore);
  if (participant) {
    return participant;
  }
  participant = {
    role: endpointStore,
    mutations: void 0,
    toString() {
      return endpointStore.toString();
    },
    commit1() {
    },
    async commit2() {
      if (this.mutations) {
        await endpointStore.set(this.mutations);
      }
    },
    rollback() {
      delete this.mutations;
    }
  };
  transaction.addParticipants(participant);
  return participant;
}
export {
  DatasourceStore
};
