// @matter/protocol/dist/esm/protocol/DeviceCommissioner.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { AdministratorCommissioning } from "#clusters/administrator-commissioning";
import { CommissioningMode } from "#common/InstanceBroadcaster.js";
import { FabricManager } from "#fabric/FabricManager.js";
import {
  Crypto,
  Diagnostic,
  Environmental,
  InternalError,
  Lifecycle,
  Logger,
  MatterFlowError,
  ObserverGroup
} from "#general";
import { SecureChannelProtocol } from "#securechannel/SecureChannelProtocol.js";
import { PaseServer, SessionManager } from "#session/index.js";
import { StatusCode, StatusResponseError } from "#types";
import { DeviceAdvertiser } from "./DeviceAdvertiser.js";
const logger = Logger.get("DeviceCommissioner");
class CommissioningConfigProvider {
}
class DeviceCommissioner {
  #context;
  #failsafeContext;
  #windowStatus = AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
  #activeDiscriminator;
  #activeCommissioningEndCallback;
  #observers = new ObserverGroup(this);
  constructor(context) {
    this.#context = context;
    this.#observers.on(this.#context.advertiser.timedOut, this.endCommissioning);
    this.#observers.on(this.#context.fabrics.events.deleted, async () => {
      if (this.#context.fabrics.length === 0 || this.#windowStatus !== AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
        this.reactivateAdvertiser();
      }
    });
    this.#observers.on(this.#context.advertiser.operationalModeEnded, this.allowBasicCommissioning);
    this.#observers.on(this.#context.secureChannelProtocol.tooManyPaseErrors, async () => {
      logger.info("Maximum number of PASE pairing errors reached, canceling commissioning");
      await this.endCommissioning();
    });
    this.#observers.on(this.#context.sessions.sessions.deleted, (session) => {
      const currentFabricIndex = session.fabric?.fabricIndex;
      const existingSessionFabric = currentFabricIndex === void 0 ? void 0 : this.#context.fabrics.findByIndex(currentFabricIndex)?.fabricIndex;
      if (this.#context.fabrics.length > 0 || session.isPase || !existingSessionFabric) {
        this.#context.advertiser.startAdvertising().catch((error) => logger.warn(`Error while announcing`, error));
      }
    });
  }
  static [Environmental.create](env) {
    const instance = new DeviceCommissioner({
      fabrics: env.get(FabricManager),
      sessions: env.get(SessionManager),
      advertiser: env.get(DeviceAdvertiser),
      secureChannelProtocol: env.get(SecureChannelProtocol),
      commissioningConfig: env.get(CommissioningConfigProvider)
    });
    env.set(DeviceCommissioner, instance);
    return instance;
  }
  get failsafeContext() {
    this.assertFailsafeArmed();
    return this.#failsafeContext;
  }
  async allowEnhancedCommissioning(discriminator, paseServer, commissioningEndCallback) {
    if (this.#windowStatus === AdministratorCommissioning.CommissioningWindowStatus.BasicWindowOpen) {
      throw new MatterFlowError(
        "Basic commissioning window is already open! Cannot set Enhanced commissioning mode."
      );
    }
    this.#context.secureChannelProtocol.setPaseCommissioner(paseServer);
    await this.#becomeCommissionable(
      AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen,
      commissioningEndCallback,
      discriminator
    );
  }
  async allowBasicCommissioning(commissioningEndCallback) {
    if (this.#windowStatus === AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen) {
      throw new MatterFlowError(
        "Enhanced commissioning window is already open! Cannot set Basic commissioning mode."
      );
    }
    this.#context.secureChannelProtocol.setPaseCommissioner(
      await PaseServer.fromPin(this.#context.sessions, this.#context.commissioningConfig.values.passcode, {
        iterations: 1e3,
        salt: Crypto.get().getRandomData(32)
      })
    );
    await this.#becomeCommissionable(
      AdministratorCommissioning.CommissioningWindowStatus.BasicWindowOpen,
      commissioningEndCallback
    );
  }
  async beginTimed(failsafeContext) {
    await failsafeContext.construction;
    this.#failsafeContext = failsafeContext;
    this.#context.fabrics.events.added.on((fabric) => {
      const fabrics = this.#context.fabrics.fabrics;
      this.#context.advertiser.advertiseFabrics(fabrics, true).catch(
        (error) => logger.warn(`Error sending Fabric announcement for Index ${fabric.fabricIndex}`, error)
      );
      logger.info("Announce done", Diagnostic.dict({ fabric: fabric.fabricId, fabricIndex: fabric.fabricIndex }));
    });
    failsafeContext.commissioned.on(async () => await this.endCommissioning());
    failsafeContext.construction.change.on((status) => {
      if (status === Lifecycle.Status.Destroyed) {
        this.#failsafeContext = void 0;
      }
    });
  }
  get isFailsafeArmed() {
    return this.#failsafeContext !== void 0;
  }
  assertFailsafeArmed(message) {
    if (this.isFailsafeArmed) return;
    throw new StatusResponseError(
      message ?? "Failsafe timer needs to be armed to execute this action.",
      StatusCode.FailsafeRequired
    );
  }
  reactivateAdvertiser() {
    if (this.#windowStatus === AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      return;
    }
    this.#enterCommissioningMode(this.#windowStatus, this.#activeDiscriminator).catch(
      (error) => logger.warn("Error sending announcement:", error)
    );
  }
  async #enterCommissioningMode(windowStatus, discriminator) {
    this.#windowStatus = windowStatus;
    const commissioningConfig = this.#context.commissioningConfig.values;
    await this.#context.advertiser.enterCommissioningMode(
      windowStatus === AdministratorCommissioning.CommissioningWindowStatus.EnhancedWindowOpen ? CommissioningMode.Enhanced : CommissioningMode.Basic,
      {
        ...commissioningConfig.productDescription,
        discriminator: discriminator ?? commissioningConfig.discriminator
      }
    );
  }
  async #becomeCommissionable(windowStatus, activeCommissioningEndCallback, discriminator) {
    if (this.#windowStatus === windowStatus && (discriminator === void 0 || discriminator === this.#activeDiscriminator)) {
      return this.reactivateAdvertiser();
    }
    if (this.#windowStatus !== AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      throw new InternalError(`Commissioning window already open with different mode (${this.#windowStatus})!`);
    }
    if (this.#activeCommissioningEndCallback !== void 0) {
      throw new InternalError("Commissioning window already open with different callback!");
    }
    this.#activeCommissioningEndCallback = activeCommissioningEndCallback;
    this.#activeDiscriminator = discriminator;
    this.#enterCommissioningMode(windowStatus, discriminator).catch(
      (error) => logger.warn("Error sending announcement:", error)
    );
  }
  async endCommissioning() {
    if (this.#windowStatus === AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen) {
      return;
    }
    logger.debug("Commissioning mode ended, stop announcements.");
    this.#context.secureChannelProtocol.removePaseCommissioner();
    this.#windowStatus = AdministratorCommissioning.CommissioningWindowStatus.WindowNotOpen;
    if (this.#activeCommissioningEndCallback !== void 0) {
      const activeCommissioningEndCallback = this.#activeCommissioningEndCallback;
      this.#activeCommissioningEndCallback = void 0;
      await activeCommissioningEndCallback();
    }
    await this.#context.advertiser.exitCommissioningMode();
    logger.info("All commissioning announcements stopped");
  }
  async close() {
    this.#observers.close();
    await this.endCommissioning();
    if (this.#failsafeContext) {
      await this.#failsafeContext.close();
      this.#failsafeContext = void 0;
    }
  }
}
export {
  CommissioningConfigProvider,
  DeviceCommissioner
};
