// @matter/model/dist/cjs/common/ElementTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ElementTag_exports = {};
__export(ElementTag_exports, {
  ElementTag: () => ElementTag
});
module.exports = __toCommonJS(ElementTag_exports);
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ElementTag = /* @__PURE__ */ ((ElementTag2) => {
  ElementTag2["Fabric"] = "fabric";
  ElementTag2["Node"] = "node";
  ElementTag2["Endpoint"] = "endpoint";
  ElementTag2["Cluster"] = "cluster";
  ElementTag2["Command"] = "command";
  ElementTag2["Event"] = "event";
  ElementTag2["Attribute"] = "attribute";
  ElementTag2["DeviceType"] = "deviceType";
  ElementTag2["Datatype"] = "datatype";
  ElementTag2["Field"] = "field";
  ElementTag2["SemanticNamespace"] = "semanticNamespace";
  ElementTag2["SemanticTag"] = "semanticTag";
  ElementTag2["Requirement"] = "requirement";
  ElementTag2["Matter"] = "matter";
  return ElementTag2;
})(ElementTag || {});
