// @matter/model/dist/cjs/common/Specification.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Specification_exports = {};
__export(Specification_exports, {
  Specification: () => Specification
});
module.exports = __toCommonJS(Specification_exports);
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Specification = /* @__PURE__ */ ((Specification2) => {
  Specification2["Core"] = "core";
  Specification2["Cluster"] = "cluster";
  Specification2["Device"] = "device";
  Specification2["Namespace"] = "namespace";
  return Specification2;
})(Specification || {});
((Specification2) => {
  let Names;
  ((Names2) => {
    Names2["core"] = "Matter Core Specification";
    Names2["cluster"] = "Matter Application Cluster Specification";
    Names2["device"] = "Matter Device Library Specification";
    Names2["namespace"] = "Matter Standard Namespace Specification";
  })(Names = Specification2.Names || (Specification2.Names = {}));
  Specification2.REVISION = "1.4";
  Specification2.SPECIFICATION_VERSION = 17039360;
  Specification2.DATA_MODEL_REVISION = 18;
  Specification2.INTERACTION_MODEL_REVISION = 12;
})(Specification || (Specification = {}));
