// @matter/general/dist/esm/net/UdpInterface.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ChannelType } from "./Channel.js";
import { NetworkError } from "./Network.js";
class UdpInterface {
  constructor(server) {
    this.server = server;
  }
  static async create(network, type, port, host, netInterface) {
    return new UdpInterface(
      await network.createUdpChannel({ listeningPort: port, type, netInterface, listeningAddress: host })
    );
  }
  supports(type, address) {
    return this.server.supports(type, address);
  }
  async openChannel(address) {
    if (address.type !== "udp") {
      throw new NetworkError(`Unsupported address type ${address.type}`);
    }
    const { ip, port } = address;
    return Promise.resolve(new UdpConnection(this.server, ip, port));
  }
  onData(listener) {
    return this.server.onData(
      (_netInterface, peerHost, peerPort, data) => listener(new UdpConnection(this.server, peerHost, peerPort), data)
    );
  }
  get port() {
    return this.server.port;
  }
  close() {
    return this.server.close();
  }
}
class UdpConnection {
  constructor(server, peerAddress, peerPort) {
    this.server = server;
    this.peerAddress = peerAddress;
    this.peerPort = peerPort;
  }
  isReliable = false;
  type = ChannelType.UDP;
  get maxPayloadSize() {
    return this.server.maxPayloadSize;
  }
  send(data) {
    return this.server.send(this.peerAddress, this.peerPort, data);
  }
  get name() {
    return `${this.type}://[${this.peerAddress}]:${this.peerPort}`;
  }
  get networkAddress() {
    return { type: "udp", ip: this.peerAddress, port: this.peerPort };
  }
  async close() {
  }
}
export {
  UdpInterface
};
