// @matter/general/dist/cjs/storage/Storage.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Storage_exports = {};
__export(Storage_exports, {
  CloneableStorage: () => CloneableStorage,
  MaybeAsyncStorage: () => MaybeAsyncStorage,
  StorageError: () => StorageError,
  SyncStorage: () => SyncStorage
});
module.exports = __toCommonJS(Storage_exports);
var import_MatterError = require("../MatterError.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class StorageError extends import_MatterError.MatterError {
}
var CloneableStorage;
((CloneableStorage2) => {
  function is(storage) {
    return "clone" in storage && typeof storage.clone === "function";
  }
  CloneableStorage2.is = is;
  function assert(storage) {
    if (!is(storage)) {
      throw new import_MatterError.ImplementationError("Storage does not support required snapshotting function");
    }
  }
  CloneableStorage2.assert = assert;
})(CloneableStorage || (CloneableStorage = {}));
class MaybeAsyncStorage {
}
class SyncStorage {
}
