// @matter/node/dist/cjs/endpoints/aggregator.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var aggregator_exports = {};
__export(aggregator_exports, {
  AggregatorEndpoint: () => AggregatorEndpoint,
  AggregatorEndpointDefinition: () => AggregatorEndpointDefinition,
  AggregatorRequirements: () => AggregatorRequirements
});
module.exports = __toCommonJS(aggregator_exports);
var import_PartsBehavior = require("../behavior/system/parts/PartsBehavior.js");
var import_IndexBehavior = require("../behavior/system/index/IndexBehavior.js");
var import_ActionsServer = require("../behaviors/actions/ActionsServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_CommissionerControlServer = require("../behaviors/commissioner-control/CommissionerControlServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var AggregatorRequirements;
((AggregatorRequirements2) => {
  AggregatorRequirements2.ActionsServer = import_ActionsServer.ActionsServer;
  AggregatorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  AggregatorRequirements2.CommissionerControlServer = import_CommissionerControlServer.CommissionerControlServer;
  AggregatorRequirements2.server = {
    mandatory: { Parts: import_PartsBehavior.PartsBehavior, Index: import_IndexBehavior.IndexBehavior },
    optional: { Actions: AggregatorRequirements2.ActionsServer, Identify: AggregatorRequirements2.IdentifyServer, CommissionerControl: AggregatorRequirements2.CommissionerControlServer }
  };
})(AggregatorRequirements || (AggregatorRequirements = {}));
const AggregatorEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "Aggregator",
  deviceType: 14,
  deviceRevision: 2,
  requirements: AggregatorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    AggregatorRequirements.server.mandatory.Parts,
    AggregatorRequirements.server.mandatory.Index
  )
});
const AggregatorEndpoint = AggregatorEndpointDefinition;
