// @matter/node/dist/cjs/endpoints/bridged-node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bridged_node_exports = {};
__export(bridged_node_exports, {
  BridgedNodeEndpoint: () => BridgedNodeEndpoint,
  BridgedNodeEndpointDefinition: () => BridgedNodeEndpointDefinition,
  BridgedNodeRequirements: () => BridgedNodeRequirements
});
module.exports = __toCommonJS(bridged_node_exports);
var import_PartsBehavior = require("../behavior/system/parts/PartsBehavior.js");
var import_IndexBehavior = require("../behavior/system/index/IndexBehavior.js");
var import_BridgedDeviceBasicInformationServer = require("../behaviors/bridged-device-basic-information/BridgedDeviceBasicInformationServer.js");
var import_PowerSourceConfigurationServer = require("../behaviors/power-source-configuration/PowerSourceConfigurationServer.js");
var import_PowerSourceServer = require("../behaviors/power-source/PowerSourceServer.js");
var import_EcosystemInformationServer = require("../behaviors/ecosystem-information/EcosystemInformationServer.js");
var import_AdministratorCommissioningServer = require("../behaviors/administrator-commissioning/AdministratorCommissioningServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_model = require("#model");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var BridgedNodeRequirements;
((BridgedNodeRequirements2) => {
  BridgedNodeRequirements2.BridgedDeviceBasicInformationServer = import_BridgedDeviceBasicInformationServer.BridgedDeviceBasicInformationServer;
  BridgedNodeRequirements2.PowerSourceConfigurationServer = import_PowerSourceConfigurationServer.PowerSourceConfigurationServer;
  BridgedNodeRequirements2.PowerSourceServer = import_PowerSourceServer.PowerSourceServer;
  BridgedNodeRequirements2.EcosystemInformationServer = import_EcosystemInformationServer.EcosystemInformationServer;
  BridgedNodeRequirements2.AdministratorCommissioningServer = import_AdministratorCommissioningServer.AdministratorCommissioningServer;
  BridgedNodeRequirements2.server = {
    mandatory: {
      Parts: import_PartsBehavior.PartsBehavior,
      Index: import_IndexBehavior.IndexBehavior,
      BridgedDeviceBasicInformation: BridgedNodeRequirements2.BridgedDeviceBasicInformationServer
    },
    optional: {
      PowerSourceConfiguration: BridgedNodeRequirements2.PowerSourceConfigurationServer,
      PowerSource: BridgedNodeRequirements2.PowerSourceServer,
      EcosystemInformation: BridgedNodeRequirements2.EcosystemInformationServer,
      AdministratorCommissioning: BridgedNodeRequirements2.AdministratorCommissioningServer
    }
  };
})(BridgedNodeRequirements || (BridgedNodeRequirements = {}));
const BridgedNodeEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "BridgedNode",
  deviceType: 19,
  deviceRevision: 3,
  deviceClass: import_model.DeviceClassification.Utility,
  requirements: BridgedNodeRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    BridgedNodeRequirements.server.mandatory.Parts,
    BridgedNodeRequirements.server.mandatory.Index,
    BridgedNodeRequirements.server.mandatory.BridgedDeviceBasicInformation
  )
});
const BridgedNodeEndpoint = BridgedNodeEndpointDefinition;
