// @matter/node/dist/cjs/endpoints/device-energy-management.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var device_energy_management_exports = {};
__export(device_energy_management_exports, {
  DeviceEnergyManagementEndpoint: () => DeviceEnergyManagementEndpoint,
  DeviceEnergyManagementEndpointDefinition: () => DeviceEnergyManagementEndpointDefinition,
  DeviceEnergyManagementRequirements: () => DeviceEnergyManagementRequirements
});
module.exports = __toCommonJS(device_energy_management_exports);
var import_DeviceEnergyManagementServer = require("../behaviors/device-energy-management/DeviceEnergyManagementServer.js");
var import_DeviceEnergyManagementModeServer = require("../behaviors/device-energy-management-mode/DeviceEnergyManagementModeServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_model = require("#model");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DeviceEnergyManagementRequirements;
((DeviceEnergyManagementRequirements2) => {
  DeviceEnergyManagementRequirements2.DeviceEnergyManagementServer = import_DeviceEnergyManagementServer.DeviceEnergyManagementServer;
  DeviceEnergyManagementRequirements2.DeviceEnergyManagementModeServer = import_DeviceEnergyManagementModeServer.DeviceEnergyManagementModeServer;
  DeviceEnergyManagementRequirements2.server = {
    mandatory: { DeviceEnergyManagement: DeviceEnergyManagementRequirements2.DeviceEnergyManagementServer },
    optional: { DeviceEnergyManagementMode: DeviceEnergyManagementRequirements2.DeviceEnergyManagementModeServer }
  };
})(DeviceEnergyManagementRequirements || (DeviceEnergyManagementRequirements = {}));
const DeviceEnergyManagementEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "DeviceEnergyManagement",
  deviceType: 1293,
  deviceRevision: 2,
  deviceClass: import_model.DeviceClassification.Utility,
  requirements: DeviceEnergyManagementRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const DeviceEnergyManagementEndpoint = DeviceEnergyManagementEndpointDefinition;
