// @matter/node/dist/cjs/endpoints/joint-fabric-administrator.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var joint_fabric_administrator_exports = {};
__export(joint_fabric_administrator_exports, {
  JointFabricAdministratorEndpoint: () => JointFabricAdministratorEndpoint,
  JointFabricAdministratorEndpointDefinition: () => JointFabricAdministratorEndpointDefinition,
  JointFabricAdministratorRequirements: () => JointFabricAdministratorRequirements
});
module.exports = __toCommonJS(joint_fabric_administrator_exports);
var import_JointFabricPkiServer = require("../behaviors/joint-fabric-pki/JointFabricPkiServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_model = require("#model");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var JointFabricAdministratorRequirements;
((JointFabricAdministratorRequirements2) => {
  JointFabricAdministratorRequirements2.JointFabricPkiServer = import_JointFabricPkiServer.JointFabricPkiServer;
  JointFabricAdministratorRequirements2.server = { mandatory: { JointFabricPki: JointFabricAdministratorRequirements2.JointFabricPkiServer } };
})(JointFabricAdministratorRequirements || (JointFabricAdministratorRequirements = {}));
const JointFabricAdministratorEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "JointFabricAdministrator",
  deviceType: 304,
  deviceRevision: 1,
  deviceClass: import_model.DeviceClassification.Utility,
  requirements: JointFabricAdministratorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(JointFabricAdministratorRequirements.server.mandatory.JointFabricPki)
});
const JointFabricAdministratorEndpoint = JointFabricAdministratorEndpointDefinition;
