// @matter/node/dist/cjs/endpoints/ota-requestor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ota_requestor_exports = {};
__export(ota_requestor_exports, {
  OtaRequestorEndpoint: () => OtaRequestorEndpoint,
  OtaRequestorEndpointDefinition: () => OtaRequestorEndpointDefinition,
  OtaRequestorRequirements: () => OtaRequestorRequirements
});
module.exports = __toCommonJS(ota_requestor_exports);
var import_OtaSoftwareUpdateRequestorServer = require("../behaviors/ota-software-update-requestor/OtaSoftwareUpdateRequestorServer.js");
var import_OtaSoftwareUpdateProviderBehavior = require("../behaviors/ota-software-update-provider/OtaSoftwareUpdateProviderBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_model = require("#model");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OtaRequestorRequirements;
((OtaRequestorRequirements2) => {
  OtaRequestorRequirements2.OtaSoftwareUpdateRequestorServer = import_OtaSoftwareUpdateRequestorServer.OtaSoftwareUpdateRequestorServer;
  OtaRequestorRequirements2.OtaSoftwareUpdateProviderBehavior = import_OtaSoftwareUpdateProviderBehavior.OtaSoftwareUpdateProviderBehavior;
  OtaRequestorRequirements2.server = { mandatory: { OtaSoftwareUpdateRequestor: OtaRequestorRequirements2.OtaSoftwareUpdateRequestorServer } };
  OtaRequestorRequirements2.client = { mandatory: { OtaSoftwareUpdateProvider: OtaRequestorRequirements2.OtaSoftwareUpdateProviderBehavior } };
})(OtaRequestorRequirements || (OtaRequestorRequirements = {}));
const OtaRequestorEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "OtaRequestor",
  deviceType: 18,
  deviceRevision: 1,
  deviceClass: import_model.DeviceClassification.Utility,
  requirements: OtaRequestorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(OtaRequestorRequirements.server.mandatory.OtaSoftwareUpdateRequestor)
});
const OtaRequestorEndpoint = OtaRequestorEndpointDefinition;
