// @matter/node/dist/cjs/endpoints/power-source.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var power_source_exports = {};
__export(power_source_exports, {
  PowerSourceEndpoint: () => PowerSourceEndpoint,
  PowerSourceEndpointDefinition: () => PowerSourceEndpointDefinition,
  PowerSourceRequirements: () => PowerSourceRequirements
});
module.exports = __toCommonJS(power_source_exports);
var import_PowerSourceServer = require("../behaviors/power-source/PowerSourceServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_model = require("#model");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var PowerSourceRequirements;
((PowerSourceRequirements2) => {
  PowerSourceRequirements2.PowerSourceServer = import_PowerSourceServer.PowerSourceServer;
  PowerSourceRequirements2.server = { mandatory: { PowerSource: PowerSourceRequirements2.PowerSourceServer } };
})(PowerSourceRequirements || (PowerSourceRequirements = {}));
const PowerSourceEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "PowerSource",
  deviceType: 17,
  deviceRevision: 1,
  deviceClass: import_model.DeviceClassification.Utility,
  requirements: PowerSourceRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const PowerSourceEndpoint = PowerSourceEndpointDefinition;
