// @matter/node/dist/cjs/endpoints/secondary-network-interface.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var secondary_network_interface_exports = {};
__export(secondary_network_interface_exports, {
  SecondaryNetworkInterfaceEndpoint: () => SecondaryNetworkInterfaceEndpoint,
  SecondaryNetworkInterfaceEndpointDefinition: () => SecondaryNetworkInterfaceEndpointDefinition,
  SecondaryNetworkInterfaceRequirements: () => SecondaryNetworkInterfaceRequirements
});
module.exports = __toCommonJS(secondary_network_interface_exports);
var import_NetworkCommissioningServer = require("../behaviors/network-commissioning/NetworkCommissioningServer.js");
var import_EthernetNetworkDiagnosticsServer = require("../behaviors/ethernet-network-diagnostics/EthernetNetworkDiagnosticsServer.js");
var import_WiFiNetworkDiagnosticsServer = require("../behaviors/wi-fi-network-diagnostics/WiFiNetworkDiagnosticsServer.js");
var import_ThreadNetworkDiagnosticsServer = require("../behaviors/thread-network-diagnostics/ThreadNetworkDiagnosticsServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_model = require("#model");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var SecondaryNetworkInterfaceRequirements;
((SecondaryNetworkInterfaceRequirements2) => {
  SecondaryNetworkInterfaceRequirements2.NetworkCommissioningServer = import_NetworkCommissioningServer.NetworkCommissioningServer;
  SecondaryNetworkInterfaceRequirements2.EthernetNetworkDiagnosticsServer = import_EthernetNetworkDiagnosticsServer.EthernetNetworkDiagnosticsServer;
  SecondaryNetworkInterfaceRequirements2.WiFiNetworkDiagnosticsServer = import_WiFiNetworkDiagnosticsServer.WiFiNetworkDiagnosticsServer;
  SecondaryNetworkInterfaceRequirements2.ThreadNetworkDiagnosticsServer = import_ThreadNetworkDiagnosticsServer.ThreadNetworkDiagnosticsServer;
  SecondaryNetworkInterfaceRequirements2.server = {
    mandatory: { NetworkCommissioning: SecondaryNetworkInterfaceRequirements2.NetworkCommissioningServer },
    optional: {
      EthernetNetworkDiagnostics: SecondaryNetworkInterfaceRequirements2.EthernetNetworkDiagnosticsServer,
      WiFiNetworkDiagnostics: SecondaryNetworkInterfaceRequirements2.WiFiNetworkDiagnosticsServer,
      ThreadNetworkDiagnostics: SecondaryNetworkInterfaceRequirements2.ThreadNetworkDiagnosticsServer
    }
  };
})(SecondaryNetworkInterfaceRequirements || (SecondaryNetworkInterfaceRequirements = {}));
const SecondaryNetworkInterfaceEndpointDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "SecondaryNetworkInterface",
  deviceType: 25,
  deviceRevision: 1,
  deviceClass: import_model.DeviceClassification.Utility,
  requirements: SecondaryNetworkInterfaceRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const SecondaryNetworkInterfaceEndpoint = SecondaryNetworkInterfaceEndpointDefinition;
