// @matter/types/dist/esm/clusters/laundry-dryer-controls.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MutableCluster } from "../cluster/mutation/MutableCluster.js";
import { Attribute, WritableAttribute } from "../cluster/Cluster.js";
import { TlvArray } from "../tlv/TlvArray.js";
import { TlvEnum } from "../tlv/TlvNumber.js";
import { TlvNullable } from "../tlv/TlvNullable.js";
import { ClusterRegistry } from "../cluster/ClusterRegistry.js";
var LaundryDryerControls;
((LaundryDryerControls2) => {
  let DrynessLevel;
  ((DrynessLevel2) => {
    DrynessLevel2[DrynessLevel2["Low"] = 0] = "Low";
    DrynessLevel2[DrynessLevel2["Normal"] = 1] = "Normal";
    DrynessLevel2[DrynessLevel2["Extra"] = 2] = "Extra";
    DrynessLevel2[DrynessLevel2["Max"] = 3] = "Max";
  })(DrynessLevel = LaundryDryerControls2.DrynessLevel || (LaundryDryerControls2.DrynessLevel = {}));
  LaundryDryerControls2.ClusterInstance = MutableCluster({
    id: 74,
    name: "LaundryDryerControls",
    revision: 1,
    attributes: {
      /**
       * Indicates the list of supported dryness levels available to the appliance in the currently selected mode.
       * The dryness level values are determined by the manufacturer. At least one dryness level value shall be
       * provided in the SupportedDrynessLevels list. The list of dryness levels may change depending on the
       * currently-selected Laundry Dryer mode.
       *
       * @see {@link MatterSpecification.v14.Cluster} ยง 8.9.5.1
       */
      supportedDrynessLevels: Attribute(0, TlvArray(TlvEnum(), { minLength: 1, maxLength: 4 })),
      /**
       * Indicates the currently-selected dryness level and it shall be the index into the SupportedDrynessLevels
       * list of the selected dryness level.
       *
       * If an attempt is made to write this attribute with a value other than null or a value contained in
       * SupportedDrynessLevels, a CONSTRAINT_ERROR response shall be sent as the response. If an attempt is made
       * to write this attribute while the device is not in a state that supports modifying the dryness level, an
       * INVALID_IN_STATE error shall be sent as the response. A value of null shall indicate that there will be
       * no dryness level setting for the current mode.
       *
       * @see {@link MatterSpecification.v14.Cluster} ยง 8.9.5.2
       */
      selectedDrynessLevel: WritableAttribute(1, TlvNullable(TlvEnum()))
    }
  });
  LaundryDryerControls2.Cluster = LaundryDryerControls2.ClusterInstance;
  LaundryDryerControls2.Complete = LaundryDryerControls2.Cluster;
})(LaundryDryerControls || (LaundryDryerControls = {}));
const LaundryDryerControlsCluster = LaundryDryerControls.Cluster;
ClusterRegistry.register(LaundryDryerControls.Complete);
export {
  LaundryDryerControls,
  LaundryDryerControlsCluster
};
