// @matter/node/dist/esm/behaviors/group-key-management/GroupKeyManagementServer.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ImplementationError } from "#general";
import { StatusCode, StatusResponseError } from "#types";
import { GroupKeyManagementBehavior } from "./GroupKeyManagementBehavior.js";
class GroupKeyManagementServer extends GroupKeyManagementBehavior {
  initialize() {
    if (this.state.maxGroupKeysPerFabric !== 1) {
      throw new ImplementationError("maxGroupKeysPerFabric must be 1 for now.");
    }
    if (this.state.maxGroupsPerFabric !== 0) {
      throw new ImplementationError("maxGroupsPerFabric must be 0 for now.");
    }
    this.state.groupTable = [];
  }
  keySetWrite() {
    throw new StatusResponseError(
      "We do not support additional groups beyond the IPK",
      StatusCode.ResourceExhausted
    );
  }
  keySetRead({
    groupKeySetId
  }) {
    if (this.context.session === void 0) {
      throw new ImplementationError("Session must be defined");
    }
    const fabric = this.context.session.associatedFabric;
    const groupKeySet = fabric.getGroupKeySet(groupKeySetId);
    if (groupKeySet === void 0) {
      throw new StatusResponseError(`GroupKeySet ${groupKeySetId} not found`, StatusCode.NotFound);
    }
    return {
      groupKeySet: {
        ...groupKeySet,
        epochKey0: null,
        epochKey1: null,
        epochKey2: null
      }
    };
  }
  keySetRemove({ groupKeySetId }) {
    if (groupKeySetId === 0) {
      throw new StatusResponseError(`GroupKeySet ${groupKeySetId} cannot be removed`, StatusCode.InvalidCommand);
    }
    throw new StatusResponseError(`GroupKeySet ${groupKeySetId} not found`, StatusCode.NotFound);
  }
  keySetReadAllIndices() {
    if (this.context.session === void 0) {
      throw new ImplementationError("Session must be defined");
    }
    const fabric = this.context.session.associatedFabric;
    const groupKeySetIDs = fabric.getAllGroupKeySets().map(({ groupKeySetId }) => groupKeySetId);
    return {
      groupKeySetIDs
    };
  }
}
export {
  GroupKeyManagementServer
};
