// @matter/protocol/dist/cjs/ble/BleConsts.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BleConsts_exports = {};
__export(BleConsts_exports, {
  BLE_MATTER_C1_CHARACTERISTIC_UUID: () => BLE_MATTER_C1_CHARACTERISTIC_UUID,
  BLE_MATTER_C2_CHARACTERISTIC_UUID: () => BLE_MATTER_C2_CHARACTERISTIC_UUID,
  BLE_MATTER_C3_CHARACTERISTIC_UUID: () => BLE_MATTER_C3_CHARACTERISTIC_UUID,
  BLE_MATTER_SERVICE_UUID: () => BLE_MATTER_SERVICE_UUID,
  BLE_MAXIMUM_BTP_MTU: () => BLE_MAXIMUM_BTP_MTU,
  BLE_MAX_MATTER_PAYLOAD_SIZE: () => BLE_MAX_MATTER_PAYLOAD_SIZE,
  BLE_MINIMUM_ATT_MTU: () => BLE_MINIMUM_ATT_MTU,
  BTP_ACK_TIMEOUT_MS: () => BTP_ACK_TIMEOUT_MS,
  BTP_CONN_IDLE_TIMEOUT: () => BTP_CONN_IDLE_TIMEOUT,
  BTP_CONN_RSP_TIMEOUT_MS: () => BTP_CONN_RSP_TIMEOUT_MS,
  BTP_MAXIMUM_WINDOW_SIZE: () => BTP_MAXIMUM_WINDOW_SIZE,
  BTP_SEND_ACK_TIMEOUT_MS: () => BTP_SEND_ACK_TIMEOUT_MS
});
module.exports = __toCommonJS(BleConsts_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const BLE_MATTER_SERVICE_UUID = "fff6";
const BLE_MATTER_C1_CHARACTERISTIC_UUID = "18EE2EF5-263D-4559-959F-4F9C429F9D11";
const BLE_MATTER_C2_CHARACTERISTIC_UUID = "18EE2EF5-263D-4559-959F-4F9C429F9D12";
const BLE_MATTER_C3_CHARACTERISTIC_UUID = "64630238-8772-45F2-B87D-748A83218F04";
const BLE_MINIMUM_ATT_MTU = 20;
const BLE_MAXIMUM_BTP_MTU = 244;
const BTP_MAXIMUM_WINDOW_SIZE = 255;
const BTP_CONN_RSP_TIMEOUT_MS = 5e3;
const BTP_ACK_TIMEOUT_MS = 15e3;
const BTP_SEND_ACK_TIMEOUT_MS = BTP_ACK_TIMEOUT_MS / 3;
const BTP_CONN_IDLE_TIMEOUT = 3e4;
const BLE_MAX_MATTER_PAYLOAD_SIZE = import_general.MAX_UDP_MESSAGE_SIZE;
