// @matter/node/dist/esm/devices/battery-storage.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var BatteryStorageRequirements;
((BatteryStorageRequirements2) => {
  BatteryStorageRequirements2.IdentifyServer = BaseIdentifyServer;
  BatteryStorageRequirements2.server = { optional: { Identify: BatteryStorageRequirements2.IdentifyServer }, mandatory: {} };
})(BatteryStorageRequirements || (BatteryStorageRequirements = {}));
const BatteryStorageDeviceDefinition = MutableEndpoint({
  name: "BatteryStorage",
  deviceType: 24,
  deviceRevision: 1,
  requirements: BatteryStorageRequirements,
  behaviors: SupportedBehaviors()
});
const BatteryStorageDevice = BatteryStorageDeviceDefinition;
export {
  BatteryStorageDevice,
  BatteryStorageDeviceDefinition,
  BatteryStorageRequirements
};
