// @matter/node/dist/esm/devices/color-dimmer-switch.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../behaviors/identify/IdentifyBehavior.js";
import { OnOffBehavior as BaseOnOffBehavior } from "../behaviors/on-off/OnOffBehavior.js";
import { LevelControlBehavior as BaseLevelControlBehavior } from "../behaviors/level-control/LevelControlBehavior.js";
import { ColorControlBehavior as BaseColorControlBehavior } from "../behaviors/color-control/ColorControlBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../behaviors/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../behaviors/scenes-management/ScenesManagementBehavior.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var ColorDimmerSwitchRequirements;
((ColorDimmerSwitchRequirements2) => {
  ColorDimmerSwitchRequirements2.IdentifyServer = BaseIdentifyServer;
  ColorDimmerSwitchRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  ColorDimmerSwitchRequirements2.OnOffBehavior = BaseOnOffBehavior;
  ColorDimmerSwitchRequirements2.LevelControlBehavior = BaseLevelControlBehavior;
  ColorDimmerSwitchRequirements2.ColorControlBehavior = BaseColorControlBehavior;
  ColorDimmerSwitchRequirements2.GroupsBehavior = BaseGroupsBehavior;
  ColorDimmerSwitchRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  ColorDimmerSwitchRequirements2.server = { mandatory: { Identify: ColorDimmerSwitchRequirements2.IdentifyServer } };
  ColorDimmerSwitchRequirements2.client = {
    mandatory: {
      Identify: ColorDimmerSwitchRequirements2.IdentifyBehavior,
      OnOff: ColorDimmerSwitchRequirements2.OnOffBehavior,
      LevelControl: ColorDimmerSwitchRequirements2.LevelControlBehavior,
      ColorControl: ColorDimmerSwitchRequirements2.ColorControlBehavior
    },
    optional: { Groups: ColorDimmerSwitchRequirements2.GroupsBehavior, ScenesManagement: ColorDimmerSwitchRequirements2.ScenesManagementBehavior }
  };
})(ColorDimmerSwitchRequirements || (ColorDimmerSwitchRequirements = {}));
const ColorDimmerSwitchDeviceDefinition = MutableEndpoint({
  name: "ColorDimmerSwitch",
  deviceType: 261,
  deviceRevision: 3,
  requirements: ColorDimmerSwitchRequirements,
  behaviors: SupportedBehaviors(ColorDimmerSwitchRequirements.server.mandatory.Identify)
});
const ColorDimmerSwitchDevice = ColorDimmerSwitchDeviceDefinition;
export {
  ColorDimmerSwitchDevice,
  ColorDimmerSwitchDeviceDefinition,
  ColorDimmerSwitchRequirements
};
