// @matter/node/dist/esm/devices/cooktop.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { OnOffServer as BaseOnOffServer } from "../behaviors/on-off/OnOffServer.js";
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var CooktopRequirements;
((CooktopRequirements2) => {
  CooktopRequirements2.OnOffServer = BaseOnOffServer.with("OffOnly");
  CooktopRequirements2.IdentifyServer = BaseIdentifyServer;
  CooktopRequirements2.server = { mandatory: { OnOff: CooktopRequirements2.OnOffServer }, optional: { Identify: CooktopRequirements2.IdentifyServer } };
})(CooktopRequirements || (CooktopRequirements = {}));
const CooktopDeviceDefinition = MutableEndpoint({
  name: "Cooktop",
  deviceType: 120,
  deviceRevision: 1,
  requirements: CooktopRequirements,
  behaviors: SupportedBehaviors(CooktopRequirements.server.mandatory.OnOff)
});
const CooktopDevice = CooktopDeviceDefinition;
export {
  CooktopDevice,
  CooktopDeviceDefinition,
  CooktopRequirements
};
