// @matter/node/dist/esm/devices/extended-color-light.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { GroupsServer as BaseGroupsServer } from "../behaviors/groups/GroupsServer.js";
import { OnOffServer as BaseOnOffServer } from "../behaviors/on-off/OnOffServer.js";
import { LevelControlServer as BaseLevelControlServer } from "../behaviors/level-control/LevelControlServer.js";
import { ColorControlServer as BaseColorControlServer } from "../behaviors/color-control/ColorControlServer.js";
import {
  ScenesManagementServer as BaseScenesManagementServer
} from "../behaviors/scenes-management/ScenesManagementServer.js";
import {
  OccupancySensingBehavior as BaseOccupancySensingBehavior
} from "../behaviors/occupancy-sensing/OccupancySensingBehavior.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var ExtendedColorLightRequirements;
((ExtendedColorLightRequirements2) => {
  ExtendedColorLightRequirements2.IdentifyServer = BaseIdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  ExtendedColorLightRequirements2.GroupsServer = BaseGroupsServer;
  ExtendedColorLightRequirements2.OnOffServer = BaseOnOffServer.with("Lighting");
  ExtendedColorLightRequirements2.LevelControlServer = BaseLevelControlServer.with("OnOff", "Lighting").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  ExtendedColorLightRequirements2.ColorControlServer = BaseColorControlServer.with("Xy", "ColorTemperature").alter({ attributes: { remainingTime: { optional: false } } });
  ExtendedColorLightRequirements2.ScenesManagementServer = BaseScenesManagementServer;
  ExtendedColorLightRequirements2.OccupancySensingBehavior = BaseOccupancySensingBehavior;
  ExtendedColorLightRequirements2.server = {
    mandatory: {
      Identify: ExtendedColorLightRequirements2.IdentifyServer,
      Groups: ExtendedColorLightRequirements2.GroupsServer,
      OnOff: ExtendedColorLightRequirements2.OnOffServer,
      LevelControl: ExtendedColorLightRequirements2.LevelControlServer,
      ColorControl: ExtendedColorLightRequirements2.ColorControlServer
    },
    optional: { ScenesManagement: ExtendedColorLightRequirements2.ScenesManagementServer }
  };
  ExtendedColorLightRequirements2.client = { optional: { OccupancySensing: ExtendedColorLightRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(ExtendedColorLightRequirements || (ExtendedColorLightRequirements = {}));
const ExtendedColorLightDeviceDefinition = MutableEndpoint({
  name: "ExtendedColorLight",
  deviceType: 269,
  deviceRevision: 4,
  requirements: ExtendedColorLightRequirements,
  behaviors: SupportedBehaviors(
    ExtendedColorLightRequirements.server.mandatory.Identify,
    ExtendedColorLightRequirements.server.mandatory.Groups,
    ExtendedColorLightRequirements.server.mandatory.OnOff,
    ExtendedColorLightRequirements.server.mandatory.LevelControl,
    ExtendedColorLightRequirements.server.mandatory.ColorControl
  )
});
const ExtendedColorLightDevice = ExtendedColorLightDeviceDefinition;
export {
  ExtendedColorLightDevice,
  ExtendedColorLightDeviceDefinition,
  ExtendedColorLightRequirements
};
