// @matter/node/dist/esm/devices/humidity-sensor.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import {
  RelativeHumidityMeasurementServer as BaseRelativeHumidityMeasurementServer
} from "../behaviors/relative-humidity-measurement/RelativeHumidityMeasurementServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var HumiditySensorRequirements;
((HumiditySensorRequirements2) => {
  HumiditySensorRequirements2.IdentifyServer = BaseIdentifyServer;
  HumiditySensorRequirements2.RelativeHumidityMeasurementServer = BaseRelativeHumidityMeasurementServer;
  HumiditySensorRequirements2.server = {
    mandatory: { Identify: HumiditySensorRequirements2.IdentifyServer, RelativeHumidityMeasurement: HumiditySensorRequirements2.RelativeHumidityMeasurementServer }
  };
})(HumiditySensorRequirements || (HumiditySensorRequirements = {}));
const HumiditySensorDeviceDefinition = MutableEndpoint({
  name: "HumiditySensor",
  deviceType: 775,
  deviceRevision: 2,
  requirements: HumiditySensorRequirements,
  behaviors: SupportedBehaviors(
    HumiditySensorRequirements.server.mandatory.Identify,
    HumiditySensorRequirements.server.mandatory.RelativeHumidityMeasurement
  )
});
const HumiditySensorDevice = HumiditySensorDeviceDefinition;
export {
  HumiditySensorDevice,
  HumiditySensorDeviceDefinition,
  HumiditySensorRequirements
};
