// @matter/node/dist/esm/devices/mode-select.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ModeSelectServer as BaseModeSelectServer } from "../behaviors/mode-select/ModeSelectServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var ModeSelectRequirements;
((ModeSelectRequirements2) => {
  ModeSelectRequirements2.ModeSelectServer = BaseModeSelectServer;
  ModeSelectRequirements2.server = { mandatory: { ModeSelect: ModeSelectRequirements2.ModeSelectServer } };
})(ModeSelectRequirements || (ModeSelectRequirements = {}));
const ModeSelectDeviceDefinition = MutableEndpoint({
  name: "ModeSelect",
  deviceType: 39,
  deviceRevision: 1,
  requirements: ModeSelectRequirements,
  behaviors: SupportedBehaviors(ModeSelectRequirements.server.mandatory.ModeSelect)
});
const ModeSelectDevice = ModeSelectDeviceDefinition;
export {
  ModeSelectDevice,
  ModeSelectDeviceDefinition,
  ModeSelectRequirements
};
