// @matter/node/dist/esm/devices/on-off-sensor.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../behaviors/identify/IdentifyBehavior.js";
import { OnOffBehavior as BaseOnOffBehavior } from "../behaviors/on-off/OnOffBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../behaviors/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../behaviors/scenes-management/ScenesManagementBehavior.js";
import { LevelControlBehavior as BaseLevelControlBehavior } from "../behaviors/level-control/LevelControlBehavior.js";
import { ColorControlBehavior as BaseColorControlBehavior } from "../behaviors/color-control/ColorControlBehavior.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var OnOffSensorRequirements;
((OnOffSensorRequirements2) => {
  OnOffSensorRequirements2.IdentifyServer = BaseIdentifyServer;
  OnOffSensorRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  OnOffSensorRequirements2.OnOffBehavior = BaseOnOffBehavior;
  OnOffSensorRequirements2.GroupsBehavior = BaseGroupsBehavior;
  OnOffSensorRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  OnOffSensorRequirements2.LevelControlBehavior = BaseLevelControlBehavior;
  OnOffSensorRequirements2.ColorControlBehavior = BaseColorControlBehavior;
  OnOffSensorRequirements2.server = { mandatory: { Identify: OnOffSensorRequirements2.IdentifyServer } };
  OnOffSensorRequirements2.client = {
    mandatory: { Identify: OnOffSensorRequirements2.IdentifyBehavior, OnOff: OnOffSensorRequirements2.OnOffBehavior },
    optional: {
      Groups: OnOffSensorRequirements2.GroupsBehavior,
      ScenesManagement: OnOffSensorRequirements2.ScenesManagementBehavior,
      LevelControl: OnOffSensorRequirements2.LevelControlBehavior,
      ColorControl: OnOffSensorRequirements2.ColorControlBehavior
    }
  };
})(OnOffSensorRequirements || (OnOffSensorRequirements = {}));
const OnOffSensorDeviceDefinition = MutableEndpoint({
  name: "OnOffSensor",
  deviceType: 2128,
  deviceRevision: 3,
  requirements: OnOffSensorRequirements,
  behaviors: SupportedBehaviors(OnOffSensorRequirements.server.mandatory.Identify)
});
const OnOffSensorDevice = OnOffSensorDeviceDefinition;
export {
  OnOffSensorDevice,
  OnOffSensorDeviceDefinition,
  OnOffSensorRequirements
};
