// @matter/node/dist/esm/devices/on-off-light-switch.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { IdentifyServer as BaseIdentifyServer } from "../behaviors/identify/IdentifyServer.js";
import { IdentifyBehavior as BaseIdentifyBehavior } from "../behaviors/identify/IdentifyBehavior.js";
import { OnOffBehavior as BaseOnOffBehavior } from "../behaviors/on-off/OnOffBehavior.js";
import { GroupsBehavior as BaseGroupsBehavior } from "../behaviors/groups/GroupsBehavior.js";
import {
  ScenesManagementBehavior as BaseScenesManagementBehavior
} from "../behaviors/scenes-management/ScenesManagementBehavior.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var OnOffLightSwitchRequirements;
((OnOffLightSwitchRequirements2) => {
  OnOffLightSwitchRequirements2.IdentifyServer = BaseIdentifyServer;
  OnOffLightSwitchRequirements2.IdentifyBehavior = BaseIdentifyBehavior;
  OnOffLightSwitchRequirements2.OnOffBehavior = BaseOnOffBehavior;
  OnOffLightSwitchRequirements2.GroupsBehavior = BaseGroupsBehavior;
  OnOffLightSwitchRequirements2.ScenesManagementBehavior = BaseScenesManagementBehavior;
  OnOffLightSwitchRequirements2.server = { mandatory: { Identify: OnOffLightSwitchRequirements2.IdentifyServer } };
  OnOffLightSwitchRequirements2.client = {
    mandatory: { Identify: OnOffLightSwitchRequirements2.IdentifyBehavior, OnOff: OnOffLightSwitchRequirements2.OnOffBehavior },
    optional: { Groups: OnOffLightSwitchRequirements2.GroupsBehavior, ScenesManagement: OnOffLightSwitchRequirements2.ScenesManagementBehavior }
  };
})(OnOffLightSwitchRequirements || (OnOffLightSwitchRequirements = {}));
const OnOffLightSwitchDeviceDefinition = MutableEndpoint({
  name: "OnOffLightSwitch",
  deviceType: 259,
  deviceRevision: 3,
  requirements: OnOffLightSwitchRequirements,
  behaviors: SupportedBehaviors(OnOffLightSwitchRequirements.server.mandatory.Identify)
});
const OnOffLightSwitchDevice = OnOffLightSwitchDeviceDefinition;
export {
  OnOffLightSwitchDevice,
  OnOffLightSwitchDeviceDefinition,
  OnOffLightSwitchRequirements
};
