// @matter/node/dist/esm/devices/video-remote-control.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { OnOffBehavior as BaseOnOffBehavior } from "../behaviors/on-off/OnOffBehavior.js";
import { MediaPlaybackBehavior as BaseMediaPlaybackBehavior } from "../behaviors/media-playback/MediaPlaybackBehavior.js";
import { KeypadInputBehavior as BaseKeypadInputBehavior } from "../behaviors/keypad-input/KeypadInputBehavior.js";
import { LevelControlBehavior as BaseLevelControlBehavior } from "../behaviors/level-control/LevelControlBehavior.js";
import { WakeOnLanBehavior as BaseWakeOnLanBehavior } from "../behaviors/wake-on-lan/WakeOnLanBehavior.js";
import { ChannelBehavior as BaseChannelBehavior } from "../behaviors/channel/ChannelBehavior.js";
import {
  TargetNavigatorBehavior as BaseTargetNavigatorBehavior
} from "../behaviors/target-navigator/TargetNavigatorBehavior.js";
import { MediaInputBehavior as BaseMediaInputBehavior } from "../behaviors/media-input/MediaInputBehavior.js";
import { LowPowerBehavior as BaseLowPowerBehavior } from "../behaviors/low-power/LowPowerBehavior.js";
import {
  ContentLauncherBehavior as BaseContentLauncherBehavior
} from "../behaviors/content-launcher/ContentLauncherBehavior.js";
import { AudioOutputBehavior as BaseAudioOutputBehavior } from "../behaviors/audio-output/AudioOutputBehavior.js";
import {
  ApplicationLauncherBehavior as BaseApplicationLauncherBehavior
} from "../behaviors/application-launcher/ApplicationLauncherBehavior.js";
import { AccountLoginBehavior as BaseAccountLoginBehavior } from "../behaviors/account-login/AccountLoginBehavior.js";
import {
  ContentControlBehavior as BaseContentControlBehavior
} from "../behaviors/content-control/ContentControlBehavior.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var VideoRemoteControlRequirements;
((VideoRemoteControlRequirements2) => {
  VideoRemoteControlRequirements2.OnOffBehavior = BaseOnOffBehavior;
  VideoRemoteControlRequirements2.MediaPlaybackBehavior = BaseMediaPlaybackBehavior;
  VideoRemoteControlRequirements2.KeypadInputBehavior = BaseKeypadInputBehavior;
  VideoRemoteControlRequirements2.LevelControlBehavior = BaseLevelControlBehavior;
  VideoRemoteControlRequirements2.WakeOnLanBehavior = BaseWakeOnLanBehavior;
  VideoRemoteControlRequirements2.ChannelBehavior = BaseChannelBehavior;
  VideoRemoteControlRequirements2.TargetNavigatorBehavior = BaseTargetNavigatorBehavior;
  VideoRemoteControlRequirements2.MediaInputBehavior = BaseMediaInputBehavior;
  VideoRemoteControlRequirements2.LowPowerBehavior = BaseLowPowerBehavior;
  VideoRemoteControlRequirements2.ContentLauncherBehavior = BaseContentLauncherBehavior;
  VideoRemoteControlRequirements2.AudioOutputBehavior = BaseAudioOutputBehavior;
  VideoRemoteControlRequirements2.ApplicationLauncherBehavior = BaseApplicationLauncherBehavior;
  VideoRemoteControlRequirements2.AccountLoginBehavior = BaseAccountLoginBehavior;
  VideoRemoteControlRequirements2.ContentControlBehavior = BaseContentControlBehavior;
  VideoRemoteControlRequirements2.client = {
    mandatory: { OnOff: VideoRemoteControlRequirements2.OnOffBehavior, MediaPlayback: VideoRemoteControlRequirements2.MediaPlaybackBehavior, KeypadInput: VideoRemoteControlRequirements2.KeypadInputBehavior },
    optional: {
      LevelControl: VideoRemoteControlRequirements2.LevelControlBehavior,
      WakeOnLan: VideoRemoteControlRequirements2.WakeOnLanBehavior,
      Channel: VideoRemoteControlRequirements2.ChannelBehavior,
      TargetNavigator: VideoRemoteControlRequirements2.TargetNavigatorBehavior,
      MediaInput: VideoRemoteControlRequirements2.MediaInputBehavior,
      LowPower: VideoRemoteControlRequirements2.LowPowerBehavior,
      ContentLauncher: VideoRemoteControlRequirements2.ContentLauncherBehavior,
      AudioOutput: VideoRemoteControlRequirements2.AudioOutputBehavior,
      ApplicationLauncher: VideoRemoteControlRequirements2.ApplicationLauncherBehavior,
      AccountLogin: VideoRemoteControlRequirements2.AccountLoginBehavior,
      ContentControl: VideoRemoteControlRequirements2.ContentControlBehavior
    }
  };
})(VideoRemoteControlRequirements || (VideoRemoteControlRequirements = {}));
const VideoRemoteControlDeviceDefinition = MutableEndpoint({
  name: "VideoRemoteControl",
  deviceType: 42,
  deviceRevision: 2,
  requirements: VideoRemoteControlRequirements,
  behaviors: SupportedBehaviors()
});
const VideoRemoteControlDevice = VideoRemoteControlDeviceDefinition;
export {
  VideoRemoteControlDevice,
  VideoRemoteControlDeviceDefinition,
  VideoRemoteControlRequirements
};
