// @matter/types/dist/cjs/common/StatusCode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var StatusCode_exports = {};
__export(StatusCode_exports, {
  ReceivedStatusResponseError: () => ReceivedStatusResponseError,
  StatusCode: () => StatusCode,
  StatusResponseError: () => StatusResponseError
});
module.exports = __toCommonJS(StatusCode_exports);
var import_general = require("#general");
var import_Status = require("../globals/Status.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const StatusCode = import_Status.Status;
class StatusResponseError extends import_general.MatterError {
  constructor(message, code, clusterCode) {
    super();
    this.code = code;
    this.clusterCode = clusterCode;
    this.message = `(${code}${clusterCode !== void 0 ? `/${clusterCode}` : ""}) ${message}`;
  }
  static is(error, ...codes) {
    return error instanceof StatusResponseError && (!codes.length || codes.includes(error.code));
  }
}
class ReceivedStatusResponseError extends StatusResponseError {
}
