// @matter/general/dist/esm/net/mock/NetworkSimulator.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Logger } from "../../log/Logger.js";
import { MockNetwork } from "./MockNetwork.js";
const logger = Logger.get("SimulatedNetwork");
class NetworkSimulator {
  listenersMap = /* @__PURE__ */ new Map();
  onUdpData(host, port, listener) {
    const ipPort = `${host ?? "*"}:${port}`;
    let listeners = this.listenersMap.get(ipPort);
    if (listeners === void 0) {
      listeners = new Array();
      this.listenersMap.set(ipPort, listeners);
    }
    listeners.push(listener);
    return {
      close: async () => this.offUdpData(host, port, listener)
    };
  }
  offUdpData(host, port, listenerToRemove) {
    const ipPort = `${host ?? "*"}:${port}`;
    const listeners = this.listenersMap.get(ipPort);
    if (listeners === void 0) return;
    const newListeners = listeners.filter((listener) => listener !== listenerToRemove);
    if (newListeners.length === 0) {
      this.listenersMap.delete(ipPort);
      return;
    }
    this.listenersMap.set(ipPort, newListeners);
  }
  sendUdp(localAddress, localPort, remoteAddress, remotePort, data) {
    [`${remoteAddress}:${remotePort}`, `*:${remotePort}`].forEach(
      (ipPort) => this.listenersMap.get(ipPort)?.forEach((listener) => {
        try {
          listener("fake0", localAddress, localPort, data);
        } catch (error) {
          logger.error(error);
        }
      })
    );
  }
  addHost(lastIdentifierByte) {
    const byte = lastIdentifierByte.toString(16).padStart(2, "0");
    return new MockNetwork(this, `00:11:22:33:44:${byte}`, [
      `1111:2222:3333:4444:5555:6666:7777:88${byte}`,
      `10.10.10.${lastIdentifierByte}`
    ]);
  }
}
export {
  NetworkSimulator
};
