// @matter/model/dist/cjs/standard/elements/AirPurifierDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AirPurifierDT_exports = {};
__export(AirPurifierDT_exports, {
  AirPurifierDt: () => AirPurifierDt
});
module.exports = __toCommonJS(AirPurifierDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AirPurifierDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "AirPurifier",
    id: 45,
    category: "HVAC",
    classification: "simple",
    details: "An Air Purifier is a standalone device that is designed to clean the air in a room.\nIt is a device that has a fan to control the air speed while it is operating. Optionally, it can report on the condition of its filters.",
    xref: { document: "device", section: "9.3" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 45, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OnOff",
    id: 6,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FanControl",
    id: 514,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "HepaFilterMonitoring",
    id: 113,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ActivatedCarbonFilterMonitoring",
    id: 114,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(AirPurifierDt);
