// @matter/model/dist/cjs/standard/elements/AirQualitySensorDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AirQualitySensorDT_exports = {};
__export(AirQualitySensorDT_exports, {
  AirQualitySensorDt: () => AirQualitySensorDt
});
module.exports = __toCommonJS(AirQualitySensorDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AirQualitySensorDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "AirQualitySensor",
    id: 44,
    category: "Sensor",
    classification: "simple",
    details: "This defines conformance for the Air Quality Sensor device type.\nAn air quality sensor is a device designed to monitor and measure various parameters related to the quality of ambient air in indoor or outdoor environments.",
    xref: { document: "device", section: "7.10" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 44, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "AirQuality",
    id: 91,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RelativeHumidityMeasurement",
    id: 1029,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "CarbonMonoxideConcentrationMeasurement",
    id: 1036,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "CarbonDioxideConcentrationMeasurement",
    id: 1037,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "NitrogenDioxideConcentrationMeasurement",
    id: 1043,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OzoneConcentrationMeasurement",
    id: 1045,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FormaldehydeConcentrationMeasurement",
    id: 1067,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Pm1ConcentrationMeasurement",
    id: 1068,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Pm25ConcentrationMeasurement",
    id: 1066,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Pm10ConcentrationMeasurement",
    id: 1069,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RadonConcentrationMeasurement",
    id: 1071,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TotalVolatileOrganicCompoundsConcentrationMeasurement",
    id: 1070,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(AirQualitySensorDt);
