// @matter/model/dist/cjs/standard/elements/AreaNamespaceNS.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AreaNamespaceNS_exports = {};
__export(AreaNamespaceNS_exports, {
  AreaNamespaceNs: () => AreaNamespaceNs
});
module.exports = __toCommonJS(AreaNamespaceNS_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AreaNamespaceNs = (0, import_elements.SemanticNamespaceElement)(
  {
    name: "AreaNamespace",
    id: 16,
    details: "The tags contained in this namespace may be used in any domain or context, to indicate an association with an indoor or outdoor area of a home.",
    xref: { document: "namespace", section: "13" }
  },
  (0, import_elements.SemanticTagElement)({ name: "Aisle", id: 0 }),
  (0, import_elements.SemanticTagElement)({ name: "Attic", id: 1 }),
  (0, import_elements.SemanticTagElement)({ name: "Back Door", id: 2 }),
  (0, import_elements.SemanticTagElement)({ name: "Back Yard", id: 3 }),
  (0, import_elements.SemanticTagElement)({ name: "Balcony", id: 4 }),
  (0, import_elements.SemanticTagElement)({ name: "Ballroom", id: 5 }),
  (0, import_elements.SemanticTagElement)({ name: "Bathroom", id: 6, description: "Also known as Restroom" }),
  (0, import_elements.SemanticTagElement)({ name: "Bedroom", id: 7 }),
  (0, import_elements.SemanticTagElement)({ name: "Border", id: 8 }),
  (0, import_elements.SemanticTagElement)({ name: "Boxroom", id: 9, description: "A small room typically used for storage" }),
  (0, import_elements.SemanticTagElement)({ name: "Breakfast Room", id: 10 }),
  (0, import_elements.SemanticTagElement)({ name: "Carport", id: 11 }),
  (0, import_elements.SemanticTagElement)({ name: "Cellar", id: 12 }),
  (0, import_elements.SemanticTagElement)({ name: "Cloakroom", id: 13 }),
  (0, import_elements.SemanticTagElement)({ name: "Closet", id: 14 }),
  (0, import_elements.SemanticTagElement)({ name: "Conservatory", id: 15 }),
  (0, import_elements.SemanticTagElement)({ name: "Corridor", id: 16 }),
  (0, import_elements.SemanticTagElement)({ name: "Craft Room", id: 17 }),
  (0, import_elements.SemanticTagElement)({ name: "Cupboard", id: 18 }),
  (0, import_elements.SemanticTagElement)({ name: "Deck", id: 19 }),
  (0, import_elements.SemanticTagElement)({
    name: "Den",
    id: 20,
    description: "A small, comfortable room for individual activities such as work or hobbies"
  }),
  (0, import_elements.SemanticTagElement)({ name: "Dining", id: 21 }),
  (0, import_elements.SemanticTagElement)({ name: "Drawing Room", id: 22 }),
  (0, import_elements.SemanticTagElement)({ name: "Dressing Room", id: 23 }),
  (0, import_elements.SemanticTagElement)({ name: "Driveway", id: 24 }),
  (0, import_elements.SemanticTagElement)({ name: "Elevator", id: 25 }),
  (0, import_elements.SemanticTagElement)({ name: "Ensuite", id: 26, description: "A bathroom directly accessible from a bedroom" }),
  (0, import_elements.SemanticTagElement)({ name: "Entrance", id: 27 }),
  (0, import_elements.SemanticTagElement)({ name: "Entryway", id: 28 }),
  (0, import_elements.SemanticTagElement)({ name: "Family Room", id: 29 }),
  (0, import_elements.SemanticTagElement)({ name: "Foyer", id: 30 }),
  (0, import_elements.SemanticTagElement)({ name: "Front Door", id: 31 }),
  (0, import_elements.SemanticTagElement)({ name: "Front Yard", id: 32 }),
  (0, import_elements.SemanticTagElement)({ name: "Game Room", id: 33 }),
  (0, import_elements.SemanticTagElement)({ name: "Garage", id: 34 }),
  (0, import_elements.SemanticTagElement)({ name: "Garage Door", id: 35 }),
  (0, import_elements.SemanticTagElement)({ name: "Garden", id: 36 }),
  (0, import_elements.SemanticTagElement)({ name: "Garden Door", id: 37 }),
  (0, import_elements.SemanticTagElement)({ name: "Guest Bathroom", id: 38, description: "Also known as Guest Restroom" }),
  (0, import_elements.SemanticTagElement)({ name: "Guest Bedroom", id: 39 }),
  (0, import_elements.SemanticTagElement)({ name: "Guest Room", id: 40 }),
  (0, import_elements.SemanticTagElement)({ name: "Gym", id: 41 }),
  (0, import_elements.SemanticTagElement)({ name: "Hallway", id: 42 }),
  (0, import_elements.SemanticTagElement)(
    { name: "Hearth Room", id: 43, description: "A cozy room containing a fireplace or other point heat source" }
  ),
  (0, import_elements.SemanticTagElement)({ name: "Kids Room", id: 44 }),
  (0, import_elements.SemanticTagElement)({ name: "Kids Bedroom", id: 45 }),
  (0, import_elements.SemanticTagElement)({ name: "Kitchen", id: 46 }),
  (0, import_elements.SemanticTagElement)({ name: "Laundry Room", id: 47 }),
  (0, import_elements.SemanticTagElement)({ name: "Lawn", id: 48 }),
  (0, import_elements.SemanticTagElement)({ name: "Library", id: 49 }),
  (0, import_elements.SemanticTagElement)({ name: "Living Room", id: 50 }),
  (0, import_elements.SemanticTagElement)({ name: "Lounge", id: 51 }),
  (0, import_elements.SemanticTagElement)({ name: "Media/TV Room", id: 52 }),
  (0, import_elements.SemanticTagElement)({
    name: "Mud Room",
    id: 53,
    description: "A space used to remove soiled garments prior to entering the domicile proper"
  }),
  (0, import_elements.SemanticTagElement)({ name: "Music Room", id: 54 }),
  (0, import_elements.SemanticTagElement)({ name: "Nursery", id: 55 }),
  (0, import_elements.SemanticTagElement)({ name: "Office", id: 56 }),
  (0, import_elements.SemanticTagElement)({ name: "Outdoor Kitchen", id: 57 }),
  (0, import_elements.SemanticTagElement)({ name: "Outside", id: 58 }),
  (0, import_elements.SemanticTagElement)({ name: "Pantry", id: 59, description: "AKA a larder, a place where food is stored" }),
  (0, import_elements.SemanticTagElement)({ name: "Parking Lot", id: 60 }),
  (0, import_elements.SemanticTagElement)({ name: "Parlor", id: 61 }),
  (0, import_elements.SemanticTagElement)({ name: "Patio", id: 62 }),
  (0, import_elements.SemanticTagElement)({ name: "Play Room", id: 63 }),
  (0, import_elements.SemanticTagElement)({ name: "Pool Room", id: 64 }),
  (0, import_elements.SemanticTagElement)({ name: "Porch", id: 65 }),
  (0, import_elements.SemanticTagElement)({ name: "Primary Bathroom", id: 66 }),
  (0, import_elements.SemanticTagElement)({ name: "Primary Bedroom", id: 67 }),
  (0, import_elements.SemanticTagElement)({ name: "Ramp", id: 68 }),
  (0, import_elements.SemanticTagElement)({ name: "Reception Room", id: 69 }),
  (0, import_elements.SemanticTagElement)({ name: "Recreation Room", id: 70 }),
  (0, import_elements.SemanticTagElement)({ name: "Roof", id: 71 }),
  (0, import_elements.SemanticTagElement)({ name: "Sauna", id: 72 }),
  (0, import_elements.SemanticTagElement)({ name: "Scullery", id: 73, description: "A utility space for cleaning dishes and laundry" }),
  (0, import_elements.SemanticTagElement)({ name: "Sewing Room", id: 74 }),
  (0, import_elements.SemanticTagElement)({ name: "Shed", id: 75 }),
  (0, import_elements.SemanticTagElement)({ name: "Side Door", id: 76 }),
  (0, import_elements.SemanticTagElement)({ name: "Side Yard", id: 77 }),
  (0, import_elements.SemanticTagElement)({ name: "Sitting Room", id: 78 }),
  (0, import_elements.SemanticTagElement)({
    name: "Snug",
    id: 79,
    description: "An informal space meant to be 'cozy', 'snug', relaxed, meant to share with family or friends"
  }),
  (0, import_elements.SemanticTagElement)({ name: "Spa", id: 80 }),
  (0, import_elements.SemanticTagElement)({ name: "Staircase", id: 81 }),
  (0, import_elements.SemanticTagElement)({ name: "Steam Room", id: 82 }),
  (0, import_elements.SemanticTagElement)({ name: "Storage Room", id: 83 }),
  (0, import_elements.SemanticTagElement)({ name: "Studio", id: 84 }),
  (0, import_elements.SemanticTagElement)({ name: "Study", id: 85 }),
  (0, import_elements.SemanticTagElement)({ name: "Sun Room", id: 86 }),
  (0, import_elements.SemanticTagElement)({ name: "Swimming Pool", id: 87 }),
  (0, import_elements.SemanticTagElement)({ name: "Terrace", id: 88 }),
  (0, import_elements.SemanticTagElement)({ name: "Toilet", id: 89, description: "A room dedicated to a toilet; a water closet / WC" }),
  (0, import_elements.SemanticTagElement)({ name: "Utility Room", id: 90 }),
  (0, import_elements.SemanticTagElement)({ name: "Ward", id: 91 }),
  (0, import_elements.SemanticTagElement)({ name: "Workshop", id: 92 })
);
import_MatterDefinition.MatterDefinition.children.push(AreaNamespaceNs);
