// @matter/model/dist/cjs/standard/elements/AtomicAttributeStatusStruct.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AtomicAttributeStatusStruct_exports = {};
__export(AtomicAttributeStatusStruct_exports, {
  AtomicAttributeStatusStruct: () => AtomicAttributeStatusStruct
});
module.exports = __toCommonJS(AtomicAttributeStatusStruct_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AtomicAttributeStatusStruct = (0, import_elements.DatatypeElement)(
  {
    name: "AtomicAttributeStatusStruct",
    type: "struct",
    details: "This struct indicates the status of an attribute during an atomic write.",
    xref: { document: "core", section: "7.15.5" }
  },
  (0, import_elements.FieldElement)({
    name: "AttributeId",
    id: 0,
    type: "attrib-id",
    conformance: "M",
    details: "This field shall indicate the ID of the attribute with the associated StatusCode.",
    xref: { document: "core", section: "7.15.5.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "StatusCode",
    id: 1,
    type: "status",
    conformance: "M",
    details: "This field shall indicate the atomic status of an attribute.",
    xref: { document: "core", section: "7.15.5.2" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(AtomicAttributeStatusStruct);
