// @matter/model/dist/cjs/standard/elements/BaseDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BaseDT_exports = {};
__export(BaseDT_exports, {
  BaseDt: () => BaseDt
});
module.exports = __toCommonJS(BaseDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const BaseDt = (0, import_elements.DeviceTypeElement)(
  { name: "Base", classification: "base", xref: { document: "device", section: "1.1" } },
  (0, import_elements.FieldElement)(
    { name: "conditions", type: "enum8" },
    (0, import_elements.FieldElement)({
      name: "Zha",
      description: "Zigbee Home Automation standard",
      xref: { document: "device", section: "1.1.3.1" }
    }),
    (0, import_elements.FieldElement)({ name: "Zse", description: "Zigbee Smart Energy standard", xref: { document: "device", section: "1.1.3.1" } }),
    (0, import_elements.FieldElement)({ name: "Gp", description: "Zigbee Green Power standard", xref: { document: "device", section: "1.1.3.1" } }),
    (0, import_elements.FieldElement)({ name: "Zigbee", description: "Zigbee standard", xref: { document: "device", section: "1.1.3.1" } }),
    (0, import_elements.FieldElement)({ name: "SuZi", description: "Zigbee PRO Sub-GHz standard", xref: { document: "device", section: "1.1.3.1" } }),
    (0, import_elements.FieldElement)({ name: "Matter", description: "Matter standard", xref: { document: "device", section: "1.1.3.1" } }),
    (0, import_elements.FieldElement)({
      name: "LanguageLocale",
      description: "The node supports localization for conveying text to the user",
      xref: { document: "device", section: "1.1.3.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "TimeLocale",
      description: "The node supports localization for conveying time to the user",
      xref: { document: "device", section: "1.1.3.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "UnitLocale",
      description: "The node supports localization for conveying units of measure to the user",
      xref: { document: "device", section: "1.1.3.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "Sit",
      description: "The node is a short idle time intermittently connected device",
      xref: { document: "device", section: "1.1.4" }
    }),
    (0, import_elements.FieldElement)({
      name: "Lit",
      description: "The node is a long idle time intermittently connected device",
      xref: { document: "device", section: "1.1.4" }
    }),
    (0, import_elements.FieldElement)({
      name: "Active",
      description: "The node is always able to communicate",
      xref: { document: "device", section: "1.1.4" }
    }),
    (0, import_elements.FieldElement)({ name: "Node", xref: { document: "device", section: "1.1.5" } }),
    (0, import_elements.FieldElement)({ name: "App", xref: { document: "device", section: "1.1.5" } }),
    (0, import_elements.FieldElement)({ name: "Simple", xref: { document: "device", section: "1.1.5" } }),
    (0, import_elements.FieldElement)({ name: "Dynamic", xref: { document: "device", section: "1.1.5" } }),
    (0, import_elements.FieldElement)({ name: "Composed", xref: { document: "device", section: "1.1.5" } }),
    (0, import_elements.FieldElement)({ name: "Client", xref: { document: "device", section: "1.1.6" } }),
    (0, import_elements.FieldElement)({ name: "Server", xref: { document: "device", section: "1.1.6" } }),
    (0, import_elements.FieldElement)({ name: "Duplicate", xref: { document: "device", section: "1.1.6" } }),
    (0, import_elements.FieldElement)({ name: "BridgedPowerSourceInfo", xref: { document: "device", section: "1.1.6" } })
  ),
  (0, import_elements.RequirementElement)(
    {
      name: "Descriptor",
      id: 29,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "1.1.7" }
    },
    (0, import_elements.RequirementElement)({ name: "TAGLIST", conformance: "Duplicate", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Binding",
    id: 30,
    conformance: "Simple & Client",
    element: "serverCluster",
    xref: { document: "device", section: "1.1.7" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FixedLabel",
    id: 64,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "1.1.7" }
  }),
  (0, import_elements.RequirementElement)({
    name: "UserLabel",
    id: 65,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "1.1.7" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(BaseDt);
