// @matter/model/dist/cjs/standard/elements/bool.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bool_exports = {};
__export(bool_exports, {
  bool: () => bool
});
module.exports = __toCommonJS(bool_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const bool = (0, import_elements.DatatypeElement)({
  name: "bool",
  description: "Boolean",
  isSeed: true,
  metatype: "boolean",
  details: "The Boolean type represents a logical value, either FALSE or TRUE.\n  \u2022 FALSE shall be equivalent to the value 0 (zero).\n  \u2022 TRUE shall be equivalent to the value 1 (one).",
  xref: { document: "core", section: "7.19.1.1" }
});
import_MatterDefinition.MatterDefinition.children.push(bool);
