// @matter/model/dist/cjs/standard/elements/BridgedNodeDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BridgedNodeDT_exports = {};
__export(BridgedNodeDT_exports, {
  BridgedNodeDt: () => BridgedNodeDt
});
module.exports = __toCommonJS(BridgedNodeDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const BridgedNodeDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "BridgedNode",
    id: 19,
    category: "Utility",
    classification: "utility",
    details: 'This defines conformance for a Bridged Node root endpoint. This endpoint is akin to a "read me first" endpoint that describes itself and any other endpoints that make up the Bridged Node. A Bridged Node endpoint represents a device on a foreign network, but is not the root endpoint of the bridge itself.',
    xref: { document: "device", section: "2.5" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 19, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "BridgedDeviceBasicInformation",
    id: 57,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PowerSourceConfiguration",
    id: 46,
    conformance: "BridgedPowerSourceInfo, D",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PowerSource",
    id: 47,
    conformance: "BridgedPowerSourceInfo",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "EcosystemInformation",
    id: 1872,
    conformance: "FabricSynchronizedNode, O",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "AdministratorCommissioning",
    id: 60,
    conformance: "FabricSynchronizedNode",
    element: "serverCluster",
    xref: { document: "device", section: "2.5.5" }
  }),
  (0, import_elements.FieldElement)(
    { name: "conditions", type: "enum8" },
    (0, import_elements.FieldElement)({
      name: "FabricSynchronizedNode",
      description: "See description below.",
      xref: { document: "device", section: "2.5.3" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(BridgedNodeDt);
