// @matter/model/dist/cjs/standard/elements/ContentLauncher.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ContentLauncher_exports = {};
__export(ContentLauncher_exports, {
  ContentLauncher: () => ContentLauncher
});
module.exports = __toCommonJS(ContentLauncher_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ContentLauncher = (0, import_elements.ClusterElement)(
  {
    name: "ContentLauncher",
    id: 1290,
    classification: "application",
    pics: "CONTENTLAUNCHER",
    details: "This cluster provides an interface for launching content on a Video Player device such as a Streaming Media Player, Smart TV or Smart Screen.\nThis cluster would be supported on a Video Player device or devices that can playback content, such as a Streaming Media Player, Smart TV or Smart Screen. This cluster supports playing back content referenced by URL. It supports finding content by type and global identifier, and either playing the content or displaying the search results.\nThe cluster server for Content Launcher is implemented by an endpoint that can launch content, such as a Video Player, or an endpoint representing a Content App on such a device.\nWhen this cluster is implemented for an Content App Endpoint (Endpoint with type \u201CContent App\u201D and having an Application Basic cluster), the Video Player device shall launch the application when a client invokes the LaunchContent or LaunchURL commands.",
    xref: { document: "cluster", section: "6.7" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap", xref: { document: "cluster", section: "6.7.4" } },
    (0, import_elements.FieldElement)({
      name: "CS",
      constraint: "0",
      description: "ContentSearch",
      details: "Device supports content search (non-app specific)"
    }),
    (0, import_elements.FieldElement)({
      name: "UP",
      constraint: "1",
      description: "UrlPlayback",
      details: "Device supports basic URL-based file playback"
    }),
    (0, import_elements.FieldElement)({
      name: "AS",
      constraint: "2",
      description: "AdvancedSeek",
      details: 'Enables clients to implement more advanced media seeking behavior in their user interface, such as for example a "seek bar".'
    }),
    (0, import_elements.FieldElement)({ name: "TT", constraint: "3", description: "TextTracks", details: "Device or app supports Text Tracks." }),
    (0, import_elements.FieldElement)({ name: "AT", constraint: "4", description: "AudioTracks", details: "Device or app supports Audio Tracks." })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "AcceptHeader",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "UP",
      constraint: "max 100[max 1024]",
      default: [],
      quality: "N",
      details: 'This attribute shall provide a list of content types supported by the Video Player or Content App\nin the form of entries in the HTTP "Accept" request header.',
      xref: { document: "cluster", section: "6.7.6.1" }
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "string" })
  ),
  (0, import_elements.AttributeElement)({
    name: "SupportedStreamingProtocols",
    id: 1,
    type: "SupportedProtocolsBitmap",
    access: "R V",
    conformance: "UP",
    default: 0,
    quality: "N",
    details: "This attribute shall provide information about supported streaming protocols.",
    xref: { document: "cluster", section: "6.7.6.2" }
  }),
  (0, import_elements.CommandElement)(
    {
      name: "LaunchContent",
      id: 0,
      access: "O",
      conformance: "CS",
      direction: "request",
      response: "LauncherResponse",
      details: "Upon receipt, this shall launch the specified content with optional search criteria. This command returns a Launch Response.",
      xref: { document: "cluster", section: "6.7.7.1" }
    },
    (0, import_elements.FieldElement)({
      name: "Search",
      id: 0,
      type: "ContentSearchStruct",
      conformance: "M",
      constraint: "desc",
      details: "This field shall indicate the content to launch.",
      xref: { document: "cluster", section: "6.7.7.1.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "AutoPlay",
      id: 1,
      type: "bool",
      conformance: "M",
      constraint: "desc",
      details: "This field shall indicate whether to automatically start playing content, where:\n  \u2022 TRUE means best match should start playing automatically.\n  \u2022 FALSE means matches should be displayed on screen for user selection.",
      xref: { document: "cluster", section: "6.7.7.1.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Data",
      id: 2,
      type: "string",
      conformance: "O",
      details: "This field, if present, shall indicate app-specific data.",
      xref: { document: "cluster", section: "6.7.7.1.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "PlaybackPreferences",
      id: 3,
      type: "PlaybackPreferencesStruct",
      conformance: "O",
      details: "This field, if present, shall indicate the user\u2019s preferred Text/AudioTracks and playbackPosition for the media, sent from the client to the server. If the server does not find an available track for the title being played exactly matching a Track requested here, in the list of available tracks, it may default to picking another track that closely matches the requested track. Alternately, it may go with user preferences set on the server side (it will use this option if these PlaybackPreferences are not specified). In the case of text tracks, that may mean that the subtitle text is not displayed at all. In the cases where the preferred Text/AudioTracks are not available, the server shall return the TextTrackNotAvailable and/or AudioTrackNotAvailable Status(es) in the LauncherResponse.",
      xref: { document: "cluster", section: "6.7.7.1.4" }
    }),
    (0, import_elements.FieldElement)({
      name: "UseCurrentContext",
      id: 4,
      type: "bool",
      conformance: "O",
      constraint: "desc",
      default: true,
      details: "This field, if present, shall indicate whether to consider the context of current ongoing activity on the receiver to fulfill the request. For example if the request only includes data in ContentSearch that specifies an Episode number, and UseCurrentContent is set to TRUE, if there is a TV series on going, the request refers to the specific episode of the ongoing season of the TV series. TRUE means current activity context may be considered FALSE means current activity context shall NOT be considered",
      xref: { document: "cluster", section: "6.7.7.1.5" }
    })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "LaunchUrl",
      id: 1,
      access: "O",
      conformance: "UP",
      direction: "request",
      response: "LauncherResponse",
      details: "Upon receipt, this shall launch content from the specified URL.\nThe content types supported include those identified in the AcceptHeader and SupportedStreamingProtocols attributes.\nA check shall be made to ensure the URL is secure (uses HTTPS).\nWhen playing a video stream in response to this command, an indication (ex. visual) of the identity of the origin node of the video stream shall be provided. This could be in the form of a friendly name label which uniquely identifies the node to the user. This friendly name label is typically assigned by the Matter Admin (ex. TV) at the time of commissioning and, when it\u2019s a device, is often editable by the user. It might be a combination of a company name and friendly name, for example, \u201DAcme\u201D or \u201CAcme Streaming Service on Alice\u2019s Phone\u201D.\nThis command returns a Launch Response.",
      xref: { document: "cluster", section: "6.7.7.2" }
    },
    (0, import_elements.FieldElement)({
      name: "ContentUrl",
      id: 0,
      type: "string",
      conformance: "M",
      constraint: "any",
      details: "This field shall indicate the URL of content to launch. The syntax of this field shall follow the syntax as specified in RFC 1738 and shall use the https scheme.",
      xref: { document: "cluster", section: "6.7.7.2.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "DisplayString",
      id: 1,
      type: "string",
      conformance: "O",
      constraint: "any",
      details: "This field, if present, shall provide a string that may be used to describe the content being accessed at the given URL.",
      xref: { document: "cluster", section: "6.7.7.2.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "BrandingInformation",
      id: 2,
      type: "BrandingInformationStruct",
      conformance: "O",
      constraint: "any",
      details: "This field, if present, shall indicate the branding information that may be displayed when playing back the given content.",
      xref: { document: "cluster", section: "6.7.7.2.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "PlaybackPreferences",
      id: 3,
      type: "PlaybackPreferencesStruct",
      conformance: "O",
      constraint: "any",
      details: "This field, if present, shall indicate the user\u2019s preferred Text/AudioTracks and playbackPosition for the media, sent from the client to the server. If the server does not find an available track for the title being played exactly matching a Track requested here, in the list of available tracks, it may default to picking another track that closely matches the requested track. Alternately, it may go with user preferences set on the server side (it will use this option if these PlaybackPreferences are not specified). In the case of text tracks, that may mean that the subtitle text is not displayed at all. In the cases where the preferred Text/AudioTracks are not available, the server shall return the TextTrackNotAvailable and/or AudioTrackNotAvailable Status(es) in the LauncherResponse.",
      xref: { document: "cluster", section: "6.7.7.2.4" }
    })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "LauncherResponse",
      id: 2,
      conformance: "CS | UP",
      direction: "response",
      details: "This command shall be generated in response to LaunchContent and LaunchURL commands.",
      xref: { document: "cluster", section: "6.7.7.3" }
    },
    (0, import_elements.FieldElement)({
      name: "Status",
      id: 0,
      type: "StatusEnum",
      conformance: "M",
      details: "This field shall indicate the status of the command which resulted in this response.",
      xref: { document: "cluster", section: "6.7.7.3.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Data",
      id: 1,
      type: "string",
      conformance: "O",
      details: "This field shall indicate Optional app-specific data.",
      xref: { document: "cluster", section: "6.7.7.3.2" }
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "SupportedProtocolsBitmap", type: "map32", xref: { document: "cluster", section: "6.7.5.1" } },
    (0, import_elements.FieldElement)({ name: "Dash", constraint: "0", description: "Device supports Dynamic Adaptive Streaming over HTTP (DASH)" }),
    (0, import_elements.FieldElement)({ name: "Hls", constraint: "1", description: "Device supports HTTP Live Streaming (HLS)" })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "StatusEnum", type: "enum8", xref: { document: "cluster", section: "6.7.5.2" } },
    (0, import_elements.FieldElement)({ name: "Success", id: 0, conformance: "M", description: "Command succeeded" }),
    (0, import_elements.FieldElement)({
      name: "UrlNotAvailable",
      id: 1,
      conformance: "M",
      description: "Requested URL could not be reached by device."
    }),
    (0, import_elements.FieldElement)({ name: "AuthFailed", id: 2, conformance: "M", description: "Requested URL returned 401 error code." }),
    (0, import_elements.FieldElement)({
      name: "TextTrackNotAvailable",
      id: 3,
      conformance: "TT",
      description: "Requested Text Track (in PlaybackPreferences) not available"
    }),
    (0, import_elements.FieldElement)({
      name: "AudioTrackNotAvailable",
      id: 4,
      conformance: "AT",
      description: "Requested Audio Track (in PlaybackPreferences) not available"
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "ParameterEnum", type: "enum8", xref: { document: "cluster", section: "6.7.5.3" } },
    (0, import_elements.FieldElement)({
      name: "Actor",
      id: 0,
      conformance: "M",
      description: "Actor represents an actor credited in video media content; for example, \u201CGaby Hoffman\u201D"
    }),
    (0, import_elements.FieldElement)({
      name: "Channel",
      id: 1,
      conformance: "M",
      description: 'Channel represents the identifying data for a television channel; for example, "PBS"'
    }),
    (0, import_elements.FieldElement)({
      name: "Character",
      id: 2,
      conformance: "M",
      description: "A character represented in video media content; for example, \u201CSnow White\u201D"
    }),
    (0, import_elements.FieldElement)({
      name: "Director",
      id: 3,
      conformance: "M",
      description: "A director of the video media content; for example, \u201CSpike Lee\u201D"
    }),
    (0, import_elements.FieldElement)({
      name: "Event",
      id: 4,
      conformance: "M",
      description: `An event is a reference to a type of event; examples would include sports, music, or other types of events. For example, searching for "Football games" would search for a 'game' event entity and a 'football' sport entity.`
    }),
    (0, import_elements.FieldElement)({
      name: "Franchise",
      id: 5,
      conformance: "M",
      description: `A franchise is a video entity which can represent a number of video entities, like movies or TV shows. For example, take the fictional franchise "Intergalactic Wars" which represents a collection of movie trilogies, as well as animated and live action TV shows. This entity type was introduced to account for requests by customers such as "Find Intergalactic Wars movies", which would search for all 'Intergalactic Wars' programs of the MOVIE MediaType, rather than attempting to match to a single title.`
    }),
    (0, import_elements.FieldElement)({
      name: "Genre",
      id: 6,
      conformance: "M",
      description: "Genre represents the genre of video media content such as action, drama or comedy."
    }),
    (0, import_elements.FieldElement)({
      name: "League",
      id: 7,
      conformance: "M",
      description: 'League represents the categorical information for a sporting league; for example, "NCAA"'
    }),
    (0, import_elements.FieldElement)({
      name: "Popularity",
      id: 8,
      conformance: "M",
      description: "Popularity indicates whether the user asks for popular content."
    }),
    (0, import_elements.FieldElement)({
      name: "Provider",
      id: 9,
      conformance: "M",
      description: 'The provider (MSP) the user wants this media to be played on; for example, "Netflix".'
    }),
    (0, import_elements.FieldElement)({
      name: "Sport",
      id: 10,
      conformance: "M",
      description: "Sport represents the categorical information of a sport; for example, football"
    }),
    (0, import_elements.FieldElement)({
      name: "SportsTeam",
      id: 11,
      conformance: "M",
      description: 'SportsTeam represents the categorical information of a professional sports team; for example, "University of Washington Huskies"'
    }),
    (0, import_elements.FieldElement)({
      name: "Type",
      id: 12,
      conformance: "M",
      description: 'The type of content requested. Supported types are "Movie", "MovieSeries", "TVSeries", "TVSeason", "TVEpisode", "Trailer", "SportsEvent", "LiveEvent", and "Video"'
    }),
    (0, import_elements.FieldElement)({
      name: "Video",
      id: 13,
      conformance: "M",
      description: 'Video represents the identifying data for a specific piece of video content; for example, "Manchester by the Sea".'
    }),
    (0, import_elements.FieldElement)({
      name: "Season",
      id: 14,
      conformance: "O",
      description: "Season represents the specific season number within a TV series."
    }),
    (0, import_elements.FieldElement)({
      name: "Episode",
      id: 15,
      conformance: "O",
      description: "Episode represents a specific episode number within a Season in a TV series."
    }),
    (0, import_elements.FieldElement)({
      name: "Any",
      id: 16,
      conformance: "O",
      description: "Represents a search text input across many parameter types or even outside of the defined param types."
    })
  ),
  (0, import_elements.DatatypeElement)(
    { name: "MetricTypeEnum", type: "enum8", xref: { document: "cluster", section: "6.7.5.4" } },
    (0, import_elements.FieldElement)({
      name: "Pixels",
      id: 0,
      conformance: "M",
      description: "Dimensions defined in a number of Pixels",
      details: "This value is used for dimensions defined in a number of Pixels.",
      xref: { document: "cluster", section: "6.7.5.4.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Percentage",
      id: 1,
      conformance: "M",
      description: "Dimensions defined as a percentage",
      details: 'This value is for dimensions defined as a percentage of the overall display dimensions. For example, if using a Percentage Metric type for a Width measurement of 50.0, against a display width of 1920 pixels, then the resulting value used would be 960 pixels (50.0% of 1920) for that dimension. Whenever a measurement uses this Metric type, the resulting values shall be rounded ("floored") towards 0 if the measurement requires an integer final value.',
      xref: { document: "cluster", section: "6.7.5.4.2" }
    })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "AdditionalInfoStruct",
      type: "struct",
      details: "This object defines additional name=value pairs that can be used for identifying content.",
      xref: { document: "cluster", section: "6.7.5.5" }
    },
    (0, import_elements.FieldElement)({
      name: "Name",
      id: 0,
      type: "string",
      conformance: "M",
      constraint: "max 256",
      details: 'This field shall indicate the name of external id, ex. "musicbrainz".',
      xref: { document: "cluster", section: "6.7.5.5.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Value",
      id: 1,
      type: "string",
      conformance: "M",
      constraint: "max 8192",
      details: 'This field shall indicate the value for external id, ex. "ST0000000666661".',
      xref: { document: "cluster", section: "6.7.5.5.2" }
    })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "ParameterStruct",
      type: "struct",
      details: "This object defines inputs to a search for content for display or playback.",
      xref: { document: "cluster", section: "6.7.5.6" }
    },
    (0, import_elements.FieldElement)({
      name: "Type",
      id: 0,
      type: "ParameterEnum",
      conformance: "M",
      details: "This field shall indicate the entity type.",
      xref: { document: "cluster", section: "6.7.5.6.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Value",
      id: 1,
      type: "string",
      conformance: "M",
      constraint: "max 1024",
      details: "This field shall indicate the entity value, which is a search string, ex. \u201CManchester by the Sea\u201D.",
      xref: { document: "cluster", section: "6.7.5.6.2" }
    }),
    (0, import_elements.FieldElement)(
      {
        name: "ExternalIdList",
        id: 2,
        type: "list",
        conformance: "O",
        default: [],
        details: "This field shall indicate the list of additional external content identifiers.",
        xref: { document: "cluster", section: "6.7.5.6.3" }
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "AdditionalInfoStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "ContentSearchStruct",
      type: "struct",
      details: "This object defines inputs to a search for content for display or playback.",
      xref: { document: "cluster", section: "6.7.5.7" }
    },
    (0, import_elements.FieldElement)(
      {
        name: "ParameterList",
        id: 0,
        type: "list",
        conformance: "M",
        details: "This field shall indicate the list of parameters comprising the search. If multiple parameters are provided, the search parameters shall be joined with 'AND' logic. e.g. action movies with Tom Cruise will be represented as [{Actor: 'Tom Cruise'}, {Type: 'Movie'}, {Genre: 'Action'}]",
        xref: { document: "cluster", section: "6.7.5.7.1" }
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "ParameterStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "DimensionStruct",
      type: "struct",
      details: "This object defines dimension which can be used for defining Size of background images.",
      xref: { document: "cluster", section: "6.7.5.8" }
    },
    (0, import_elements.FieldElement)({
      name: "Width",
      id: 0,
      type: "double",
      conformance: "M",
      details: "This field shall indicate the width using the metric defined in Metric",
      xref: { document: "cluster", section: "6.7.5.8.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Height",
      id: 1,
      type: "double",
      conformance: "M",
      details: "This field shall indicate the height using the metric defined in Metric",
      xref: { document: "cluster", section: "6.7.5.8.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Metric",
      id: 2,
      type: "MetricTypeEnum",
      conformance: "M",
      details: "This field shall indicate metric used for defining Height/Width.",
      xref: { document: "cluster", section: "6.7.5.8.3" }
    })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "StyleInformationStruct",
      type: "struct",
      details: "This object defines style information which can be used by content providers to change the Media Player\u2019s style related properties.",
      xref: { document: "cluster", section: "6.7.5.9" }
    },
    (0, import_elements.FieldElement)({
      name: "ImageUrl",
      id: 0,
      type: "string",
      conformance: "O",
      constraint: "max 8192",
      details: "This field shall indicate the URL of image used for Styling different Video Player sections like Logo, Watermark etc. The syntax of this field shall follow the syntax as specified in RFC 1738 and shall use the https scheme.",
      xref: { document: "cluster", section: "6.7.5.9.1" }
    }),
    (0, import_elements.FieldElement)(
      {
        name: "Color",
        id: 1,
        type: "string",
        conformance: "O",
        constraint: "7, 9",
        details: "This field shall indicate the color, in RGB or RGBA, used for styling different Video Player sections like Logo, Watermark, etc. The value shall conform to the 6-digit or 8-digit format defined for CSS sRGB hexadecimal color notation [https://www.w3.org/TR/css-color-4/#hex-notation]. Examples:\n  \u2022 #76DE19 for R=0x76, G=0xDE, B=0x19, A absent\n  \u2022 #76DE1980 for R=0x76, G=0xDE, B=0x19, A=0x80",
        xref: { document: "cluster", section: "6.7.5.9.2" }
      }
    ),
    (0, import_elements.FieldElement)({
      name: "Size",
      id: 2,
      type: "DimensionStruct",
      conformance: "O",
      details: "This field shall indicate the size of the image used for Styling different Video Player sections like\nLogo, Watermark etc.",
      xref: { document: "cluster", section: "6.7.5.9.3" }
    })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "BrandingInformationStruct",
      type: "struct",
      details: "This object defines Branding Information which can be provided by the client in order to customize the skin of the Video Player during playback.",
      xref: { document: "cluster", section: "6.7.5.10" }
    },
    (0, import_elements.FieldElement)({
      name: "ProviderName",
      id: 0,
      type: "string",
      conformance: "M",
      constraint: "max 256",
      details: "This field shall indicate name of the provider for the given content.",
      xref: { document: "cluster", section: "6.7.5.10.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Background",
      id: 1,
      type: "StyleInformationStruct",
      conformance: "O",
      details: "This field shall indicate background of the Video Player while content launch request is being processed by it. This background information may also be used by the Video Player when it is in idle state.",
      xref: { document: "cluster", section: "6.7.5.10.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Logo",
      id: 2,
      type: "StyleInformationStruct",
      conformance: "O",
      details: "This field shall indicate the logo shown when the Video Player is launching. This is also used when the Video Player is in the idle state and Splash field is not available.",
      xref: { document: "cluster", section: "6.7.5.10.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "ProgressBar",
      id: 3,
      type: "StyleInformationStruct",
      conformance: "O",
      details: "This field shall indicate the style of progress bar for media playback.",
      xref: { document: "cluster", section: "6.7.5.10.4" }
    }),
    (0, import_elements.FieldElement)({
      name: "Splash",
      id: 4,
      type: "StyleInformationStruct",
      conformance: "O",
      details: "This field shall indicate the screen shown when the Video Player is in an idle state. If this property is not populated, the Video Player shall default to logo or the provider name.",
      xref: { document: "cluster", section: "6.7.5.10.5" }
    }),
    (0, import_elements.FieldElement)({
      name: "WaterMark",
      id: 5,
      type: "StyleInformationStruct",
      conformance: "O",
      details: "This field shall indicate watermark shown when the media is playing.",
      xref: { document: "cluster", section: "6.7.5.10.6" }
    })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "PlaybackPreferencesStruct",
      type: "struct",
      details: "PlaybackPreferencesStruct defines the preferences sent by the client to the receiver in the ContentLauncher LaunchURL or LaunchContent commands.",
      xref: { document: "cluster", section: "6.7.5.11" }
    },
    (0, import_elements.FieldElement)({
      name: "PlaybackPosition",
      id: 0,
      type: "uint64",
      conformance: "AS",
      quality: "X",
      details: "This field shall indicate the preferred position (in milliseconds) in the media to launch playback from. In case the position falls in the middle of a frame, the server shall set the position to the beginning of that frame and set the SampledPosition attribute on the MediaPlayback cluster accordingly. A value of null shall indicate that playback position is not applicable for the current state of the media playback. (For example : Live media with no known duration and where seek is not supported).",
      xref: { document: "cluster", section: "6.7.5.11.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "TextTrack",
      id: 1,
      type: "TrackPreferenceStruct",
      conformance: "TT",
      quality: "X",
      details: "This field shall indicate the user\u2019s preferred Text Track. A value of null shall indicate that the user did not specify a preferred Text Track on the client. In such a case, the decision to display and select a Text Track is up to the server.",
      xref: { document: "cluster", section: "6.7.5.11.2" }
    }),
    (0, import_elements.FieldElement)(
      {
        name: "AudioTracks",
        id: 2,
        type: "list",
        conformance: "AT",
        quality: "X",
        details: "This field shall indicate the list of the user\u2019s preferred Audio Tracks. If the list contains multiple values, each AudioTrack must also specify a unique audioOutputIndex to play the track on. A value of null shall indicate that the user did not specify a preferred Audio Track on the client. In such a case, the decision to play and select an Audio Track is up to the server.",
        xref: { document: "cluster", section: "6.7.5.11.3" }
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "TrackPreferenceStruct" })
    )
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "TrackPreferenceStruct",
      type: "struct",
      details: "This structure defines Text/Audio Track preferences.",
      xref: { document: "cluster", section: "6.7.5.12" }
    },
    (0, import_elements.FieldElement)({
      name: "LanguageCode",
      id: 0,
      type: "string",
      conformance: "M",
      constraint: "max 32",
      details: "This field shall contain one of the standard Tags for Identifying Languages RFC 5646, which identifies the primary language used in the Track.",
      xref: { document: "cluster", section: "6.7.5.12.1" }
    }),
    (0, import_elements.FieldElement)(
      {
        name: "Characteristics",
        id: 1,
        type: "list",
        conformance: "O",
        default: null,
        quality: "X",
        details: "This field shall contain a list of enumerated CharacteristicEnum values that indicate a purpose, trait or feature associated with the Track. A value of null shall indicate that there are no Characteristics corresponding to the Track.",
        xref: { document: "cluster", section: "6.7.5.12.2" }
      },
      (0, import_elements.FieldElement)({ name: "entry", type: "MediaPlayback.CharacteristicEnum" })
    ),
    (0, import_elements.FieldElement)({
      name: "AudioOutputIndex",
      id: 2,
      type: "uint8",
      conformance: "AT",
      quality: "X",
      details: "This field if present shall indicate the index of the OutputInfoStruct from the OutputList attribute (from the AudioOutput cluster) and indicates which audio output the Audio Track should be played on.\nThis field shall NOT be present if the track is not an audio track.\nIf the track is an audio track, this field MUST be present. A value of null shall indicate that the server can choose the audio output(s) to play the Audio Track on.",
      xref: { document: "cluster", section: "6.7.5.12.3" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ContentLauncher);
