// @matter/model/dist/cjs/standard/elements/ContactSensorDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ContactSensorDT_exports = {};
__export(ContactSensorDT_exports, {
  ContactSensorDt: () => ContactSensorDt
});
module.exports = __toCommonJS(ContactSensorDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ContactSensorDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "ContactSensor",
    id: 21,
    category: "Sensor",
    classification: "simple",
    details: "This defines conformance to the Contact Sensor device type.",
    xref: { document: "device", section: "7.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 21, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "BooleanState",
    id: 69,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "BooleanStateConfiguration",
    id: 128,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.1.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(ContactSensorDt);
