// @matter/model/dist/cjs/standard/elements/ContentAppDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ContentAppDT_exports = {};
__export(ContentAppDT_exports, {
  ContentAppDt: () => ContentAppDt
});
module.exports = __toCommonJS(ContentAppDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ContentAppDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "ContentApp",
    id: 36,
    category: "Media",
    classification: "simple",
    details: "This defines conformance to the Content App device type.\nA Content App is usually an application built by a Content Provider. A Casting Video Player with a Content App Platform is able to launch Content Apps and represent these apps as separate endpoints.",
    xref: { document: "device", section: "10.5" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 36, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Binding",
    id: 30,
    conformance: "ObserverClient",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Channel",
    id: 1284,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TargetNavigator",
    id: 1285,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "MediaPlayback",
    id: 1286,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "KeypadInput",
    id: 1289,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ContentLauncher",
    id: 1290,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "ApplicationLauncher",
      id: 1292,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "10.5.4" }
    },
    (0, import_elements.RequirementElement)({ name: "APPLICATIONPLATFORM", conformance: "X", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "ApplicationBasic",
    id: 1293,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "AccountLogin",
    id: 1294,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ContentAppObserver",
    id: 1296,
    conformance: "ObserverClient",
    element: "clientCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  (0, import_elements.FieldElement)(
    { name: "conditions", type: "enum8" },
    (0, import_elements.FieldElement)({
      name: "ObserverClient",
      description: "The node is a client for ContentAppObservers.",
      xref: { document: "device", section: "10.5.3" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ContentAppDt);
