// @matter/model/dist/cjs/standard/elements/CookSurfaceDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CookSurfaceDT_exports = {};
__export(CookSurfaceDT_exports, {
  CookSurfaceDt: () => CookSurfaceDt
});
module.exports = __toCommonJS(CookSurfaceDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const CookSurfaceDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "CookSurface",
    id: 119,
    category: "Appliances",
    classification: "simple",
    details: "A Cook Surface device type represents a heating object on a cooktop or other similar device. It shall only be used when composed as part of another device type.",
    xref: { document: "device", section: "13.7" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 119, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "TemperatureControl",
    id: 86,
    conformance: "O.a+",
    element: "serverCluster",
    xref: { document: "device", section: "13.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O.a+",
    element: "serverCluster",
    xref: { document: "device", section: "13.7.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "OnOff",
      id: 6,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.7.4" }
    },
    (0, import_elements.RequirementElement)({ name: "OFFONLY", conformance: "M", element: "feature" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(CookSurfaceDt);
