// @matter/model/dist/cjs/standard/elements/date.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var date_exports = {};
__export(date_exports, {
  date: () => date
});
module.exports = __toCommonJS(date_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const date = (0, import_elements.DatatypeElement)(
  {
    name: "date",
    type: "struct",
    description: "Date",
    isSeed: true,
    metatype: "date",
    details: "This data type shall be a struct as defined below.",
    xref: { document: "core", section: "7.19.2.4" }
  },
  (0, import_elements.FieldElement)({
    name: "Year",
    id: 0,
    type: "uint8",
    conformance: "M",
    default: null,
    quality: "X",
    details: "The year subfield represents years from 1900 (0) to 2155 (255).",
    xref: { document: "core", section: "7.19.2.4.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "Month",
    id: 1,
    type: "uint8",
    conformance: "M",
    constraint: "1 to 12",
    default: null,
    quality: "X",
    details: "This field represents months January (1) to December (12).",
    xref: { document: "core", section: "7.19.2.4.2" }
  }),
  (0, import_elements.FieldElement)({
    name: "Day",
    id: 2,
    type: "uint8",
    conformance: "M",
    constraint: "1 to 31",
    default: null,
    quality: "X",
    details: "This field represents the day of the month. Note that values in the range 29 to 31 may be invalid, depending on the month and year.",
    xref: { document: "core", section: "7.19.2.4.3" }
  }),
  (0, import_elements.FieldElement)({
    name: "DayOfWeek",
    id: 3,
    type: "uint8",
    conformance: "M",
    constraint: "1 to 7",
    default: null,
    quality: "X",
    details: "This represents the day of the week from Monday (1) to Sunday (7).",
    xref: { document: "core", section: "7.19.2.4.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(date);
