// @matter/model/dist/cjs/standard/elements/DishwasherAlarm.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DishwasherAlarm_exports = {};
__export(DishwasherAlarm_exports, {
  DishwasherAlarm: () => DishwasherAlarm
});
module.exports = __toCommonJS(DishwasherAlarm_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DishwasherAlarm = (0, import_elements.ClusterElement)(
  {
    name: "DishwasherAlarm",
    id: 93,
    type: "AlarmBase",
    classification: "application",
    pics: "DISHALM",
    details: "This cluster is a derived cluster of the Alarm Base cluster and provides the alarm definition related to dishwasher devices.",
    xref: { document: "cluster", section: "8.4" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.DatatypeElement)(
    { name: "AlarmBitmap", type: "map32", xref: { document: "cluster", section: "8.4.4.1" } },
    (0, import_elements.FieldElement)({ name: "InflowError", constraint: "0", description: "Water inflow is abnormal" }),
    (0, import_elements.FieldElement)({ name: "DrainError", constraint: "1", description: "Water draining is abnormal" }),
    (0, import_elements.FieldElement)({ name: "DoorError", constraint: "2", description: "Door or door lock is abnormal" }),
    (0, import_elements.FieldElement)({ name: "TempTooLow", constraint: "3", description: "Unable to reach normal temperature" }),
    (0, import_elements.FieldElement)({ name: "TempTooHigh", constraint: "4", description: "Temperature is too high" }),
    (0, import_elements.FieldElement)({ name: "WaterLevelError", constraint: "5", description: "Water level is abnormal" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(DishwasherAlarm);
