// @matter/model/dist/cjs/standard/elements/DishwasherDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DishwasherDT_exports = {};
__export(DishwasherDT_exports, {
  DishwasherDt: () => DishwasherDt
});
module.exports = __toCommonJS(DishwasherDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DishwasherDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "Dishwasher",
    id: 117,
    category: "Appliances",
    classification: "simple",
    details: "A dishwasher is a device that is generally installed in residential homes and is capable of washing dishes, cutlery, and other items associate with food preparation and consumption. The device can be permanently installed or portable and can have variety of filling and draining methods.",
    xref: { document: "device", section: "13.5" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 117, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.5.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "OnOff",
      id: 6,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.5.4" }
    },
    (0, import_elements.RequirementElement)({ name: "DEADFRONTBEHAVIOR", conformance: "M", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "TemperatureControl",
    id: 86,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.5.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "DishwasherMode",
      id: 89,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.5.4" }
    },
    (0, import_elements.RequirementElement)({ name: "StartUpMode", conformance: "X", element: "attribute" }),
    (0, import_elements.RequirementElement)({ name: "ONOFF", conformance: "X", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "DishwasherAlarm",
    id: 93,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OperationalState",
    id: 96,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.5.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(DishwasherDt);
