// @matter/model/dist/cjs/standard/elements/DoorLockControllerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DoorLockControllerDT_exports = {};
__export(DoorLockControllerDT_exports, {
  DoorLockControllerDt: () => DoorLockControllerDt
});
module.exports = __toCommonJS(DoorLockControllerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DoorLockControllerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "DoorLockController",
    id: 11,
    category: "Closure",
    classification: "simple",
    details: "A Door Lock Controller is a device capable of controlling a door lock.",
    xref: { document: "device", section: "8.2" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 11, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "8.2.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ScenesManagement",
    id: 98,
    conformance: "P, O",
    element: "clientCluster",
    xref: { document: "device", section: "8.2.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TimeSynchronization",
    id: 56,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "8.2.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "DoorLock",
    id: 257,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "8.2.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(DoorLockControllerDt);
