// @matter/model/dist/cjs/standard/elements/DoorLockDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DoorLockDT_exports = {};
__export(DoorLockDT_exports, {
  DoorLockDt: () => DoorLockDt
});
module.exports = __toCommonJS(DoorLockDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DoorLockDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "DoorLock",
    id: 10,
    category: "Closure",
    classification: "simple",
    details: "A Door Lock is a device used to secure a door. It is possible to actuate a door lock either by means of a manual or a remote method.",
    xref: { document: "device", section: "8.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 10, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "8.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "X",
    element: "serverCluster",
    xref: { document: "device", section: "8.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ScenesManagement",
    id: 98,
    conformance: "X",
    element: "serverCluster",
    xref: { document: "device", section: "8.1.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "DoorLock",
      id: 257,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "8.1.4" }
    },
    (0, import_elements.RequirementElement)({ name: "USER", conformance: "Matter & (PIN | RID | FPG | FACE | ALIRO)", element: "feature" }),
    (0, import_elements.RequirementElement)({ name: "RFIDCREDENTIAL", conformance: "P, O", element: "feature" }),
    (0, import_elements.RequirementElement)({ name: "AlarmMask", conformance: "[Alarms]", element: "attribute" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(DoorLockDt);
