// @matter/model/dist/cjs/standard/elements/elapsed-s.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var elapsed_s_exports = {};
__export(elapsed_s_exports, {
  elapsedS: () => elapsedS
});
module.exports = __toCommonJS(elapsed_s_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const elapsedS = (0, import_elements.DatatypeElement)({
  name: "elapsed-s",
  type: "uint32",
  description: "Elapsed Time in seconds",
  isSeed: true,
  details: "Elapsed time in seconds is an unsigned 32-bit value representing the time that has elapsed for an operation or other activity, as determined by the definition of the attribute using this type.",
  xref: { document: "core", section: "7.19.2.10" }
});
import_MatterDefinition.MatterDefinition.children.push(elapsedS);
