// @matter/model/dist/cjs/standard/elements/ElectricalMeasurementNS.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ElectricalMeasurementNS_exports = {};
__export(ElectricalMeasurementNS_exports, {
  ElectricalMeasurementNs: () => ElectricalMeasurementNs
});
module.exports = __toCommonJS(ElectricalMeasurementNS_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ElectricalMeasurementNs = (0, import_elements.SemanticNamespaceElement)(
  {
    name: "ElectricalMeasurement",
    id: 10,
    details: "The tags contained in this namespace are restricted for use in the electrical measurement domain and shall NOT be used in any other domain or context.",
    xref: { document: "namespace", section: "12" }
  },
  (0, import_elements.SemanticTagElement)({ name: "DC", id: 0, description: "Indicates values measured for a DC load" }),
  (0, import_elements.SemanticTagElement)({
    name: "AC",
    id: 1,
    description: "Indicates values measured for a single-phase AC load, or values measured for the collective load on a polyphase AC power supply"
  }),
  (0, import_elements.SemanticTagElement)({
    name: "ACPhase1",
    id: 2,
    description: "Indicates values measured for an AC load on phase 1 of a polyphase power supply"
  }),
  (0, import_elements.SemanticTagElement)({
    name: "ACPhase2",
    id: 3,
    description: "Indicates values measured for an AC load on phase 2 of a polyphase power supply"
  }),
  (0, import_elements.SemanticTagElement)({
    name: "ACPhase3",
    id: 4,
    description: "Indicates values measured for an AC load on phase 3 of a polyphase power supply"
  })
);
import_MatterDefinition.MatterDefinition.children.push(ElectricalMeasurementNs);
