// @matter/model/dist/cjs/standard/elements/ElectricalSensorDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ElectricalSensorDT_exports = {};
__export(ElectricalSensorDT_exports, {
  ElectricalSensorDt: () => ElectricalSensorDt
});
module.exports = __toCommonJS(ElectricalSensorDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ElectricalSensorDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "ElectricalSensor",
    id: 1296,
    category: "Utility",
    classification: "utility",
    details: "An Electrical Sensor device measures the electrical power and/or energy being imported and/or exported.",
    xref: { document: "device", section: "2.6" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 1296, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "PowerTopology",
    id: 156,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ElectricalPowerMeasurement",
    id: 144,
    conformance: "O.a+",
    element: "serverCluster",
    xref: { document: "device", section: "2.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ElectricalEnergyMeasurement",
    id: 145,
    conformance: "O.a+",
    element: "serverCluster",
    xref: { document: "device", section: "2.6.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(ElectricalSensorDt);
