// @matter/model/dist/cjs/standard/elements/endpoint-no.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var endpoint_no_exports = {};
__export(endpoint_no_exports, {
  endpointNo: () => endpointNo
});
module.exports = __toCommonJS(endpoint_no_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const endpointNo = (0, import_elements.DatatypeElement)({
  name: "endpoint-no",
  type: "uint16",
  description: "Endpoint Number",
  isSeed: true,
  details: "An unsigned number that indicates an instance of a device type. Endpoint numbers shall NOT be 0xFFFF, to allow all endpoint number values to be expressible in nullable endpoint-no fields.",
  xref: { document: "core", section: "7.19.2.23" }
});
import_MatterDefinition.MatterDefinition.children.push(endpointNo);
