// @matter/model/dist/cjs/standard/elements/EnergyEvseDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EnergyEvseDT_exports = {};
__export(EnergyEvseDT_exports, {
  EnergyEvseDt: () => EnergyEvseDt
});
module.exports = __toCommonJS(EnergyEvseDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const EnergyEvseDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "EnergyEvse",
    id: 1292,
    category: "Energy",
    classification: "simple",
    details: "An EVSE (Electric Vehicle Supply Equipment) is a device that allows an EV (Electric Vehicle) to be connected to the mains electricity supply to allow it to be charged (or discharged in case of Vehicle to Grid / Vehicle to Home applications).",
    xref: { document: "device", section: "14.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 1292, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "EnergyEvse",
    id: 153,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "EnergyEvseMode",
    id: 157,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(EnergyEvseDt);
