// @matter/model/dist/cjs/standard/elements/epoch-s.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var epoch_s_exports = {};
__export(epoch_s_exports, {
  epochS: () => epochS
});
module.exports = __toCommonJS(epoch_s_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const epochS = (0, import_elements.DatatypeElement)({
  name: "epoch-s",
  type: "uint32",
  description: "Epoch Time in seconds",
  isSeed: true,
  details: "This type has the same semantics as Epoch Time in Microseconds, except that:\n  \u2022 the value encodes an offset in seconds, rather than microseconds;\n  \u2022 the value is encoded as an unsigned 32-bit scalar, rather than 64-bit.\nThis type is employed where compactness of representation is important and where the resolution of seconds is still satisfactory.",
  xref: { document: "core", section: "7.19.2.6" }
});
import_MatterDefinition.MatterDefinition.children.push(epochS);
