// @matter/model/dist/cjs/standard/elements/ExtractorHoodDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ExtractorHoodDT_exports = {};
__export(ExtractorHoodDT_exports, {
  ExtractorHoodDt: () => ExtractorHoodDt
});
module.exports = __toCommonJS(ExtractorHoodDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ExtractorHoodDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "ExtractorHood",
    id: 122,
    category: "Appliances",
    classification: "simple",
    details: "An Extractor Hood is a device that is generally installed above a cooking surface in residential kitchens. An Extractor Hood\u2019s primary purpose is to reduce odors that arise during the cooking process by either extracting the air above the cooking surface or by recirculating and filtering it. It may also contain a light for illuminating the cooking surface.\nExtractor Hoods may also be known by the following names:\n  \u2022 Hoods\n  \u2022 Extractor Fans\n  \u2022 Extractors\n  \u2022 Range Hoods\n  \u2022 Telescoping Hoods\n  \u2022 Telescoping Extractors",
    xref: { document: "device", section: "13.10" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 122, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.10.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "HepaFilterMonitoring",
    id: 113,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.10.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ActivatedCarbonFilterMonitoring",
    id: 114,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.10.5" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "FanControl",
      id: 514,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.10.5" }
    },
    (0, import_elements.RequirementElement)({ name: "ROCKING", conformance: "X", element: "feature" }),
    (0, import_elements.RequirementElement)({ name: "WIND", conformance: "X", element: "feature" }),
    (0, import_elements.RequirementElement)({ name: "AIRFLOWDIRECTION", conformance: "X", element: "feature" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ExtractorHoodDt);
