// @matter/model/dist/cjs/standard/elements/fabric-idx.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fabric_idx_exports = {};
__export(fabric_idx_exports, {
  fabricIdx: () => fabricIdx
});
module.exports = __toCommonJS(fabric_idx_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const fabricIdx = (0, import_elements.DatatypeElement)({
  name: "fabric-idx",
  type: "uint8",
  description: "Fabric Index",
  isSeed: true,
  details: "This is an index that maps to a particular fabric on the node, see Fabric-Index. It is used for:\n  \u2022 the accessing fabric index of an interaction\n  \u2022 the FabricIndex global field in fabric-scoped data",
  xref: { document: "core", section: "7.19.2.20" }
});
import_MatterDefinition.MatterDefinition.children.push(fabricIdx);
