// @matter/model/dist/cjs/standard/elements/FabricIndex.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FabricIndex_exports = {};
__export(FabricIndex_exports, {
  FabricIndex: () => FabricIndex
});
module.exports = __toCommonJS(FabricIndex_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const FabricIndex = (0, import_elements.FieldElement)({
  name: "FabricIndex",
  id: 254,
  type: "fabric-idx",
  access: "R F V",
  conformance: "M",
  constraint: "1 to 254",
  isSeed: true,
  details: "This field shall be present for fabric-scoped data. This field does not have to be defined explicitly in the field table for fabric-scoped data.\nThis field shall NOT be present in a write interaction. For a write interaction, the server shall provide the value of the accessing fabric-index as the FabricIndex field value to processing logic, after receipt of the interaction. For a read interaction this field shall be included in all reported data that is defined as fabric-scoped.",
  xref: { document: "core", section: "7.13.6" }
});
import_MatterDefinition.MatterDefinition.children.push(FabricIndex);
