// @matter/model/dist/cjs/standard/elements/FanDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FanDT_exports = {};
__export(FanDT_exports, {
  FanDt: () => FanDt
});
module.exports = __toCommonJS(FanDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const FanDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "Fan",
    id: 43,
    category: "HVAC",
    classification: "simple",
    details: "A Fan device is typically standalone or mounted on a ceiling or wall and is used to circulate air in a room.",
    xref: { document: "device", section: "9.2" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 43, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.2.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.2.5" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OnOff",
    id: 6,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.2.5" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "FanControl",
      id: 514,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.2.5" }
    },
    (0, import_elements.RequirementElement)({ name: "FanModeSequence", access: "R V", conformance: "Matter", element: "attribute" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(FanDt);
