// @matter/model/dist/cjs/standard/elements/FixedLabel.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FixedLabel_exports = {};
__export(FixedLabel_exports, {
  FixedLabel: () => FixedLabel
});
module.exports = __toCommonJS(FixedLabel_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const FixedLabel = (0, import_elements.ClusterElement)(
  {
    name: "FixedLabel",
    id: 64,
    type: "Label",
    classification: "endpoint",
    pics: "FLABEL",
    details: 'This cluster is derived from the Label cluster and provides a feature for the device to tag an endpoint with zero or more read-only labels.\nExamples:\n  \u2022 A bridge can use this to indicate grouping of bridged devices. For example: All bridged devices     whose endpoints have an entry in their LabelList "room":"bedroom 2" are in the same (bed)room.\n  \u2022 A manufacturer can use this to identify a characteristic of an endpoint. For example to identify     the endpoints of a luminaire, one pointing up, the other pointing down, one of the endpoints     would have a LabelList entry "orientation":"up" while the other would have "orientation":"down".     Using such indication, the user interface of a Node controlling this luminaire\nknows which of the endpoints is which of the lights.\nNote that the TagList in the Descriptor cluster provides an alternative mechanism for such self- description using standardized tags rather than manufacturer-selected strings, yielding a standardized mechanism for features defined in the various namespaces. The second example above can be implemented using semantic tags Direction.Upward and Direction.Downward instead of (or in addition to) the Fixed Label cluster.',
    xref: { document: "core", section: "9.8" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    {
      name: "LabelList",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      default: [],
      quality: "N",
      xref: { document: "core", section: "9.8.4" }
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "LabelStruct" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(FixedLabel);
