// @matter/model/dist/cjs/standard/elements/FlowMeasurement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FlowMeasurement_exports = {};
__export(FlowMeasurement_exports, {
  FlowMeasurement: () => FlowMeasurement
});
module.exports = __toCommonJS(FlowMeasurement_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const FlowMeasurement = (0, import_elements.ClusterElement)(
  {
    name: "FlowMeasurement",
    id: 1028,
    classification: "application",
    pics: "FLW",
    details: "This cluster provides an interface to flow measurement functionality, including configuration and provision of notifications of flow measurements.",
    xref: { document: "cluster", section: "2.5" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    {
      name: "MeasuredValue",
      id: 0,
      type: "uint16",
      access: "R V",
      conformance: "M",
      constraint: "minMeasuredValue to maxMeasuredValue",
      default: null,
      quality: "X P",
      details: "Indicates the flow in m/h as follows:\nMeasuredValue = 10 x Flow\nThe null value indicates that the flow measurement is unknown, otherwise the range shall be as described in Measured Value.",
      xref: { document: "cluster", section: "2.5.4.1" }
    }
  ),
  (0, import_elements.AttributeElement)({
    name: "MinMeasuredValue",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "max 65533",
    quality: "X",
    details: "Indicates the minimum value of MeasuredValue that can be measured. See Measured Value for more details.\nThe null value indicates that the value is not available.",
    xref: { document: "cluster", section: "2.5.4.2" }
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxMeasuredValue",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min minMeasuredValue + 1",
    quality: "X",
    details: "Indicates the maximum value of MeasuredValue that can be measured. See\nMeasured Value for more details.\nThe null value indicates that the value is not available.",
    xref: { document: "cluster", section: "2.5.4.3" }
  }),
  (0, import_elements.AttributeElement)({
    name: "Tolerance",
    id: 3,
    type: "uint16",
    access: "R V",
    conformance: "O",
    constraint: "max 2048",
    default: 0,
    details: "See Measured Value.",
    xref: { document: "cluster", section: "2.5.4.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(FlowMeasurement);
