// @matter/model/dist/cjs/standard/elements/GeneratedCommandList.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GeneratedCommandList_exports = {};
__export(GeneratedCommandList_exports, {
  GeneratedCommandList: () => GeneratedCommandList
});
module.exports = __toCommonJS(GeneratedCommandList_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const GeneratedCommandList = (0, import_elements.AttributeElement)(
  {
    name: "GeneratedCommandList",
    id: 65528,
    type: "list",
    access: "R V",
    conformance: "M",
    isSeed: true,
    quality: "F",
    details: "This attribute is a list of server generated commands. A server generated command is a server to client command.\nEach instance of a cluster shall support this attribute.\nThis attribute shall be a list of the command IDs for server generated commands.\nFor each command in this list that is a response to a client command request, the request command shall be indicated in the AcceptedCommandList attribute.\nIf any attribute on a server supports atomic writes, this attribute shall contain the command ID for AtomicResponse.",
    xref: { document: "core", section: "7.13.5" }
  },
  (0, import_elements.FieldElement)({ name: "entry", type: "command-id" })
);
import_MatterDefinition.MatterDefinition.children.push(GeneratedCommandList);
