// @matter/model/dist/cjs/standard/elements/HeatPumpDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var HeatPumpDT_exports = {};
__export(HeatPumpDT_exports, {
  HeatPumpDt: () => HeatPumpDt
});
module.exports = __toCommonJS(HeatPumpDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const HeatPumpDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "HeatPump",
    id: 777,
    category: "Energy",
    classification: "simple",
    details: "A Heat Pump device is a device that uses electrical energy to heat either spaces or water tanks using ground, water or air as the heat source. These typically can heat the air or can pump water via central heating radiators or underfloor heating systems. It is typical to also heat hot water and store the heat in a hot water tank.\nNote that the Water Heater device type can also be heated by a heat pump and has similar requirements, but that cannot be used for space heating.",
    xref: { document: "device", section: "14.5" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 777, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.5.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Thermostat",
    id: 513,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "14.5.6" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(HeatPumpDt);
