// @matter/model/dist/cjs/standard/elements/IlluminanceMeasurement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var IlluminanceMeasurement_exports = {};
__export(IlluminanceMeasurement_exports, {
  IlluminanceMeasurement: () => IlluminanceMeasurement
});
module.exports = __toCommonJS(IlluminanceMeasurement_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const IlluminanceMeasurement = (0, import_elements.ClusterElement)(
  {
    name: "IlluminanceMeasurement",
    id: 1024,
    classification: "application",
    pics: "ILL",
    details: "The Illuminance Measurement cluster provides an interface to illuminance measurement functionality, including configuration and provision of notifications of illuminance measurements.",
    xref: { document: "cluster", section: "2.2" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    {
      name: "MeasuredValue",
      id: 0,
      type: "uint16",
      access: "R V",
      conformance: "M",
      constraint: "0, minMeasuredValue to maxMeasuredValue",
      default: 0,
      quality: "X P",
      details: "Indicates the illuminance in Lux (symbol lx) as follows:\n  \u2022 MeasuredValue = 10,000 x log10(illuminance) + 1,\nwhere 1 lx <= illuminance <= 3.576 Mlx, corresponding to a MeasuredValue in the range 1 to 0xFFFE. The MeasuredValue attribute can take the following values:\n  \u2022 0 indicates a value of illuminance that is too low to be measured,\n  \u2022 MinMeasuredValue <= MeasuredValue <= MaxMeasuredValue under normal circumstances,\n  \u2022 null indicates that the illuminance measurement is invalid.\nThe MeasuredValue attribute is updated continuously as new measurements are made.",
      xref: { document: "cluster", section: "2.2.5.1" }
    }
  ),
  (0, import_elements.AttributeElement)({
    name: "MinMeasuredValue",
    id: 1,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "1 to 65533",
    quality: "X",
    details: "Indicates the minimum value of MeasuredValue that can be measured. A value of null indicates that this attribute is not defined. See Measured Value for more details.",
    xref: { document: "cluster", section: "2.2.5.2" }
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxMeasuredValue",
    id: 2,
    type: "uint16",
    access: "R V",
    conformance: "M",
    constraint: "min minMeasuredValue + 1",
    quality: "X",
    details: "Indicates the maximum value of MeasuredValue that can be measured. A value of null indicates that this attribute is not defined. See Measured Value for more details.",
    xref: { document: "cluster", section: "2.2.5.3" }
  }),
  (0, import_elements.AttributeElement)({
    name: "Tolerance",
    id: 3,
    type: "uint16",
    access: "R V",
    conformance: "O",
    constraint: "max 2048",
    details: "See Measured Value.",
    xref: { document: "cluster", section: "2.2.5.4" }
  }),
  (0, import_elements.AttributeElement)({
    name: "LightSensorType",
    id: 4,
    type: "uint8",
    access: "R V",
    conformance: "O",
    default: null,
    quality: "X",
    details: "Indicates the electronic type of the light sensor. This attribute shall be set to one of the non-reserved values listed in LightSensorTypeEnum or null in case the sensor type is unknown.",
    xref: { document: "cluster", section: "2.2.5.5" }
  }),
  (0, import_elements.DatatypeElement)(
    { name: "LightSensorTypeEnum", type: "enum8", xref: { document: "cluster", section: "2.2.4.1" } },
    (0, import_elements.FieldElement)({ name: "Photodiode", id: 0, conformance: "M", description: "Indicates photodiode sensor type" }),
    (0, import_elements.FieldElement)({ name: "Cmos", id: 1, conformance: "M", description: "Indicates CMOS sensor type" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(IlluminanceMeasurement);
