// @matter/model/dist/cjs/standard/elements/ipv4adr.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ipv4adr_exports = {};
__export(ipv4adr_exports, {
  ipv4Adr: () => ipv4Adr
});
module.exports = __toCommonJS(ipv4adr_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ipv4Adr = (0, import_elements.DatatypeElement)({
  name: "ipv4adr",
  type: "octstr",
  description: "IPv4 Address",
  isSeed: true,
  details: "The IPv4 address data type is derived from an octet string. The octets shall correspond to the four octets in network byte order that comprise an IPv4 address represented utilizing quad-dotted notation.\nExamples of encoding:\n  \u2022 Address 192.168.2.235 \u2192 C0A802EB\n  \u2022 Address 10.4.200.75 \u2192 0A04C84B",
  xref: { document: "core", section: "7.19.2.38" }
});
import_MatterDefinition.MatterDefinition.children.push(ipv4Adr);
