// @matter/model/dist/cjs/standard/elements/JointFabricPki.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var JointFabricPki_exports = {};
__export(JointFabricPki_exports, {
  JointFabricPki: () => JointFabricPki
});
module.exports = __toCommonJS(JointFabricPki_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const JointFabricPki = (0, import_elements.ClusterElement)(
  {
    name: "JointFabricPki",
    id: 1875,
    classification: "node",
    pics: "JFPKI",
    details: "An instance of the Joint Fabric PKI Cluster only applies to Joint Fabric Administrator nodes fulfilling the role of Anchor CA.\nNOTE Support for Joint Fabric PKI Cluster is provisional.",
    xref: { document: "core", section: "11.25" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.CommandElement)(
    {
      name: "IcacsrRequest",
      id: 0,
      access: "A",
      conformance: "M",
      direction: "request",
      response: "IcacsrResponse",
      details: "This command shall be generated and executed during the Joint Commissioning Method steps and subsequently respond in the form of an ICACSRResponse command.\nCheck ICA Cross Signing for details about the generation and contents of the ICACSR.",
      xref: { document: "core", section: "11.25.5.1" }
    },
    (0, import_elements.FieldElement)({ name: "Icacsr", id: 0, type: "octstr", conformance: "M", constraint: "max 400" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "IcacsrResponse",
      id: 1,
      access: "A",
      conformance: "M",
      direction: "response",
      details: "This command shall be generated in response to the ICACSRRequest command. Check ICA Cross Signing for details about the generation and contents of ICAC.",
      xref: { document: "core", section: "11.25.5.2" }
    },
    (0, import_elements.FieldElement)({
      name: "StatusCode",
      id: 0,
      type: "IcacsrRequestStatusEnum",
      conformance: "M",
      details: "This field shall contain an ICACSRRequestStatusEnum value representing the status of the Section 11.25.5.1, \u201CICACSRRequest Command\u201D operation.",
      xref: { document: "core", section: "11.25.5.2.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Icac",
      id: 1,
      type: "octstr",
      conformance: "O",
      constraint: "max 400",
      details: "If present, it shall contain the NOC Issuer Certificate in PEM format.",
      xref: { document: "core", section: "11.25.5.2.2" }
    })
  ),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorRequest",
    id: 2,
    access: "A",
    conformance: "M",
    direction: "request",
    response: "TransferAnchorResponse",
    xref: { document: "core", section: "11.25.5" }
  }),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorResponse",
    id: 3,
    access: "A",
    conformance: "M",
    direction: "response",
    xref: { document: "core", section: "11.25.5" }
  }),
  (0, import_elements.CommandElement)({
    name: "TransferAnchorComplete",
    id: 4,
    access: "A",
    conformance: "M",
    direction: "request",
    response: "status",
    xref: { document: "core", section: "11.25.5" }
  }),
  (0, import_elements.DatatypeElement)(
    {
      name: "IcacsrRequestStatusEnum",
      type: "enum8",
      details: "This enumeration is used by the ICACSRResponse command to convey the detailed outcome of this cluster\u2019s ICACSRRequest command.",
      xref: { document: "core", section: "11.25.4.1" }
    },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M", description: "No error" }),
    (0, import_elements.FieldElement)({
      name: "InvalidIcaCsrFormat",
      id: 1,
      conformance: "M",
      description: "The ICACSR in the request is not compliant to PKCS #10 rules"
    }),
    (0, import_elements.FieldElement)({
      name: "InvalidIcaCsrSignature",
      id: 2,
      conformance: "M",
      description: "The ICACSR in the request has an incorrect signature"
    }),
    (0, import_elements.FieldElement)({
      name: "FailedDclVendorIdValidation",
      id: 3,
      conformance: "M",
      description: "DCL Vendor ID validation failed"
    }),
    (0, import_elements.FieldElement)({ name: "NotAnIcac", id: 4, conformance: "M", description: "DCL returned certificate is not an ICAC" }),
    (0, import_elements.FieldElement)({
      name: "BusyAnchorTransfer",
      id: 5,
      conformance: "M",
      description: "Error due to an in progress Anchor Transfer"
    }),
    (0, import_elements.FieldElement)({ name: "IcaCsrSigningFailed", id: 6, conformance: "M", description: "Signing the ICA CSR failed" }),
    (0, import_elements.FieldElement)({ name: "IcaCsrRequestNoUserConsent", id: 7, conformance: "M", description: "No user consent" })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "TransferAnchorResponseStatusEnum",
      type: "enum8",
      details: "This enumeration is used by the TransferAnchorResponse command to convey the detailed outcome of this cluster\u2019s TransferAnchorRequest command.",
      xref: { document: "core", section: "11.25.4.2" }
    },
    (0, import_elements.FieldElement)({ name: "Ok", id: 0, conformance: "M", description: "No error" }),
    (0, import_elements.FieldElement)({
      name: "TransferAnchorStatusDatastoreBusy",
      id: 1,
      conformance: "M",
      description: "Anchor Transfer was not started due to on- going Datastore operations"
    }),
    (0, import_elements.FieldElement)({
      name: "TransferAnchorStatusNoUserConsent",
      id: 2,
      conformance: "M",
      description: "User has not consented for Anchor Transfer"
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(JointFabricPki);
