// @matter/model/dist/cjs/standard/elements/LaundryDryerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LaundryDryerDT_exports = {};
__export(LaundryDryerDT_exports, {
  LaundryDryerDt: () => LaundryDryerDt
});
module.exports = __toCommonJS(LaundryDryerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LaundryDryerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "LaundryDryer",
    id: 124,
    category: "Appliances",
    classification: "simple",
    details: "A Laundry Dryer represents a device that is capable of drying laundry items.",
    xref: { document: "device", section: "13.6" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 124, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "LaundryWasherMode",
      id: 81,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.6.4" }
    },
    (0, import_elements.RequirementElement)({ name: "StartUpMode", conformance: "X", element: "attribute" }),
    (0, import_elements.RequirementElement)({ name: "ONOFF", conformance: "X", element: "feature" })
  ),
  (0, import_elements.RequirementElement)(
    {
      name: "OnOff",
      id: 6,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.6.4" }
    },
    (0, import_elements.RequirementElement)({ name: "DEADFRONTBEHAVIOR", conformance: "M", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "LaundryDryerControls",
    id: 74,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureControl",
    id: 86,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OperationalState",
    id: 96,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(LaundryDryerDt);
