// @matter/model/dist/cjs/standard/elements/LaundryWasherDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LaundryWasherDT_exports = {};
__export(LaundryWasherDT_exports, {
  LaundryWasherDt: () => LaundryWasherDt
});
module.exports = __toCommonJS(LaundryWasherDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LaundryWasherDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "LaundryWasher",
    id: 115,
    category: "Appliances",
    classification: "simple",
    details: "A Laundry Washer represents a device that is capable of laundering consumer items. Any laundry washer product may utilize this device type.\nA Laundry Washer shall be composed of at least one endpoint with the Laundry Washer device type.",
    xref: { document: "device", section: "13.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 115, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.1.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "LaundryWasherMode",
      id: 81,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.1.4" }
    },
    (0, import_elements.RequirementElement)({ name: "StartUpMode", conformance: "X", element: "attribute" }),
    (0, import_elements.RequirementElement)({ name: "ONOFF", conformance: "X", element: "feature" })
  ),
  (0, import_elements.RequirementElement)(
    {
      name: "OnOff",
      id: 6,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.1.4" }
    },
    (0, import_elements.RequirementElement)({ name: "DEADFRONTBEHAVIOR", conformance: "M", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "LaundryWasherControls",
    id: 83,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureControl",
    id: 86,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OperationalState",
    id: 96,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.1.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(LaundryWasherDt);
