// @matter/model/dist/cjs/standard/elements/LaundryDryerControls.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LaundryDryerControls_exports = {};
__export(LaundryDryerControls_exports, {
  LaundryDryerControls: () => LaundryDryerControls
});
module.exports = __toCommonJS(LaundryDryerControls_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LaundryDryerControls = (0, import_elements.ClusterElement)(
  {
    name: "LaundryDryerControls",
    id: 74,
    classification: "application",
    pics: "DRYERCTRL",
    details: "This cluster provides a way to access options associated with the operation of a laundry dryer device type.",
    xref: { document: "cluster", section: "8.9" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    {
      name: "SupportedDrynessLevels",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "1 to 4",
      details: "Indicates the list of supported dryness levels available to the appliance in the currently selected mode. The dryness level values are determined by the manufacturer. At least one dryness level value shall be provided in the SupportedDrynessLevels list. The list of dryness levels may change depending on the currently-selected Laundry Dryer mode.",
      xref: { document: "cluster", section: "8.9.5.1" }
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "DrynessLevelEnum" })
  ),
  (0, import_elements.AttributeElement)({
    name: "SelectedDrynessLevel",
    id: 1,
    type: "DrynessLevelEnum",
    access: "RW VO",
    conformance: "M",
    constraint: "desc",
    quality: "X",
    details: "Indicates the currently-selected dryness level and it shall be the index into the SupportedDrynessLevels list of the selected dryness level.\nIf an attempt is made to write this attribute with a value other than null or a value contained in SupportedDrynessLevels, a CONSTRAINT_ERROR response shall be sent as the response. If an attempt is made to write this attribute while the device is not in a state that supports modifying the dryness level, an INVALID_IN_STATE error shall be sent as the response. A value of null shall indicate that there will be no dryness level setting for the current mode.",
    xref: { document: "cluster", section: "8.9.5.2" }
  }),
  (0, import_elements.DatatypeElement)(
    {
      name: "DrynessLevelEnum",
      type: "enum8",
      details: "This enum provides a representation of the level of dryness that will be used while drying in a selected mode.\nIt is up to the device manufacturer to determine the mapping between the enum values and the corresponding temperature level.",
      xref: { document: "cluster", section: "8.9.4.1" }
    },
    (0, import_elements.FieldElement)({ name: "Low", id: 0, conformance: "M", description: "Provides a low dryness level for the selected mode" }),
    (0, import_elements.FieldElement)({
      name: "Normal",
      id: 1,
      conformance: "M",
      description: "Provides the normal level of dryness for the selected mode"
    }),
    (0, import_elements.FieldElement)({
      name: "Extra",
      id: 2,
      conformance: "M",
      description: "Provides an extra dryness level for the selected mode"
    }),
    (0, import_elements.FieldElement)({ name: "Max", id: 3, conformance: "M", description: "Provides the max dryness level for the selected mode" })
  )
);
import_MatterDefinition.MatterDefinition.children.push(LaundryDryerControls);
