// @matter/model/dist/cjs/standard/elements/LaundryWasherMode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LaundryWasherMode_exports = {};
__export(LaundryWasherMode_exports, {
  LaundryWasherMode: () => LaundryWasherMode
});
module.exports = __toCommonJS(LaundryWasherMode_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LaundryWasherMode = (0, import_elements.ClusterElement)(
  {
    name: "LaundryWasherMode",
    id: 81,
    type: "ModeBase",
    classification: "application",
    pics: "LWM",
    details: "This cluster is derived from the Mode Base cluster and defines additional mode tags and namespaced enumerated values for laundry washer as well as laundry dryer devices.",
    xref: { document: "cluster", section: "8.5" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap", xref: { document: "cluster", section: "8.5.4" } },
    (0, import_elements.FieldElement)({
      name: "DEPONOFF",
      conformance: "X",
      constraint: "0",
      description: "OnOff",
      details: "Dependency with the OnOff cluster"
    })
  ),
  (0, import_elements.AttributeElement)({
    name: "SupportedModes",
    id: 0,
    conformance: "M",
    details: "At least one entry in the SupportedModes attribute shall include the Normal mode tag in the ModeTags field list.",
    xref: { document: "cluster", section: "8.5.6.1" }
  }),
  (0, import_elements.AttributeElement)({ name: "CurrentMode", id: 1, conformance: "M", xref: { document: "cluster", section: "8.5.6" } }),
  (0, import_elements.AttributeElement)({ name: "StartUpMode", id: 2, conformance: "X", xref: { document: "cluster", section: "8.5.6" } }),
  (0, import_elements.AttributeElement)({ name: "OnMode", id: 3, conformance: "X", xref: { document: "cluster", section: "8.5.6" } }),
  (0, import_elements.DatatypeElement)({
    name: "ModeOptionStruct",
    type: "ModeOptionStruct",
    details: "The table below lists the changes relative to the Mode Base cluster for the fields of the ModeOptionStruct type. A blank field indicates no change.",
    xref: { document: "cluster", section: "8.5.5.1" }
  }),
  (0, import_elements.DatatypeElement)(
    { name: "ModeTag", type: "enum16" },
    (0, import_elements.FieldElement)({ name: "Auto", id: 0, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Quick", id: 1, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Quiet", id: 2, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "LowNoise", id: 3, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "LowEnergy", id: 4, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Vacation", id: 5, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Min", id: 6, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Max", id: 7, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Night", id: 8, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({ name: "Day", id: 9, xref: { document: "cluster", section: "8.5.7.1" } }),
    (0, import_elements.FieldElement)({
      name: "Normal",
      id: 16384,
      details: "The normal regime of operation.",
      xref: { document: "cluster", section: "8.5.7.1.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "Delicate",
      id: 16385,
      details: "Mode optimized for washing delicate garments.",
      xref: { document: "cluster", section: "8.5.7.1.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Heavy",
      id: 16386,
      details: "Mode optimized for heavy washing.",
      xref: { document: "cluster", section: "8.5.7.1.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "Whites",
      id: 16387,
      details: "Mode optimized for stain removal on white fabrics.",
      xref: { document: "cluster", section: "8.5.7.1.4" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(LaundryWasherMode);
