// @matter/model/dist/cjs/standard/elements/LocationNS.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LocationNS_exports = {};
__export(LocationNS_exports, {
  LocationNs: () => LocationNs
});
module.exports = __toCommonJS(LocationNS_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LocationNs = (0, import_elements.SemanticNamespaceElement)(
  {
    name: "Location",
    id: 6,
    details: "The tags contained in this namespace may be used in any domain or context, to indicate an association with a location of a device (e.g. an outdoor temperature sensor).",
    xref: { document: "namespace", section: "7" }
  },
  (0, import_elements.SemanticTagElement)({
    name: "Indoor",
    id: 0,
    description: 'Element is indoors or related to indoor equipment/conditions (e.g. the "indoor" temperature).'
  }),
  (0, import_elements.SemanticTagElement)({
    name: "Outdoor",
    id: 1,
    description: 'Element is outdoors or related to outdoor equipment/conditions (e.g. the "outdoor" temperature).'
  }),
  (0, import_elements.SemanticTagElement)({
    name: "Inside",
    id: 2,
    description: 'Element is located inside the equipment (e.g. a sensor "inside" a cabinet).'
  }),
  (0, import_elements.SemanticTagElement)({
    name: "Outside",
    id: 3,
    description: 'Element is located outside the equipment (e.g. a sensor "outside" a cabinet)'
  })
);
import_MatterDefinition.MatterDefinition.children.push(LocationNs);
