// @matter/model/dist/cjs/standard/elements/MicrowaveOvenDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MicrowaveOvenDT_exports = {};
__export(MicrowaveOvenDT_exports, {
  MicrowaveOvenDt: () => MicrowaveOvenDt
});
module.exports = __toCommonJS(MicrowaveOvenDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const MicrowaveOvenDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "MicrowaveOven",
    id: 121,
    category: "Appliances",
    classification: "simple",
    details: "This defines conformance to the Microwave Oven device type.\nA Microwave Oven is a device with the primary function of heating foods and beverages using a magnetron.",
    xref: { document: "device", section: "13.11" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 121, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.11.6" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "OperationalState",
      id: 96,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.11.6" }
    },
    (0, import_elements.RequirementElement)({ name: "CountdownTime", conformance: "M", element: "attribute" })
  ),
  (0, import_elements.RequirementElement)(
    {
      name: "FanControl",
      id: 514,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.11.6" }
    },
    (0, import_elements.RequirementElement)({ name: "WIND", conformance: "X", element: "feature" }),
    (0, import_elements.RequirementElement)({ name: "AIRFLOWDIRECTION", conformance: "X", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "MicrowaveOvenMode",
    id: 94,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.11.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "MicrowaveOvenControl",
    id: 95,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.11.6" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(MicrowaveOvenDt);
