// @matter/model/dist/cjs/standard/elements/ModeSelectDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ModeSelectDT_exports = {};
__export(ModeSelectDT_exports, {
  ModeSelectDt: () => ModeSelectDt
});
module.exports = __toCommonJS(ModeSelectDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ModeSelectDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "ModeSelect",
    id: 39,
    category: "Generic",
    classification: "simple",
    details: "This defines conformance to the Mode Select device type.",
    xref: { document: "device", section: "11.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 39, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "ModeSelect",
    id: 80,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "11.1.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(ModeSelectDt);
