// @matter/model/dist/cjs/standard/elements/NetworkInfrastructureManagerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NetworkInfrastructureManagerDT_exports = {};
__export(NetworkInfrastructureManagerDT_exports, {
  NetworkInfrastructureManagerDt: () => NetworkInfrastructureManagerDt
});
module.exports = __toCommonJS(NetworkInfrastructureManagerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const NetworkInfrastructureManagerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "NetworkInfrastructureManager",
    id: 144,
    category: "Network Infrastructure",
    classification: "simple",
    details: "A Network Infrastructure Manager provides interfaces that allow for the management of the Wi-Fi, Thread, and Ethernet networks underlying a Matter deployment, realizing the Star Network Topology described in [MatterCore].\nExamples of physical devices that implement the Matter Network Infrastructure Manager device type include Wi-Fi gateway routers.\nRelevant hardware and software requirements for Network Infrastructure Manager devices are defined in Section 15.2.6, \u201COther Requirements\u201D and within the clusters mandated by this device type.\nA Network Infrastructure Manager device may be managed by a service associated with the device vendor, for example, an Internet Service Provider. Sometimes this managing service will have policies that require the use of the Managed Device feature of the Access Control Cluster (see Section 15.2.5.1, \u201CAccess Control MNGD Conformance\u201D). Consequently, Commissioners of this device type should be aware of this feature and its use.",
    xref: { document: "device", section: "15.2" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 144, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "WiFiNetworkManagement",
    id: 1105,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "15.2.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ThreadBorderRouterManagement",
    id: 1106,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "15.2.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ThreadNetworkDirectory",
    id: 1107,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "15.2.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(NetworkInfrastructureManagerDt);
