// @matter/model/dist/cjs/standard/elements/OccupancySensorDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OccupancySensorDT_exports = {};
__export(OccupancySensorDT_exports, {
  OccupancySensorDt: () => OccupancySensorDt
});
module.exports = __toCommonJS(OccupancySensorDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const OccupancySensorDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "OccupancySensor",
    id: 263,
    category: "Sensor",
    classification: "simple",
    details: "An Occupancy Sensor is a measurement and sensing device that is capable of measuring and reporting the occupancy state in a designated area.",
    xref: { document: "device", section: "7.3" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 263, revision: 4 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.3.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "BooleanStateConfiguration",
    id: 128,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.3.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OccupancySensing",
    id: 1030,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.3.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(OccupancySensorDt);
