// @matter/model/dist/cjs/standard/elements/octstr.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var octstr_exports = {};
__export(octstr_exports, {
  octstr: () => octstr
});
module.exports = __toCommonJS(octstr_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const octstr = (0, import_elements.DatatypeElement)({
  name: "octstr",
  description: "Octet string",
  isSeed: true,
  metatype: "bytes",
  details: "The octet string data type defines a sequence of octets with a finite octet count from 0 to 65534. It is recommended to define a constraint on the maximum possible count.",
  xref: { document: "core", section: "7.19.1.7" }
});
import_MatterDefinition.MatterDefinition.children.push(octstr);
