// @matter/model/dist/cjs/standard/elements/OtaProviderDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OtaProviderDT_exports = {};
__export(OtaProviderDT_exports, {
  OtaProviderDt: () => OtaProviderDt
});
module.exports = __toCommonJS(OtaProviderDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const OtaProviderDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "OtaProvider",
    id: 20,
    category: "Utility",
    classification: "utility",
    details: "An OTA Provider is a node that is capable of providing an OTA software update to other nodes on the same fabric.",
    xref: { document: "device", section: "2.4" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 20, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "OtaSoftwareUpdateRequestor",
    id: 42,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "2.4.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OtaSoftwareUpdateProvider",
    id: 41,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.4.3" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(OtaProviderDt);
