// @matter/model/dist/cjs/standard/elements/posix-ms.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var posix_ms_exports = {};
__export(posix_ms_exports, {
  posixMs: () => posixMs
});
module.exports = __toCommonJS(posix_ms_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const posixMs = (0, import_elements.DatatypeElement)({
  name: "posix-ms",
  type: "uint64",
  description: "POSIX Time in milliseconds",
  isSeed: true,
  details: "This type represents an offset, in milliseconds, from the UNIX epoch (1970-01-01 00:00:00 UTC), encoded as an unsigned 64-bit scalar value.\nThis type is employed for compatibility reasons.",
  xref: { document: "core", section: "7.19.2.7" }
});
import_MatterDefinition.MatterDefinition.children.push(posixMs);
