// @matter/model/dist/cjs/standard/elements/PowerSourceDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PowerSourceDT_exports = {};
__export(PowerSourceDT_exports, {
  PowerSourceDt: () => PowerSourceDt
});
module.exports = __toCommonJS(PowerSourceDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PowerSourceDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "PowerSource",
    id: 17,
    category: "Utility",
    classification: "utility",
    xref: { document: "device", section: "2.2" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 17, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "PowerSource",
    id: 47,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.2.3" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(PowerSourceDt);
