// @matter/model/dist/cjs/standard/elements/power-mW.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var power_mW_exports = {};
__export(power_mW_exports, {
  powerMW: () => powerMW
});
module.exports = __toCommonJS(power_mW_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const powerMW = (0, import_elements.DatatypeElement)({
  name: "power-mW",
  type: "int64",
  description: "Power",
  isSeed: true,
  details: "This type represents power measured in milliwatts.",
  xref: { document: "core", section: "7.19.2.12" }
});
import_MatterDefinition.MatterDefinition.children.push(powerMW);
