// @matter/model/dist/cjs/standard/elements/PressureMeasurement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PressureMeasurement_exports = {};
__export(PressureMeasurement_exports, {
  PressureMeasurement: () => PressureMeasurement
});
module.exports = __toCommonJS(PressureMeasurement_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PressureMeasurement = (0, import_elements.ClusterElement)(
  {
    name: "PressureMeasurement",
    id: 1027,
    classification: "application",
    pics: "PRS",
    details: "This cluster provides an interface to pressure measurement functionality, including configuration and provision of notifications of pressure measurements.",
    xref: { document: "cluster", section: "2.4" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 3 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap", xref: { document: "cluster", section: "2.4.4" } },
    (0, import_elements.FieldElement)({
      name: "EXT",
      conformance: "O",
      constraint: "0",
      description: "Extended",
      details: "Extended range and resolution"
    })
  ),
  (0, import_elements.AttributeElement)(
    {
      name: "MeasuredValue",
      id: 0,
      type: "int16",
      access: "R V",
      conformance: "M",
      constraint: "minMeasuredValue to maxMeasuredValue",
      quality: "X P",
      details: "Indicates the pressure in kPa as follows:\nMeasuredValue = 10 x Pressure [kPa]\nThe null value indicates that the value is not available.",
      xref: { document: "cluster", section: "2.4.5.1" }
    }
  ),
  (0, import_elements.AttributeElement)({
    name: "MinMeasuredValue",
    id: 1,
    type: "int16",
    access: "R V",
    conformance: "M",
    constraint: "max 32766",
    quality: "X",
    details: "Indicates the minimum value of MeasuredValue that can be measured. See Measured Value for more details.\nThe null value indicates that the value is not available.",
    xref: { document: "cluster", section: "2.4.5.2" }
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxMeasuredValue",
    id: 2,
    type: "int16",
    access: "R V",
    conformance: "M",
    constraint: "minMeasuredValue + 1 to 32767",
    quality: "X",
    details: "Indicates the maximum value of MeasuredValue that can be measured. See Measured Value for more details.\nThe null value indicates that the value is not available.",
    xref: { document: "cluster", section: "2.4.5.3" }
  }),
  (0, import_elements.AttributeElement)({
    name: "Tolerance",
    id: 3,
    type: "uint16",
    access: "R V",
    conformance: "O",
    constraint: "max 2048",
    default: 0,
    details: "See Measured Value.",
    xref: { document: "cluster", section: "2.4.5.4" }
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "ScaledValue",
      id: 16,
      type: "int16",
      access: "R V",
      conformance: "EXT",
      constraint: "minScaledValue to maxScaledValue",
      default: 0,
      quality: "X",
      details: "Indicates the pressure in Pascals as follows:\nScaledValue = 10Scale x Pressure [Pa]\nThe null value indicates that the value is not available.",
      xref: { document: "cluster", section: "2.4.5.5" }
    }
  ),
  (0, import_elements.AttributeElement)({
    name: "MinScaledValue",
    id: 17,
    type: "int16",
    access: "R V",
    conformance: "EXT",
    constraint: "max 32766",
    default: 0,
    quality: "X",
    details: "Indicates the minimum value of ScaledValue that can be measured. The null value indicates that the value is not available.",
    xref: { document: "cluster", section: "2.4.5.6" }
  }),
  (0, import_elements.AttributeElement)({
    name: "MaxScaledValue",
    id: 18,
    type: "int16",
    access: "R V",
    conformance: "EXT",
    constraint: "minScaledValue + 1 to 32767",
    default: 0,
    quality: "X",
    details: "Indicates the maximum value of ScaledValue that can be measured. The null value indicates that the value is not available.",
    xref: { document: "cluster", section: "2.4.5.7" }
  }),
  (0, import_elements.AttributeElement)({
    name: "ScaledTolerance",
    id: 19,
    type: "uint16",
    access: "R V",
    conformance: "[EXT]",
    constraint: "max 2048",
    default: 0,
    details: "Indicates the magnitude of the possible error that is associated with Scaled\nValue. The true value is located in the range\n(ScaledValue \u2013 ScaledTolerance) to (ScaledValue + ScaledTolerance).",
    xref: { document: "cluster", section: "2.4.5.8" }
  }),
  (0, import_elements.AttributeElement)({
    name: "Scale",
    id: 20,
    type: "int8",
    access: "R V",
    conformance: "EXT",
    constraint: "min -127",
    default: 0,
    details: "Indicates the base 10 exponent used to obtain ScaledValue (see ScaledValue).",
    xref: { document: "cluster", section: "2.4.5.9" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(PressureMeasurement);
