// @matter/model/dist/cjs/standard/elements/PumpControllerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PumpControllerDT_exports = {};
__export(PumpControllerDT_exports, {
  PumpControllerDt: () => PumpControllerDt
});
module.exports = __toCommonJS(PumpControllerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PumpControllerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "PumpController",
    id: 772,
    category: "Switches and Controls",
    classification: "simple",
    details: "A Pump Controller device is capable of configuring and controlling a Pump device.",
    xref: { document: "device", section: "6.5" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 772, revision: 4 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PumpConfigurationAndControl",
    id: 512,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "LevelControl",
    id: 8,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ScenesManagement",
    id: 98,
    conformance: "P, O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(PumpControllerDt);
