// @matter/model/dist/cjs/standard/elements/PumpDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PumpDT_exports = {};
__export(PumpDT_exports, {
  PumpDt: () => PumpDt
});
module.exports = __toCommonJS(PumpDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PumpDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "Pump",
    id: 771,
    category: "Smart Plugs/Outlets and other Actuators",
    classification: "simple",
    details: "A Pump device is a pump that may have variable speed. It may have optional built-in sensors and a regulation mechanism. It is typically used for pumping fluids like water.",
    xref: { document: "device", section: "5.5" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 771, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PumpConfigurationAndControl",
    id: 512,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "LevelControl",
    id: 8,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ScenesManagement",
    id: 98,
    conformance: "P, O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OccupancySensing",
    id: 1030,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(PumpDt);
