// @matter/model/dist/cjs/standard/elements/RefrigeratorAlarm.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RefrigeratorAlarm_exports = {};
__export(RefrigeratorAlarm_exports, {
  RefrigeratorAlarm: () => RefrigeratorAlarm
});
module.exports = __toCommonJS(RefrigeratorAlarm_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RefrigeratorAlarm = (0, import_elements.ClusterElement)(
  {
    name: "RefrigeratorAlarm",
    id: 87,
    type: "AlarmBase",
    classification: "application",
    pics: "REFALM",
    details: "This cluster is a derived cluster of Alarm Base cluster and provides the alarm definition related to refrigerator and temperature controlled cabinet devices.",
    xref: { document: "cluster", section: "8.8" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)(
    { name: "FeatureMap", id: 65532, type: "FeatureMap", xref: { document: "cluster", section: "8.8.4" } },
    (0, import_elements.FieldElement)({
      name: "RESET",
      conformance: "X",
      constraint: "0",
      description: "Reset",
      details: "Supports the ability to reset alarms"
    })
  ),
  (0, import_elements.CommandElement)({ name: "ModifyEnabledAlarms", id: 1, conformance: "X", xref: { document: "cluster", section: "8.8.7" } }),
  (0, import_elements.DatatypeElement)(
    { name: "AlarmBitmap", type: "map32", xref: { document: "cluster", section: "8.8.5.1" } },
    (0, import_elements.FieldElement)({
      name: "DoorOpen",
      constraint: "0",
      description: "The cabinet\u2019s door has been open for a vendor defined amount of time."
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(RefrigeratorAlarm);
