// @matter/model/dist/cjs/standard/elements/RefrigeratorDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RefrigeratorDT_exports = {};
__export(RefrigeratorDT_exports, {
  RefrigeratorDt: () => RefrigeratorDt
});
module.exports = __toCommonJS(RefrigeratorDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RefrigeratorDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "Refrigerator",
    id: 112,
    category: "Appliances",
    classification: "simple",
    details: "A refrigerator represents a device that contains one or more cabinets that are capable of chilling or freezing food. Examples of consumer products that may make use of this device type include refrigerators, freezers, and wine coolers.",
    xref: { document: "device", section: "13.2" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 112, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.2.6" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "RefrigeratorAndTemperatureControlledCabinetMode",
      id: 82,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.2.6" }
    },
    (0, import_elements.RequirementElement)({ name: "StartUpMode", conformance: "X", element: "attribute" }),
    (0, import_elements.RequirementElement)({ name: "ONOFF", conformance: "X", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "RefrigeratorAlarm",
    id: 87,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.2.6" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(RefrigeratorDt);
