// @matter/model/dist/cjs/standard/elements/RoboticVacuumCleanerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RoboticVacuumCleanerDT_exports = {};
__export(RoboticVacuumCleanerDT_exports, {
  RoboticVacuumCleanerDt: () => RoboticVacuumCleanerDt
});
module.exports = __toCommonJS(RoboticVacuumCleanerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RoboticVacuumCleanerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "RoboticVacuumCleaner",
    id: 116,
    category: "Robotic",
    classification: "simple",
    details: "This defines conformance for the Robotic Vacuum Cleaner device type.",
    xref: { document: "device", section: "12.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 116, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RvcRunMode",
    id: 84,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RvcCleanMode",
    id: 85,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RvcOperationalState",
    id: 97,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ServiceArea",
    id: 336,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(RoboticVacuumCleanerDt);
