// @matter/model/dist/cjs/standard/elements/RoomAirConditionerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RoomAirConditionerDT_exports = {};
__export(RoomAirConditionerDT_exports, {
  RoomAirConditionerDt: () => RoomAirConditionerDt
});
module.exports = __toCommonJS(RoomAirConditionerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const RoomAirConditionerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "RoomAirConditioner",
    id: 114,
    category: "Appliances",
    classification: "simple",
    details: "This defines conformance to the Room Air Conditioner device type.\nA Room Air Conditioner is a device with the primary function of controlling the air temperature in a single room.",
    xref: { document: "device", section: "13.3" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 114, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ScenesManagement",
    id: 98,
    conformance: "P, O",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "OnOff",
      id: 6,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    },
    (0, import_elements.RequirementElement)({ name: "DEADFRONTBEHAVIOR", conformance: "M", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Thermostat",
    id: 513,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FanControl",
    id: 514,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "ThermostatUserInterfaceConfiguration",
      id: 516,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.3.6" }
    },
    (0, import_elements.RequirementElement)({ name: "KeypadLockout", conformance: "O", element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RelativeHumidityMeasurement",
    id: 1029,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.3.6" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(RoomAirConditionerDt);
