// @matter/model/dist/cjs/standard/elements/SecondaryNetworkInterfaceDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SecondaryNetworkInterfaceDT_exports = {};
__export(SecondaryNetworkInterfaceDT_exports, {
  SecondaryNetworkInterfaceDt: () => SecondaryNetworkInterfaceDt
});
module.exports = __toCommonJS(SecondaryNetworkInterfaceDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SecondaryNetworkInterfaceDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "SecondaryNetworkInterface",
    id: 25,
    category: "Utility",
    classification: "utility",
    details: "A Secondary Network Interface device provides an additional network interface supported by the Node, supplementing the primary interface hosted by the Root Node endpoint.\nA Node supporting multiple network interfaces shall include the primary interface on the Root Node endpoint, along with secondary interfaces on other endpoints. The priorities of these network interfaces are determined by the order of their endpoints, where interfaces with smaller endpoint numbers are higher priority.",
    xref: { document: "device", section: "2.8" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 25, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "NetworkCommissioning",
    id: 49,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.8.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "EthernetNetworkDiagnostics",
    id: 55,
    conformance: "[Ethernet]",
    element: "serverCluster",
    xref: { document: "device", section: "2.8.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "WiFiNetworkDiagnostics",
    id: 54,
    conformance: "[Wi, Fi]",
    element: "serverCluster",
    xref: { document: "device", section: "2.8.3" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ThreadNetworkDiagnostics",
    id: 53,
    conformance: "[Thread]",
    element: "serverCluster",
    xref: { document: "device", section: "2.8.3" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(SecondaryNetworkInterfaceDt);
