// @matter/model/dist/cjs/standard/elements/SpeakerDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SpeakerDT_exports = {};
__export(SpeakerDT_exports, {
  SpeakerDt: () => SpeakerDt
});
module.exports = __toCommonJS(SpeakerDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SpeakerDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "Speaker",
    id: 34,
    category: "Media",
    classification: "simple",
    details: "This defines conformance to the Speaker device type. This feature controls the speaker volume of the device.\nTo control unmute/mute, the On/Off cluster shall be used. A value of TRUE for the OnOff attribute shall represent the volume on (not muted) state, while a value of FALSE shall represent the volume off (muted) state. For volume level control, the Level cluster shall be used.\nA dedicated endpoint is needed because the On/Off cluster can also be used for other purposes, such as for power control.\nThe decision to use Level and On/Off clusters for volume (rather than defining a new audio control cluster) was made in order to treat volume in a fashion consistent with lighting which also uses these clusters and has matching functional requirements.",
    xref: { document: "device", section: "10.4" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 34, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "10.4.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "LevelControl",
    id: 8,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "10.4.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(SpeakerDt);
