// @matter/model/dist/cjs/standard/elements/status.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var status_exports = {};
__export(status_exports, {
  status: () => status
});
module.exports = __toCommonJS(status_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const status = (0, import_elements.DatatypeElement)(
  {
    name: "status",
    type: "enum8",
    description: "Status Code",
    isSeed: true,
    metatype: "enum",
    details: "An enumeration value that means a success or error status. A status code is indicated as a response to an action in an interaction (see Interaction Model).\nA status code shall be one of:\n  \u2022 a common status code from the set defined in the Interaction Model status code table;\n  \u2022 a cluster status code that is scoped to a particular cluster.\nThe following table defines the enumeration ranges for status codes.\nStatus codes in an undefined range, or status codes undefined within a range are reserved and shall NOT be indicated.",
    xref: { document: "core", section: "7.19.2.18" }
  },
  (0, import_elements.FieldElement)({
    name: "Success",
    id: 0,
    description: "Operation was successful.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "Failure",
    id: 1,
    description: "Operation was not successful.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "InvalidSubscription",
    id: 125,
    description: "Subscription ID is not active.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedAccess",
    id: 126,
    description: "The sender of the action or command does not have authorization or access.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedEndpoint",
    id: 127,
    description: "The endpoint indicated is unsupported on the node.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "InvalidAction",
    id: 128,
    description: "The action is malformed, has missing fields, or fields with invalid values. Action not carried out.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedCommand",
    id: 129,
    description: "The indicated command ID is not supported on the cluster instance. Command not carried out.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "InvalidCommand",
    id: 133,
    description: "The cluster command is malformed, has missing fields, or fields with invalid values. Command not carried out.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedAttribute",
    id: 134,
    description: "The indicated attribute ID, field ID or list entry does not exist for an attribute path.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "ConstraintError",
    id: 135,
    description: "Out of range error or set to a reserved value. Attribute keeps its old value. Note that an attribute value may be out of range if an attribute is related to another, e.g. with minimum and maximum attributes. See the individual attribute descriptions for specific details.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedWrite",
    id: 136,
    description: "Attempt to write a read-only attribute.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "ResourceExhausted",
    id: 137,
    description: "An action or operation failed due to insufficient available resources.INSUFFICIENT_SPACE is anobsolete name for this error code.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "NotFound",
    id: 139,
    description: "The indicated data field or entry could not be found.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnreportableAttribute",
    id: 140,
    description: "Reports cannot be issued for this attribute.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "InvalidDataType",
    id: 141,
    description: "The data type indicated is undefined or invalid for the indicated data field. Command or action not carried out.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedRead",
    id: 143,
    description: "Attempt to read a write-only attribute.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "DataVersionMismatch",
    id: 146,
    description: "Cluster instance data version did not match request path",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "Timeout",
    id: 148,
    description: "The transaction was aborted due to time being exceeded.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedNode",
    id: 155,
    description: "The node ID indicated is not supported on the node.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "Busy",
    id: 156,
    description: "The receiver is busy processing another action that prevents the execution of the incoming action.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "AccessRestricted",
    id: 157,
    description: "The access to the action or command by the sender is permitted by the ACL but restricted by the ARL.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedCluster",
    id: 195,
    description: "The cluster indicated is not supported on the endpoint.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "NoUpstreamSubscription",
    id: 197,
    description: "Used by proxies to convey to clients the lack of an upstream subscription to a source.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "NeedsTimedInteraction",
    id: 198,
    description: "A Untimed Write or Untimed Invoke interaction was used for an attribute or command that requires a Timed Write or Timed Invoke.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "UnsupportedEvent",
    id: 199,
    description: "The indicated event ID is not supported on the cluster instance.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "PathsExhausted",
    id: 200,
    description: "The receiver has insufficient resources to support the specified number of paths in the request",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "TimedRequestMismatch",
    id: 201,
    description: "A request with TimedRequest field set to TRUE was issued outside a Timed transaction or a request with TimedRequest set to FALSE was issued inside a Timed transaction.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "FailsafeRequired",
    id: 202,
    description: "A request requiring a Fail-safe context was invoked without the Fail-Safe context.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "InvalidInState",
    id: 203,
    description: "The received request cannot be handled due to the current operational state of the device",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "NoCommandResponse",
    id: 204,
    description: "A CommandDataIB is missing a response in the InvokeResponses of an Invoke Response action.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "TermsAndConditionsChanged",
    id: 205,
    description: "The node requires updated TC acceptance. The user MAY be directed to visit the EnhancedSetupFlowMaintenanceUrl to complete this.",
    xref: { document: "core", section: "8.10.1" }
  }),
  (0, import_elements.FieldElement)({
    name: "MaintenanceRequired",
    id: 206,
    description: "The node requires the user to visit the EnhancedSetupFlowMaintenanceUrl for instructions on further action.",
    xref: { document: "core", section: "8.10.1" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(status);
