// @matter/model/dist/cjs/standard/elements/systime-us.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var systime_us_exports = {};
__export(systime_us_exports, {
  systimeUs: () => systimeUs
});
module.exports = __toCommonJS(systime_us_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const systimeUs = (0, import_elements.DatatypeElement)({
  name: "systime-us",
  type: "uint64",
  description: "System Time in microseconds",
  isSeed: true,
  details: "System time in microseconds is an unsigned 64-bit value representing the number of microseconds since boot.",
  xref: { document: "core", section: "7.19.2.8" }
});
import_MatterDefinition.MatterDefinition.children.push(systimeUs);
