// @matter/model/dist/cjs/standard/elements/TemperatureSensorDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TemperatureSensorDT_exports = {};
__export(TemperatureSensorDT_exports, {
  TemperatureSensorDt: () => TemperatureSensorDt
});
module.exports = __toCommonJS(TemperatureSensorDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TemperatureSensorDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "TemperatureSensor",
    id: 770,
    category: "Sensor",
    classification: "simple",
    details: "A Temperature Sensor device reports measurements of temperature.",
    xref: { document: "device", section: "7.4" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 770, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.4.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.4.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(TemperatureSensorDt);
