// @matter/model/dist/cjs/standard/elements/temperature.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var temperature_exports = {};
__export(temperature_exports, {
  temperature: () => temperature
});
module.exports = __toCommonJS(temperature_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const temperature = (0, import_elements.DatatypeElement)(
  {
    name: "temperature",
    type: "int16",
    description: "Temperature",
    isSeed: true,
    details: "This type represents a temperature on the Celsius scale with a resolution of 0.01\xB0C.\n  \u2022 value = (temperature in \xB0C) x 100\nThe range is constrained by absolute zero: -273.15\xB0C to 327.67\xB0C.\nConversion of Temperature Values for Display\nWhen converting temperature values for display manufacturers SHOULD ensure that calculations\nround to the nearest representable value. Particular care is needed when using integer arithmetic.\nSample Conversion Code\nSample code provided to ensure consistent Fahrenheit to Celsius and vice-versa conversion between devices and across vendors.\nFor degF: the value is a int8u representing 2x temperature value in Fahrenheit (to get 0.5 resolution).\nFor degC: the value is a int16s representing Celsius in\n0.01 resolution as expected by the ZCL format.",
    xref: { document: "core", section: "7.19.2.11" }
  }
);
import_MatterDefinition.MatterDefinition.children.push(temperature);
