// @matter/model/dist/cjs/standard/elements/ThermostatDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThermostatDT_exports = {};
__export(ThermostatDT_exports, {
  ThermostatDt: () => ThermostatDt
});
module.exports = __toCommonJS(ThermostatDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ThermostatDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "Thermostat",
    id: 769,
    category: "HVAC",
    classification: "simple",
    details: "A Thermostat device is capable of having either built-in or separate sensors for temperature, humidity or occupancy. It allows the desired temperature to be set either remotely or locally. The thermostat is capable of sending heating and/or cooling requirement notifications to a heating/cooling unit (for example, an indoor air handler) or is capable of including a mechanism to control a heating or cooling unit directly.",
    xref: { document: "device", section: "9.1" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 769, revision: 4 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "Active",
    element: "serverCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "Thermostat",
      id: 513,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.1.4" }
    },
    (0, import_elements.RequirementElement)({ name: "SCHEDULECONFIGURATION", conformance: "X", element: "feature" }),
    (0, import_elements.RequirementElement)({ name: "AlarmMask", conformance: "X", element: "attribute" }),
    (0, import_elements.RequirementElement)({ name: "GetRelayStatusLog", conformance: "X", element: "command" }),
    (0, import_elements.RequirementElement)({ name: "GetRelayStatusLogResponse", conformance: "X", element: "command" })
  ),
  (0, import_elements.RequirementElement)({
    name: "ThermostatUserInterfaceConfiguration",
    id: 516,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "EnergyPreference",
    id: 155,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FanControl",
    id: 514,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "RelativeHumidityMeasurement",
    id: 1029,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "9.1.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "OccupancySensing",
    id: 1030,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "9.1.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(ThermostatDt);
