// @matter/model/dist/cjs/standard/elements/ThreadNetworkDirectory.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ThreadNetworkDirectory_exports = {};
__export(ThreadNetworkDirectory_exports, {
  ThreadNetworkDirectory: () => ThreadNetworkDirectory
});
module.exports = __toCommonJS(ThreadNetworkDirectory_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ThreadNetworkDirectory = (0, import_elements.ClusterElement)(
  {
    name: "ThreadNetworkDirectory",
    id: 1107,
    classification: "application",
    pics: "THNETDIR",
    details: "This cluster stores a list of Thread networks (including the credentials required to access each network), as well as a designation of the user\u2019s preferred network, to facilitate the sharing of Thread networks across fabrics.",
    xref: { document: "cluster", section: "10.4" }
  },
  (0, import_elements.AttributeElement)({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  (0, import_elements.AttributeElement)({
    name: "PreferredExtendedPanId",
    id: 0,
    type: "octstr",
    access: "RW VM",
    conformance: "M",
    constraint: "8",
    default: null,
    quality: "X N",
    details: "Indicates the Thread Extended PAN ID value for the Thread network designated by the user to be their preferred network for commissioning of Thread devices. If not null, the value of this attribute shall match the ExtendedPanID of a network in the ThreadNetworks attribute. A write operation with a non-null value that does not match any network in the ThreadNetworks list shall be rejected with a status of CONSTRAINT_ERROR.\nThe purpose of designating one Thread network as preferred is to help a commissioner to select a Thread network when a Thread device is within suitable range of more than one Thread network which appears in the ThreadNetworks list. A value of null indicates that there is no current preferred network: All networks may be treated as equally preferred by a commissioner with access to this cluster.\nThis attribute may be automatically set to the ExtendedPanID of the first Thread network added to the ThreadNetworks list.\nA client shall obtain user consent before changing the value of this attribute from a non-null value.\nOn a factory reset this attribute shall be reset to null.",
    xref: { document: "cluster", section: "10.4.5.1" }
  }),
  (0, import_elements.AttributeElement)(
    {
      name: "ThreadNetworks",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "M",
      constraint: "max threadNetworkTableSize",
      quality: "N",
      details: "Indicates the list of Thread Networks known about by this cluster. If the node hosting this cluster includes a Thread Border Router, then an entry for its Thread Network shall be included in this list.\nThe list can be modified via the AddNetwork and RemoveNetwork commands.\nFor each entry in the list, the cluster server also stores a Thread Operational Dataset. Clients use the GetOperationalDataset command to obtain the Operational Dataset for an entry in this list.\nOn a factory reset this list shall be cleared, and any Thread Operational datasets previously stored shall be removed from the Node.",
      xref: { document: "cluster", section: "10.4.5.2" }
    },
    (0, import_elements.FieldElement)({ name: "entry", type: "ThreadNetworkStruct" })
  ),
  (0, import_elements.AttributeElement)({
    name: "ThreadNetworkTableSize",
    id: 2,
    type: "uint8",
    access: "R V",
    conformance: "M",
    constraint: "desc",
    default: 10,
    quality: "F",
    details: "Indicates the maximum number of entries that can be held in the ThreadNetworks list; it shall be at least 2 times the number of SupportedFabrics advertised in the Operational Credentials Cluster on the root endpoint of this node.",
    xref: { document: "cluster", section: "10.4.5.3" }
  }),
  (0, import_elements.CommandElement)(
    {
      name: "AddNetwork",
      id: 0,
      access: "M T",
      conformance: "M",
      direction: "request",
      response: "status",
      details: "Adds an entry to the ThreadNetworks attribute with the specified Thread Operational Dataset.\nIf there is an existing entry with the Extended PAN ID then the Thread Operational Dataset for that entry is replaced. As a result, changes to the network parameters (e.g. Channel, Network Name, PSKc, \u2026) of an existing entry with a given Extended PAN ID can be made using this command.",
      xref: { document: "cluster", section: "10.4.6.1" }
    },
    (0, import_elements.FieldElement)({
      name: "OperationalDataset",
      id: 0,
      type: "octstr",
      conformance: "M",
      constraint: "max 254",
      details: "This field shall represent the Operational Dataset for the network, using the encoding defined in the Thread specification. It shall contain at least the following sub-TLVs: Active Timestamp, Channel, Channel Mask, Extended PAN ID, Network Key, Network Mesh-Local Prefix, Network Name, PAN ID, PSKc, and Security Policy.",
      xref: { document: "cluster", section: "10.4.6.1.1" }
    })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "RemoveNetwork",
      id: 1,
      access: "M T",
      conformance: "M",
      direction: "request",
      response: "status",
      details: "Removes the network with the given Extended PAN ID from the ThreadNetworks attribute.",
      xref: { document: "cluster", section: "10.4.6.2" }
    },
    (0, import_elements.FieldElement)({ name: "ExtendedPanId", id: 0, type: "octstr", conformance: "M", constraint: "8" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "GetOperationalDataset",
      id: 2,
      access: "M",
      conformance: "M",
      direction: "request",
      response: "OperationalDatasetResponse",
      details: "Retrieves the Thread Operational Dataset with the given Extended PAN ID.",
      xref: { document: "cluster", section: "10.4.6.3" }
    },
    (0, import_elements.FieldElement)({ name: "ExtendedPanId", id: 0, type: "octstr", conformance: "M", constraint: "8" })
  ),
  (0, import_elements.CommandElement)(
    {
      name: "OperationalDatasetResponse",
      id: 3,
      conformance: "M",
      direction: "response",
      details: "Contains the Thread Operational Dataset for the Extended PAN specified in GetOperationalDataset.",
      xref: { document: "cluster", section: "10.4.6.4" }
    },
    (0, import_elements.FieldElement)({ name: "OperationalDataset", id: 0, type: "octstr", conformance: "M", constraint: "max 254" })
  ),
  (0, import_elements.DatatypeElement)(
    {
      name: "ThreadNetworkStruct",
      type: "struct",
      details: "Represents the data associated with a Thread Network.",
      xref: { document: "cluster", section: "10.4.4.1" }
    },
    (0, import_elements.FieldElement)({
      name: "ExtendedPanId",
      id: 0,
      type: "octstr",
      conformance: "M",
      constraint: "8",
      details: "This field shall indicate the Extended PAN ID from the OperationalDataset for the given Thread network.",
      xref: { document: "cluster", section: "10.4.4.1.1" }
    }),
    (0, import_elements.FieldElement)({
      name: "NetworkName",
      id: 1,
      type: "string",
      conformance: "M",
      constraint: "1 to 16",
      details: "This field shall indicate the Network Name from the OperationalDataset for the given Thread network.",
      xref: { document: "cluster", section: "10.4.4.1.2" }
    }),
    (0, import_elements.FieldElement)({
      name: "Channel",
      id: 2,
      type: "uint16",
      conformance: "M",
      details: "This field shall indicate the Channel from the OperationalDataset for the given Thread network.",
      xref: { document: "cluster", section: "10.4.4.1.3" }
    }),
    (0, import_elements.FieldElement)({
      name: "ActiveTimestamp",
      id: 3,
      type: "uint64",
      conformance: "M",
      details: "This field shall indicate the Active Timestamp from the OperationalDataset for the given Thread network.",
      xref: { document: "cluster", section: "10.4.4.1.4" }
    })
  )
);
import_MatterDefinition.MatterDefinition.children.push(ThreadNetworkDirectory);
