// @matter/model/dist/cjs/standard/elements/uint56.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var uint56_exports = {};
__export(uint56_exports, {
  uint56: () => uint56
});
module.exports = __toCommonJS(uint56_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const uint56 = (0, import_elements.DatatypeElement)({
  name: "uint56",
  byteSize: 7,
  description: "Unsigned 56-bit integer",
  isSeed: true,
  metatype: "integer",
  xref: { document: "core", section: "7.19.1" }
});
import_MatterDefinition.MatterDefinition.children.push(uint56);
