// @matter/model/dist/cjs/standard/elements/vendor-id.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var vendor_id_exports = {};
__export(vendor_id_exports, {
  vendorId: () => vendorId
});
module.exports = __toCommonJS(vendor_id_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const vendorId = (0, import_elements.DatatypeElement)({
  name: "vendor-id",
  type: "uint16",
  description: "Vendor ID",
  isSeed: true,
  details: "A Vendor ID.\nVendor IDs may be used as a prefix in a Manufacturer Extensible Identifier format.",
  xref: { document: "core", section: "7.19.2.24" }
});
import_MatterDefinition.MatterDefinition.children.push(vendorId);
