// @matter/model/dist/cjs/standard/elements/VideoRemoteControlDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VideoRemoteControlDT_exports = {};
__export(VideoRemoteControlDT_exports, {
  VideoRemoteControlDt: () => VideoRemoteControlDt
});
module.exports = __toCommonJS(VideoRemoteControlDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const VideoRemoteControlDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "VideoRemoteControl",
    id: 42,
    category: "Media",
    classification: "simple",
    details: "This defines conformance to the Video Remote Control device type.\nA Video Remote Control is a client that can control a Video Player, for example, a traditional universal remote control.",
    xref: { document: "device", section: "10.7" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 42, revision: 2 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "LevelControl",
    id: 8,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "WakeOnLan",
    id: 1283,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Channel",
    id: 1284,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "TargetNavigator",
    id: 1285,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "MediaPlayback",
    id: 1286,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "MediaInput",
    id: 1287,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "LowPower",
    id: 1288,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "KeypadInput",
    id: 1289,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ContentLauncher",
    id: 1290,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "AudioOutput",
    id: 1291,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ApplicationLauncher",
    id: 1292,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "AccountLogin",
    id: 1294,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ContentControl",
    id: 1295,
    conformance: "P, O",
    element: "clientCluster",
    xref: { document: "device", section: "10.7.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(VideoRemoteControlDt);
