// @matter/model/dist/cjs/standard/elements/WaterValveDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WaterValveDT_exports = {};
__export(WaterValveDT_exports, {
  WaterValveDt: () => WaterValveDt
});
module.exports = __toCommonJS(WaterValveDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WaterValveDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "WaterValve",
    id: 66,
    category: "Smart Plugs/Outlets and other Actuators",
    classification: "simple",
    details: "This defines conformance to the Water Valve device type.",
    xref: { document: "device", section: "5.6" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 66, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "ValveConfigurationAndControl",
    id: 129,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.6.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.6.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(WaterValveDt);
