// @matter/model/dist/cjs/standard/elements/WaterHeaterDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WaterHeaterDT_exports = {};
__export(WaterHeaterDT_exports, {
  WaterHeaterDt: () => WaterHeaterDt
});
module.exports = __toCommonJS(WaterHeaterDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WaterHeaterDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "WaterHeater",
    id: 1295,
    category: "Energy",
    classification: "simple",
    details: "A water heater is a device that is generally installed in properties to heat water for showers, baths etc.",
    xref: { document: "device", section: "14.2" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 1295, revision: 1 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.2.6" }
  }),
  (0, import_elements.RequirementElement)(
    {
      name: "Thermostat",
      id: 513,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "14.2.6" }
    },
    (0, import_elements.RequirementElement)({ name: "HEATING", conformance: "M", element: "feature" })
  ),
  (0, import_elements.RequirementElement)({
    name: "WaterHeaterManagement",
    id: 148,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.2.6" }
  }),
  (0, import_elements.RequirementElement)({
    name: "WaterHeaterMode",
    id: 158,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.2.6" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(WaterHeaterDt);
