// @matter/model/dist/cjs/standard/elements/WindowCoveringDT.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var WindowCoveringDT_exports = {};
__export(WindowCoveringDT_exports, {
  WindowCoveringDt: () => WindowCoveringDt
});
module.exports = __toCommonJS(WindowCoveringDT_exports);
var import_MatterDefinition = require("../MatterDefinition.js");
var import_elements = require("../../elements/index.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const WindowCoveringDt = (0, import_elements.DeviceTypeElement)(
  {
    name: "WindowCovering",
    id: 514,
    category: "Closure",
    classification: "simple",
    details: "This defines conformance to the Window Covering device type.",
    xref: { document: "device", section: "8.3" }
  },
  (0, import_elements.RequirementElement)(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    (0, import_elements.RequirementElement)({ name: "DeviceTypeList", default: [{ deviceType: 514, revision: 3 }], element: "attribute" })
  ),
  (0, import_elements.RequirementElement)({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "8.3.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "Groups",
    id: 4,
    conformance: "Active, O",
    element: "serverCluster",
    xref: { document: "device", section: "8.3.4" }
  }),
  (0, import_elements.RequirementElement)({
    name: "WindowCovering",
    id: 258,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "8.3.4" }
  })
);
import_MatterDefinition.MatterDefinition.children.push(WindowCoveringDt);
