// @matter/node/dist/cjs/devices/basic-video-player.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var basic_video_player_exports = {};
__export(basic_video_player_exports, {
  BasicVideoPlayerDevice: () => BasicVideoPlayerDevice,
  BasicVideoPlayerDeviceDefinition: () => BasicVideoPlayerDeviceDefinition,
  BasicVideoPlayerRequirements: () => BasicVideoPlayerRequirements
});
module.exports = __toCommonJS(basic_video_player_exports);
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_MediaPlaybackServer = require("../behaviors/media-playback/MediaPlaybackServer.js");
var import_KeypadInputServer = require("../behaviors/keypad-input/KeypadInputServer.js");
var import_WakeOnLanServer = require("../behaviors/wake-on-lan/WakeOnLanServer.js");
var import_ChannelServer = require("../behaviors/channel/ChannelServer.js");
var import_TargetNavigatorServer = require("../behaviors/target-navigator/TargetNavigatorServer.js");
var import_MediaInputServer = require("../behaviors/media-input/MediaInputServer.js");
var import_LowPowerServer = require("../behaviors/low-power/LowPowerServer.js");
var import_AudioOutputServer = require("../behaviors/audio-output/AudioOutputServer.js");
var import_ContentControlServer = require("../behaviors/content-control/ContentControlServer.js");
var import_MessagesServer = require("../behaviors/messages/MessagesServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var BasicVideoPlayerRequirements;
((BasicVideoPlayerRequirements2) => {
  BasicVideoPlayerRequirements2.OnOffServer = import_OnOffServer.OnOffServer;
  BasicVideoPlayerRequirements2.MediaPlaybackServer = import_MediaPlaybackServer.MediaPlaybackServer;
  BasicVideoPlayerRequirements2.KeypadInputServer = import_KeypadInputServer.KeypadInputServer;
  BasicVideoPlayerRequirements2.WakeOnLanServer = import_WakeOnLanServer.WakeOnLanServer;
  BasicVideoPlayerRequirements2.ChannelServer = import_ChannelServer.ChannelServer;
  BasicVideoPlayerRequirements2.TargetNavigatorServer = import_TargetNavigatorServer.TargetNavigatorServer;
  BasicVideoPlayerRequirements2.MediaInputServer = import_MediaInputServer.MediaInputServer;
  BasicVideoPlayerRequirements2.LowPowerServer = import_LowPowerServer.LowPowerServer;
  BasicVideoPlayerRequirements2.AudioOutputServer = import_AudioOutputServer.AudioOutputServer;
  BasicVideoPlayerRequirements2.ContentControlServer = import_ContentControlServer.ContentControlServer;
  BasicVideoPlayerRequirements2.MessagesServer = import_MessagesServer.MessagesServer;
  BasicVideoPlayerRequirements2.server = {
    mandatory: { OnOff: BasicVideoPlayerRequirements2.OnOffServer, MediaPlayback: BasicVideoPlayerRequirements2.MediaPlaybackServer, KeypadInput: BasicVideoPlayerRequirements2.KeypadInputServer },
    optional: {
      WakeOnLan: BasicVideoPlayerRequirements2.WakeOnLanServer,
      Channel: BasicVideoPlayerRequirements2.ChannelServer,
      TargetNavigator: BasicVideoPlayerRequirements2.TargetNavigatorServer,
      MediaInput: BasicVideoPlayerRequirements2.MediaInputServer,
      LowPower: BasicVideoPlayerRequirements2.LowPowerServer,
      AudioOutput: BasicVideoPlayerRequirements2.AudioOutputServer,
      ContentControl: BasicVideoPlayerRequirements2.ContentControlServer,
      Messages: BasicVideoPlayerRequirements2.MessagesServer
    }
  };
})(BasicVideoPlayerRequirements || (BasicVideoPlayerRequirements = {}));
const BasicVideoPlayerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "BasicVideoPlayer",
  deviceType: 40,
  deviceRevision: 2,
  requirements: BasicVideoPlayerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    BasicVideoPlayerRequirements.server.mandatory.OnOff,
    BasicVideoPlayerRequirements.server.mandatory.MediaPlayback,
    BasicVideoPlayerRequirements.server.mandatory.KeypadInput
  )
});
const BasicVideoPlayerDevice = BasicVideoPlayerDeviceDefinition;
