// @matter/node/dist/cjs/devices/casting-video-client.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var casting_video_client_exports = {};
__export(casting_video_client_exports, {
  CastingVideoClientDevice: () => CastingVideoClientDevice,
  CastingVideoClientDeviceDefinition: () => CastingVideoClientDeviceDefinition,
  CastingVideoClientRequirements: () => CastingVideoClientRequirements
});
module.exports = __toCommonJS(casting_video_client_exports);
var import_ContentAppObserverServer = require("../behaviors/content-app-observer/ContentAppObserverServer.js");
var import_OnOffBehavior = require("../behaviors/on-off/OnOffBehavior.js");
var import_KeypadInputBehavior = require("../behaviors/keypad-input/KeypadInputBehavior.js");
var import_ContentLauncherBehavior = require("../behaviors/content-launcher/ContentLauncherBehavior.js");
var import_ApplicationBasicBehavior = require("../behaviors/application-basic/ApplicationBasicBehavior.js");
var import_LevelControlBehavior = require("../behaviors/level-control/LevelControlBehavior.js");
var import_WakeOnLanBehavior = require("../behaviors/wake-on-lan/WakeOnLanBehavior.js");
var import_ChannelBehavior = require("../behaviors/channel/ChannelBehavior.js");
var import_TargetNavigatorBehavior = require("../behaviors/target-navigator/TargetNavigatorBehavior.js");
var import_MediaPlaybackBehavior = require("../behaviors/media-playback/MediaPlaybackBehavior.js");
var import_MediaInputBehavior = require("../behaviors/media-input/MediaInputBehavior.js");
var import_LowPowerBehavior = require("../behaviors/low-power/LowPowerBehavior.js");
var import_AudioOutputBehavior = require("../behaviors/audio-output/AudioOutputBehavior.js");
var import_ApplicationLauncherBehavior = require("../behaviors/application-launcher/ApplicationLauncherBehavior.js");
var import_AccountLoginBehavior = require("../behaviors/account-login/AccountLoginBehavior.js");
var import_ContentControlBehavior = require("../behaviors/content-control/ContentControlBehavior.js");
var import_MessagesBehavior = require("../behaviors/messages/MessagesBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var CastingVideoClientRequirements;
((CastingVideoClientRequirements2) => {
  CastingVideoClientRequirements2.ContentAppObserverServer = import_ContentAppObserverServer.ContentAppObserverServer;
  CastingVideoClientRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  CastingVideoClientRequirements2.KeypadInputBehavior = import_KeypadInputBehavior.KeypadInputBehavior;
  CastingVideoClientRequirements2.ContentLauncherBehavior = import_ContentLauncherBehavior.ContentLauncherBehavior;
  CastingVideoClientRequirements2.ApplicationBasicBehavior = import_ApplicationBasicBehavior.ApplicationBasicBehavior;
  CastingVideoClientRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  CastingVideoClientRequirements2.WakeOnLanBehavior = import_WakeOnLanBehavior.WakeOnLanBehavior;
  CastingVideoClientRequirements2.ChannelBehavior = import_ChannelBehavior.ChannelBehavior;
  CastingVideoClientRequirements2.TargetNavigatorBehavior = import_TargetNavigatorBehavior.TargetNavigatorBehavior;
  CastingVideoClientRequirements2.MediaPlaybackBehavior = import_MediaPlaybackBehavior.MediaPlaybackBehavior;
  CastingVideoClientRequirements2.MediaInputBehavior = import_MediaInputBehavior.MediaInputBehavior;
  CastingVideoClientRequirements2.LowPowerBehavior = import_LowPowerBehavior.LowPowerBehavior;
  CastingVideoClientRequirements2.AudioOutputBehavior = import_AudioOutputBehavior.AudioOutputBehavior;
  CastingVideoClientRequirements2.ApplicationLauncherBehavior = import_ApplicationLauncherBehavior.ApplicationLauncherBehavior;
  CastingVideoClientRequirements2.AccountLoginBehavior = import_AccountLoginBehavior.AccountLoginBehavior;
  CastingVideoClientRequirements2.ContentControlBehavior = import_ContentControlBehavior.ContentControlBehavior;
  CastingVideoClientRequirements2.MessagesBehavior = import_MessagesBehavior.MessagesBehavior;
  CastingVideoClientRequirements2.server = { optional: { ContentAppObserver: CastingVideoClientRequirements2.ContentAppObserverServer }, mandatory: {} };
  CastingVideoClientRequirements2.client = {
    mandatory: {
      OnOff: CastingVideoClientRequirements2.OnOffBehavior,
      KeypadInput: CastingVideoClientRequirements2.KeypadInputBehavior,
      ContentLauncher: CastingVideoClientRequirements2.ContentLauncherBehavior,
      ApplicationBasic: CastingVideoClientRequirements2.ApplicationBasicBehavior
    },
    optional: {
      LevelControl: CastingVideoClientRequirements2.LevelControlBehavior,
      WakeOnLan: CastingVideoClientRequirements2.WakeOnLanBehavior,
      Channel: CastingVideoClientRequirements2.ChannelBehavior,
      TargetNavigator: CastingVideoClientRequirements2.TargetNavigatorBehavior,
      MediaPlayback: CastingVideoClientRequirements2.MediaPlaybackBehavior,
      MediaInput: CastingVideoClientRequirements2.MediaInputBehavior,
      LowPower: CastingVideoClientRequirements2.LowPowerBehavior,
      AudioOutput: CastingVideoClientRequirements2.AudioOutputBehavior,
      ApplicationLauncher: CastingVideoClientRequirements2.ApplicationLauncherBehavior,
      AccountLogin: CastingVideoClientRequirements2.AccountLoginBehavior,
      ContentControl: CastingVideoClientRequirements2.ContentControlBehavior,
      Messages: CastingVideoClientRequirements2.MessagesBehavior
    }
  };
})(CastingVideoClientRequirements || (CastingVideoClientRequirements = {}));
const CastingVideoClientDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "CastingVideoClient",
  deviceType: 41,
  deviceRevision: 2,
  requirements: CastingVideoClientRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const CastingVideoClientDevice = CastingVideoClientDeviceDefinition;
