// @matter/node/dist/cjs/devices/color-temperature-light.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var color_temperature_light_exports = {};
__export(color_temperature_light_exports, {
  ColorTemperatureLightDevice: () => ColorTemperatureLightDevice,
  ColorTemperatureLightDeviceDefinition: () => ColorTemperatureLightDeviceDefinition,
  ColorTemperatureLightRequirements: () => ColorTemperatureLightRequirements
});
module.exports = __toCommonJS(color_temperature_light_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_LevelControlServer = require("../behaviors/level-control/LevelControlServer.js");
var import_ColorControlServer = require("../behaviors/color-control/ColorControlServer.js");
var import_ScenesManagementServer = require("../behaviors/scenes-management/ScenesManagementServer.js");
var import_OccupancySensingBehavior = require("../behaviors/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ColorTemperatureLightRequirements;
((ColorTemperatureLightRequirements2) => {
  ColorTemperatureLightRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  ColorTemperatureLightRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  ColorTemperatureLightRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("Lighting");
  ColorTemperatureLightRequirements2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("OnOff", "Lighting").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  ColorTemperatureLightRequirements2.ColorControlServer = import_ColorControlServer.ColorControlServer.with("ColorTemperature").alter({ attributes: { remainingTime: { optional: false } } });
  ColorTemperatureLightRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  ColorTemperatureLightRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  ColorTemperatureLightRequirements2.server = {
    mandatory: {
      Identify: ColorTemperatureLightRequirements2.IdentifyServer,
      Groups: ColorTemperatureLightRequirements2.GroupsServer,
      OnOff: ColorTemperatureLightRequirements2.OnOffServer,
      LevelControl: ColorTemperatureLightRequirements2.LevelControlServer,
      ColorControl: ColorTemperatureLightRequirements2.ColorControlServer
    },
    optional: { ScenesManagement: ColorTemperatureLightRequirements2.ScenesManagementServer }
  };
  ColorTemperatureLightRequirements2.client = { optional: { OccupancySensing: ColorTemperatureLightRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(ColorTemperatureLightRequirements || (ColorTemperatureLightRequirements = {}));
const ColorTemperatureLightDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ColorTemperatureLight",
  deviceType: 268,
  deviceRevision: 4,
  requirements: ColorTemperatureLightRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    ColorTemperatureLightRequirements.server.mandatory.Identify,
    ColorTemperatureLightRequirements.server.mandatory.Groups,
    ColorTemperatureLightRequirements.server.mandatory.OnOff,
    ColorTemperatureLightRequirements.server.mandatory.LevelControl,
    ColorTemperatureLightRequirements.server.mandatory.ColorControl
  )
});
const ColorTemperatureLightDevice = ColorTemperatureLightDeviceDefinition;
