// @matter/node/dist/cjs/devices/contact-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var contact_sensor_exports = {};
__export(contact_sensor_exports, {
  ContactSensorDevice: () => ContactSensorDevice,
  ContactSensorDeviceDefinition: () => ContactSensorDeviceDefinition,
  ContactSensorRequirements: () => ContactSensorRequirements
});
module.exports = __toCommonJS(contact_sensor_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_BooleanStateServer = require("../behaviors/boolean-state/BooleanStateServer.js");
var import_BooleanStateConfigurationServer = require("../behaviors/boolean-state-configuration/BooleanStateConfigurationServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ContactSensorRequirements;
((ContactSensorRequirements2) => {
  ContactSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  ContactSensorRequirements2.BooleanStateServer = import_BooleanStateServer.BooleanStateServer;
  ContactSensorRequirements2.BooleanStateConfigurationServer = import_BooleanStateConfigurationServer.BooleanStateConfigurationServer;
  ContactSensorRequirements2.server = {
    mandatory: { Identify: ContactSensorRequirements2.IdentifyServer, BooleanState: ContactSensorRequirements2.BooleanStateServer },
    optional: { BooleanStateConfiguration: ContactSensorRequirements2.BooleanStateConfigurationServer }
  };
})(ContactSensorRequirements || (ContactSensorRequirements = {}));
const ContactSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ContactSensor",
  deviceType: 21,
  deviceRevision: 2,
  requirements: ContactSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    ContactSensorRequirements.server.mandatory.Identify,
    ContactSensorRequirements.server.mandatory.BooleanState
  )
});
const ContactSensorDevice = ContactSensorDeviceDefinition;
