// @matter/node/dist/cjs/devices/cooktop.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var cooktop_exports = {};
__export(cooktop_exports, {
  CooktopDevice: () => CooktopDevice,
  CooktopDeviceDefinition: () => CooktopDeviceDefinition,
  CooktopRequirements: () => CooktopRequirements
});
module.exports = __toCommonJS(cooktop_exports);
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var CooktopRequirements;
((CooktopRequirements2) => {
  CooktopRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("OffOnly");
  CooktopRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  CooktopRequirements2.server = { mandatory: { OnOff: CooktopRequirements2.OnOffServer }, optional: { Identify: CooktopRequirements2.IdentifyServer } };
})(CooktopRequirements || (CooktopRequirements = {}));
const CooktopDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "Cooktop",
  deviceType: 120,
  deviceRevision: 1,
  requirements: CooktopRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(CooktopRequirements.server.mandatory.OnOff)
});
const CooktopDevice = CooktopDeviceDefinition;
