// @matter/node/dist/cjs/devices/dimmer-switch.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dimmer_switch_exports = {};
__export(dimmer_switch_exports, {
  DimmerSwitchDevice: () => DimmerSwitchDevice,
  DimmerSwitchDeviceDefinition: () => DimmerSwitchDeviceDefinition,
  DimmerSwitchRequirements: () => DimmerSwitchRequirements
});
module.exports = __toCommonJS(dimmer_switch_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_IdentifyBehavior = require("../behaviors/identify/IdentifyBehavior.js");
var import_OnOffBehavior = require("../behaviors/on-off/OnOffBehavior.js");
var import_LevelControlBehavior = require("../behaviors/level-control/LevelControlBehavior.js");
var import_GroupsBehavior = require("../behaviors/groups/GroupsBehavior.js");
var import_ScenesManagementBehavior = require("../behaviors/scenes-management/ScenesManagementBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DimmerSwitchRequirements;
((DimmerSwitchRequirements2) => {
  DimmerSwitchRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  DimmerSwitchRequirements2.IdentifyBehavior = import_IdentifyBehavior.IdentifyBehavior;
  DimmerSwitchRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  DimmerSwitchRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  DimmerSwitchRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  DimmerSwitchRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  DimmerSwitchRequirements2.server = { mandatory: { Identify: DimmerSwitchRequirements2.IdentifyServer } };
  DimmerSwitchRequirements2.client = {
    mandatory: { Identify: DimmerSwitchRequirements2.IdentifyBehavior, OnOff: DimmerSwitchRequirements2.OnOffBehavior, LevelControl: DimmerSwitchRequirements2.LevelControlBehavior },
    optional: { Groups: DimmerSwitchRequirements2.GroupsBehavior, ScenesManagement: DimmerSwitchRequirements2.ScenesManagementBehavior }
  };
})(DimmerSwitchRequirements || (DimmerSwitchRequirements = {}));
const DimmerSwitchDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "DimmerSwitch",
  deviceType: 260,
  deviceRevision: 3,
  requirements: DimmerSwitchRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(DimmerSwitchRequirements.server.mandatory.Identify)
});
const DimmerSwitchDevice = DimmerSwitchDeviceDefinition;
