// @matter/node/dist/cjs/devices/dimmable-light.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var dimmable_light_exports = {};
__export(dimmable_light_exports, {
  DimmableLightDevice: () => DimmableLightDevice,
  DimmableLightDeviceDefinition: () => DimmableLightDeviceDefinition,
  DimmableLightRequirements: () => DimmableLightRequirements
});
module.exports = __toCommonJS(dimmable_light_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_LevelControlServer = require("../behaviors/level-control/LevelControlServer.js");
var import_ScenesManagementServer = require("../behaviors/scenes-management/ScenesManagementServer.js");
var import_OccupancySensingBehavior = require("../behaviors/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DimmableLightRequirements;
((DimmableLightRequirements2) => {
  DimmableLightRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  DimmableLightRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  DimmableLightRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("Lighting");
  DimmableLightRequirements2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("Lighting", "OnOff").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  DimmableLightRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  DimmableLightRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  DimmableLightRequirements2.server = {
    mandatory: {
      Identify: DimmableLightRequirements2.IdentifyServer,
      Groups: DimmableLightRequirements2.GroupsServer,
      OnOff: DimmableLightRequirements2.OnOffServer,
      LevelControl: DimmableLightRequirements2.LevelControlServer
    },
    optional: { ScenesManagement: DimmableLightRequirements2.ScenesManagementServer }
  };
  DimmableLightRequirements2.client = { optional: { OccupancySensing: DimmableLightRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(DimmableLightRequirements || (DimmableLightRequirements = {}));
const DimmableLightDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "DimmableLight",
  deviceType: 257,
  deviceRevision: 3,
  requirements: DimmableLightRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    DimmableLightRequirements.server.mandatory.Identify,
    DimmableLightRequirements.server.mandatory.Groups,
    DimmableLightRequirements.server.mandatory.OnOff,
    DimmableLightRequirements.server.mandatory.LevelControl
  )
});
const DimmableLightDevice = DimmableLightDeviceDefinition;
