// @matter/node/dist/cjs/devices/door-lock-controller.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var door_lock_controller_exports = {};
__export(door_lock_controller_exports, {
  DoorLockControllerDevice: () => DoorLockControllerDevice,
  DoorLockControllerDeviceDefinition: () => DoorLockControllerDeviceDefinition,
  DoorLockControllerRequirements: () => DoorLockControllerRequirements
});
module.exports = __toCommonJS(door_lock_controller_exports);
var import_TimeSynchronizationServer = require("../behaviors/time-synchronization/TimeSynchronizationServer.js");
var import_DoorLockBehavior = require("../behaviors/door-lock/DoorLockBehavior.js");
var import_GroupsBehavior = require("../behaviors/groups/GroupsBehavior.js");
var import_ScenesManagementBehavior = require("../behaviors/scenes-management/ScenesManagementBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DoorLockControllerRequirements;
((DoorLockControllerRequirements2) => {
  DoorLockControllerRequirements2.TimeSynchronizationServer = import_TimeSynchronizationServer.TimeSynchronizationServer;
  DoorLockControllerRequirements2.DoorLockBehavior = import_DoorLockBehavior.DoorLockBehavior;
  DoorLockControllerRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  DoorLockControllerRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  DoorLockControllerRequirements2.server = { optional: { TimeSynchronization: DoorLockControllerRequirements2.TimeSynchronizationServer }, mandatory: {} };
  DoorLockControllerRequirements2.client = {
    mandatory: { DoorLock: DoorLockControllerRequirements2.DoorLockBehavior },
    optional: { Groups: DoorLockControllerRequirements2.GroupsBehavior, ScenesManagement: DoorLockControllerRequirements2.ScenesManagementBehavior }
  };
})(DoorLockControllerRequirements || (DoorLockControllerRequirements = {}));
const DoorLockControllerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "DoorLockController",
  deviceType: 11,
  deviceRevision: 3,
  requirements: DoorLockControllerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const DoorLockControllerDevice = DoorLockControllerDeviceDefinition;
