// @matter/node/dist/cjs/devices/door-lock.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var door_lock_exports = {};
__export(door_lock_exports, {
  DoorLockDevice: () => DoorLockDevice,
  DoorLockDeviceDefinition: () => DoorLockDeviceDefinition,
  DoorLockRequirements: () => DoorLockRequirements
});
module.exports = __toCommonJS(door_lock_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_DoorLockServer = require("../behaviors/door-lock/DoorLockServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_ScenesManagementServer = require("../behaviors/scenes-management/ScenesManagementServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var DoorLockRequirements;
((DoorLockRequirements2) => {
  DoorLockRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  DoorLockRequirements2.DoorLockServer = import_DoorLockServer.DoorLockServer;
  DoorLockRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  DoorLockRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  DoorLockRequirements2.server = {
    mandatory: { Identify: DoorLockRequirements2.IdentifyServer, DoorLock: DoorLockRequirements2.DoorLockServer },
    optional: { Groups: DoorLockRequirements2.GroupsServer, ScenesManagement: DoorLockRequirements2.ScenesManagementServer }
  };
})(DoorLockRequirements || (DoorLockRequirements = {}));
const DoorLockDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "DoorLock",
  deviceType: 10,
  deviceRevision: 3,
  requirements: DoorLockRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    DoorLockRequirements.server.mandatory.Identify,
    DoorLockRequirements.server.mandatory.DoorLock
  )
});
const DoorLockDevice = DoorLockDeviceDefinition;
