// @matter/node/dist/cjs/devices/extended-color-light.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var extended_color_light_exports = {};
__export(extended_color_light_exports, {
  ExtendedColorLightDevice: () => ExtendedColorLightDevice,
  ExtendedColorLightDeviceDefinition: () => ExtendedColorLightDeviceDefinition,
  ExtendedColorLightRequirements: () => ExtendedColorLightRequirements
});
module.exports = __toCommonJS(extended_color_light_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_LevelControlServer = require("../behaviors/level-control/LevelControlServer.js");
var import_ColorControlServer = require("../behaviors/color-control/ColorControlServer.js");
var import_ScenesManagementServer = require("../behaviors/scenes-management/ScenesManagementServer.js");
var import_OccupancySensingBehavior = require("../behaviors/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ExtendedColorLightRequirements;
((ExtendedColorLightRequirements2) => {
  ExtendedColorLightRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  ExtendedColorLightRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  ExtendedColorLightRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("Lighting");
  ExtendedColorLightRequirements2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("OnOff", "Lighting").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  ExtendedColorLightRequirements2.ColorControlServer = import_ColorControlServer.ColorControlServer.with("Xy", "ColorTemperature").alter({ attributes: { remainingTime: { optional: false } } });
  ExtendedColorLightRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  ExtendedColorLightRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  ExtendedColorLightRequirements2.server = {
    mandatory: {
      Identify: ExtendedColorLightRequirements2.IdentifyServer,
      Groups: ExtendedColorLightRequirements2.GroupsServer,
      OnOff: ExtendedColorLightRequirements2.OnOffServer,
      LevelControl: ExtendedColorLightRequirements2.LevelControlServer,
      ColorControl: ExtendedColorLightRequirements2.ColorControlServer
    },
    optional: { ScenesManagement: ExtendedColorLightRequirements2.ScenesManagementServer }
  };
  ExtendedColorLightRequirements2.client = { optional: { OccupancySensing: ExtendedColorLightRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(ExtendedColorLightRequirements || (ExtendedColorLightRequirements = {}));
const ExtendedColorLightDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ExtendedColorLight",
  deviceType: 269,
  deviceRevision: 4,
  requirements: ExtendedColorLightRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    ExtendedColorLightRequirements.server.mandatory.Identify,
    ExtendedColorLightRequirements.server.mandatory.Groups,
    ExtendedColorLightRequirements.server.mandatory.OnOff,
    ExtendedColorLightRequirements.server.mandatory.LevelControl,
    ExtendedColorLightRequirements.server.mandatory.ColorControl
  )
});
const ExtendedColorLightDevice = ExtendedColorLightDeviceDefinition;
