// @matter/node/dist/cjs/devices/extractor-hood.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var extractor_hood_exports = {};
__export(extractor_hood_exports, {
  ExtractorHoodDevice: () => ExtractorHoodDevice,
  ExtractorHoodDeviceDefinition: () => ExtractorHoodDeviceDefinition,
  ExtractorHoodRequirements: () => ExtractorHoodRequirements
});
module.exports = __toCommonJS(extractor_hood_exports);
var import_FanControlServer = require("../behaviors/fan-control/FanControlServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_HepaFilterMonitoringServer = require("../behaviors/hepa-filter-monitoring/HepaFilterMonitoringServer.js");
var import_ActivatedCarbonFilterMonitoringServer = require("../behaviors/activated-carbon-filter-monitoring/ActivatedCarbonFilterMonitoringServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ExtractorHoodRequirements;
((ExtractorHoodRequirements2) => {
  ExtractorHoodRequirements2.FanControlServer = import_FanControlServer.FanControlServer;
  ExtractorHoodRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  ExtractorHoodRequirements2.HepaFilterMonitoringServer = import_HepaFilterMonitoringServer.HepaFilterMonitoringServer;
  ExtractorHoodRequirements2.ActivatedCarbonFilterMonitoringServer = import_ActivatedCarbonFilterMonitoringServer.ActivatedCarbonFilterMonitoringServer;
  ExtractorHoodRequirements2.server = {
    mandatory: { FanControl: ExtractorHoodRequirements2.FanControlServer },
    optional: {
      Identify: ExtractorHoodRequirements2.IdentifyServer,
      HepaFilterMonitoring: ExtractorHoodRequirements2.HepaFilterMonitoringServer,
      ActivatedCarbonFilterMonitoring: ExtractorHoodRequirements2.ActivatedCarbonFilterMonitoringServer
    }
  };
})(ExtractorHoodRequirements || (ExtractorHoodRequirements = {}));
const ExtractorHoodDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ExtractorHood",
  deviceType: 122,
  deviceRevision: 1,
  requirements: ExtractorHoodRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(ExtractorHoodRequirements.server.mandatory.FanControl)
});
const ExtractorHoodDevice = ExtractorHoodDeviceDefinition;
