// @matter/node/dist/cjs/devices/fan.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fan_exports = {};
__export(fan_exports, {
  FanDevice: () => FanDevice,
  FanDeviceDefinition: () => FanDeviceDefinition,
  FanRequirements: () => FanRequirements
});
module.exports = __toCommonJS(fan_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_FanControlServer = require("../behaviors/fan-control/FanControlServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FanRequirements;
((FanRequirements2) => {
  FanRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  FanRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  FanRequirements2.FanControlServer = import_FanControlServer.FanControlServer;
  FanRequirements2.OnOffServer = import_OnOffServer.OnOffServer;
  FanRequirements2.server = {
    mandatory: { Identify: FanRequirements2.IdentifyServer, Groups: FanRequirements2.GroupsServer, FanControl: FanRequirements2.FanControlServer },
    optional: { OnOff: FanRequirements2.OnOffServer }
  };
})(FanRequirements || (FanRequirements = {}));
const FanDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "Fan",
  deviceType: 43,
  deviceRevision: 3,
  requirements: FanRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    FanRequirements.server.mandatory.Identify,
    FanRequirements.server.mandatory.Groups,
    FanRequirements.server.mandatory.FanControl
  )
});
const FanDevice = FanDeviceDefinition;
