// @matter/node/dist/cjs/devices/flow-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var flow_sensor_exports = {};
__export(flow_sensor_exports, {
  FlowSensorDevice: () => FlowSensorDevice,
  FlowSensorDeviceDefinition: () => FlowSensorDeviceDefinition,
  FlowSensorRequirements: () => FlowSensorRequirements
});
module.exports = __toCommonJS(flow_sensor_exports);
var import_FlowMeasurementServer = require("../behaviors/flow-measurement/FlowMeasurementServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FlowSensorRequirements;
((FlowSensorRequirements2) => {
  FlowSensorRequirements2.FlowMeasurementServer = import_FlowMeasurementServer.FlowMeasurementServer;
  FlowSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  FlowSensorRequirements2.server = { mandatory: { FlowMeasurement: FlowSensorRequirements2.FlowMeasurementServer, Identify: FlowSensorRequirements2.IdentifyServer } };
})(FlowSensorRequirements || (FlowSensorRequirements = {}));
const FlowSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "FlowSensor",
  deviceType: 774,
  deviceRevision: 2,
  requirements: FlowSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    FlowSensorRequirements.server.mandatory.FlowMeasurement,
    FlowSensorRequirements.server.mandatory.Identify
  )
});
const FlowSensorDevice = FlowSensorDeviceDefinition;
