// @matter/node/dist/cjs/devices/generic-switch.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var generic_switch_exports = {};
__export(generic_switch_exports, {
  GenericSwitchDevice: () => GenericSwitchDevice,
  GenericSwitchDeviceDefinition: () => GenericSwitchDeviceDefinition,
  GenericSwitchRequirements: () => GenericSwitchRequirements
});
module.exports = __toCommonJS(generic_switch_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_SwitchServer = require("../behaviors/switch/SwitchServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var GenericSwitchRequirements;
((GenericSwitchRequirements2) => {
  GenericSwitchRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  GenericSwitchRequirements2.SwitchServer = import_SwitchServer.SwitchServer;
  GenericSwitchRequirements2.server = { mandatory: { Identify: GenericSwitchRequirements2.IdentifyServer, Switch: GenericSwitchRequirements2.SwitchServer } };
})(GenericSwitchRequirements || (GenericSwitchRequirements = {}));
const GenericSwitchDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "GenericSwitch",
  deviceType: 15,
  deviceRevision: 3,
  requirements: GenericSwitchRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(GenericSwitchRequirements.server.mandatory.Identify)
});
const GenericSwitchDevice = GenericSwitchDeviceDefinition;
