// @matter/node/dist/cjs/devices/heat-pump.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var heat_pump_exports = {};
__export(heat_pump_exports, {
  HeatPumpDevice: () => HeatPumpDevice,
  HeatPumpDeviceDefinition: () => HeatPumpDeviceDefinition,
  HeatPumpRequirements: () => HeatPumpRequirements
});
module.exports = __toCommonJS(heat_pump_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_ThermostatBehavior = require("../behaviors/thermostat/ThermostatBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var HeatPumpRequirements;
((HeatPumpRequirements2) => {
  HeatPumpRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  HeatPumpRequirements2.ThermostatBehavior = import_ThermostatBehavior.ThermostatBehavior;
  HeatPumpRequirements2.server = { optional: { Identify: HeatPumpRequirements2.IdentifyServer }, mandatory: {} };
  HeatPumpRequirements2.client = { optional: { Thermostat: HeatPumpRequirements2.ThermostatBehavior }, mandatory: {} };
})(HeatPumpRequirements || (HeatPumpRequirements = {}));
const HeatPumpDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "HeatPump",
  deviceType: 777,
  deviceRevision: 1,
  requirements: HeatPumpRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const HeatPumpDevice = HeatPumpDeviceDefinition;
