// @matter/node/dist/cjs/devices/humidity-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var humidity_sensor_exports = {};
__export(humidity_sensor_exports, {
  HumiditySensorDevice: () => HumiditySensorDevice,
  HumiditySensorDeviceDefinition: () => HumiditySensorDeviceDefinition,
  HumiditySensorRequirements: () => HumiditySensorRequirements
});
module.exports = __toCommonJS(humidity_sensor_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_RelativeHumidityMeasurementServer = require("../behaviors/relative-humidity-measurement/RelativeHumidityMeasurementServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var HumiditySensorRequirements;
((HumiditySensorRequirements2) => {
  HumiditySensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  HumiditySensorRequirements2.RelativeHumidityMeasurementServer = import_RelativeHumidityMeasurementServer.RelativeHumidityMeasurementServer;
  HumiditySensorRequirements2.server = {
    mandatory: { Identify: HumiditySensorRequirements2.IdentifyServer, RelativeHumidityMeasurement: HumiditySensorRequirements2.RelativeHumidityMeasurementServer }
  };
})(HumiditySensorRequirements || (HumiditySensorRequirements = {}));
const HumiditySensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "HumiditySensor",
  deviceType: 775,
  deviceRevision: 2,
  requirements: HumiditySensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    HumiditySensorRequirements.server.mandatory.Identify,
    HumiditySensorRequirements.server.mandatory.RelativeHumidityMeasurement
  )
});
const HumiditySensorDevice = HumiditySensorDeviceDefinition;
