// @matter/node/dist/cjs/devices/laundry-washer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var laundry_washer_exports = {};
__export(laundry_washer_exports, {
  LaundryWasherDevice: () => LaundryWasherDevice,
  LaundryWasherDeviceDefinition: () => LaundryWasherDeviceDefinition,
  LaundryWasherRequirements: () => LaundryWasherRequirements
});
module.exports = __toCommonJS(laundry_washer_exports);
var import_OperationalStateServer = require("../behaviors/operational-state/OperationalStateServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_LaundryWasherModeServer = require("../behaviors/laundry-washer-mode/LaundryWasherModeServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_LaundryWasherControlsServer = require("../behaviors/laundry-washer-controls/LaundryWasherControlsServer.js");
var import_TemperatureControlServer = require("../behaviors/temperature-control/TemperatureControlServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var LaundryWasherRequirements;
((LaundryWasherRequirements2) => {
  LaundryWasherRequirements2.OperationalStateServer = import_OperationalStateServer.OperationalStateServer;
  LaundryWasherRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  LaundryWasherRequirements2.LaundryWasherModeServer = import_LaundryWasherModeServer.LaundryWasherModeServer;
  LaundryWasherRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("DeadFrontBehavior");
  LaundryWasherRequirements2.LaundryWasherControlsServer = import_LaundryWasherControlsServer.LaundryWasherControlsServer;
  LaundryWasherRequirements2.TemperatureControlServer = import_TemperatureControlServer.TemperatureControlServer;
  LaundryWasherRequirements2.server = {
    mandatory: { OperationalState: LaundryWasherRequirements2.OperationalStateServer },
    optional: {
      Identify: LaundryWasherRequirements2.IdentifyServer,
      LaundryWasherMode: LaundryWasherRequirements2.LaundryWasherModeServer,
      OnOff: LaundryWasherRequirements2.OnOffServer,
      LaundryWasherControls: LaundryWasherRequirements2.LaundryWasherControlsServer,
      TemperatureControl: LaundryWasherRequirements2.TemperatureControlServer
    }
  };
})(LaundryWasherRequirements || (LaundryWasherRequirements = {}));
const LaundryWasherDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "LaundryWasher",
  deviceType: 115,
  deviceRevision: 1,
  requirements: LaundryWasherRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(LaundryWasherRequirements.server.mandatory.OperationalState)
});
const LaundryWasherDevice = LaundryWasherDeviceDefinition;
