// @matter/node/dist/cjs/devices/laundry-dryer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var laundry_dryer_exports = {};
__export(laundry_dryer_exports, {
  LaundryDryerDevice: () => LaundryDryerDevice,
  LaundryDryerDeviceDefinition: () => LaundryDryerDeviceDefinition,
  LaundryDryerRequirements: () => LaundryDryerRequirements
});
module.exports = __toCommonJS(laundry_dryer_exports);
var import_OperationalStateServer = require("../behaviors/operational-state/OperationalStateServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_LaundryWasherModeServer = require("../behaviors/laundry-washer-mode/LaundryWasherModeServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_LaundryDryerControlsServer = require("../behaviors/laundry-dryer-controls/LaundryDryerControlsServer.js");
var import_TemperatureControlServer = require("../behaviors/temperature-control/TemperatureControlServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var LaundryDryerRequirements;
((LaundryDryerRequirements2) => {
  LaundryDryerRequirements2.OperationalStateServer = import_OperationalStateServer.OperationalStateServer;
  LaundryDryerRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  LaundryDryerRequirements2.LaundryWasherModeServer = import_LaundryWasherModeServer.LaundryWasherModeServer;
  LaundryDryerRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("DeadFrontBehavior");
  LaundryDryerRequirements2.LaundryDryerControlsServer = import_LaundryDryerControlsServer.LaundryDryerControlsServer;
  LaundryDryerRequirements2.TemperatureControlServer = import_TemperatureControlServer.TemperatureControlServer;
  LaundryDryerRequirements2.server = {
    mandatory: { OperationalState: LaundryDryerRequirements2.OperationalStateServer },
    optional: {
      Identify: LaundryDryerRequirements2.IdentifyServer,
      LaundryWasherMode: LaundryDryerRequirements2.LaundryWasherModeServer,
      OnOff: LaundryDryerRequirements2.OnOffServer,
      LaundryDryerControls: LaundryDryerRequirements2.LaundryDryerControlsServer,
      TemperatureControl: LaundryDryerRequirements2.TemperatureControlServer
    }
  };
})(LaundryDryerRequirements || (LaundryDryerRequirements = {}));
const LaundryDryerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "LaundryDryer",
  deviceType: 124,
  deviceRevision: 1,
  requirements: LaundryDryerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(LaundryDryerRequirements.server.mandatory.OperationalState)
});
const LaundryDryerDevice = LaundryDryerDeviceDefinition;
