// @matter/node/dist/cjs/devices/microwave-oven.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var microwave_oven_exports = {};
__export(microwave_oven_exports, {
  MicrowaveOvenDevice: () => MicrowaveOvenDevice,
  MicrowaveOvenDeviceDefinition: () => MicrowaveOvenDeviceDefinition,
  MicrowaveOvenRequirements: () => MicrowaveOvenRequirements
});
module.exports = __toCommonJS(microwave_oven_exports);
var import_OperationalStateServer = require("../behaviors/operational-state/OperationalStateServer.js");
var import_MicrowaveOvenModeServer = require("../behaviors/microwave-oven-mode/MicrowaveOvenModeServer.js");
var import_MicrowaveOvenControlServer = require("../behaviors/microwave-oven-control/MicrowaveOvenControlServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_FanControlServer = require("../behaviors/fan-control/FanControlServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var MicrowaveOvenRequirements;
((MicrowaveOvenRequirements2) => {
  MicrowaveOvenRequirements2.OperationalStateServer = import_OperationalStateServer.OperationalStateServer.alter({ attributes: { countdownTime: { optional: false } } });
  MicrowaveOvenRequirements2.MicrowaveOvenModeServer = import_MicrowaveOvenModeServer.MicrowaveOvenModeServer;
  MicrowaveOvenRequirements2.MicrowaveOvenControlServer = import_MicrowaveOvenControlServer.MicrowaveOvenControlServer;
  MicrowaveOvenRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  MicrowaveOvenRequirements2.FanControlServer = import_FanControlServer.FanControlServer;
  MicrowaveOvenRequirements2.server = {
    mandatory: {
      OperationalState: MicrowaveOvenRequirements2.OperationalStateServer,
      MicrowaveOvenMode: MicrowaveOvenRequirements2.MicrowaveOvenModeServer,
      MicrowaveOvenControl: MicrowaveOvenRequirements2.MicrowaveOvenControlServer
    },
    optional: { Identify: MicrowaveOvenRequirements2.IdentifyServer, FanControl: MicrowaveOvenRequirements2.FanControlServer }
  };
})(MicrowaveOvenRequirements || (MicrowaveOvenRequirements = {}));
const MicrowaveOvenDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "MicrowaveOven",
  deviceType: 121,
  deviceRevision: 1,
  requirements: MicrowaveOvenRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    MicrowaveOvenRequirements.server.mandatory.OperationalState,
    MicrowaveOvenRequirements.server.mandatory.MicrowaveOvenMode
  )
});
const MicrowaveOvenDevice = MicrowaveOvenDeviceDefinition;
