// @matter/node/dist/cjs/devices/mode-select.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mode_select_exports = {};
__export(mode_select_exports, {
  ModeSelectDevice: () => ModeSelectDevice,
  ModeSelectDeviceDefinition: () => ModeSelectDeviceDefinition,
  ModeSelectRequirements: () => ModeSelectRequirements
});
module.exports = __toCommonJS(mode_select_exports);
var import_ModeSelectServer = require("../behaviors/mode-select/ModeSelectServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var ModeSelectRequirements;
((ModeSelectRequirements2) => {
  ModeSelectRequirements2.ModeSelectServer = import_ModeSelectServer.ModeSelectServer;
  ModeSelectRequirements2.server = { mandatory: { ModeSelect: ModeSelectRequirements2.ModeSelectServer } };
})(ModeSelectRequirements || (ModeSelectRequirements = {}));
const ModeSelectDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "ModeSelect",
  deviceType: 39,
  deviceRevision: 1,
  requirements: ModeSelectRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(ModeSelectRequirements.server.mandatory.ModeSelect)
});
const ModeSelectDevice = ModeSelectDeviceDefinition;
