// @matter/node/dist/cjs/devices/mounted-on-off-control.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var mounted_on_off_control_exports = {};
__export(mounted_on_off_control_exports, {
  MountedOnOffControlDevice: () => MountedOnOffControlDevice,
  MountedOnOffControlDeviceDefinition: () => MountedOnOffControlDeviceDefinition,
  MountedOnOffControlRequirements: () => MountedOnOffControlRequirements
});
module.exports = __toCommonJS(mounted_on_off_control_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_ScenesManagementServer = require("../behaviors/scenes-management/ScenesManagementServer.js");
var import_LevelControlServer = require("../behaviors/level-control/LevelControlServer.js");
var import_OccupancySensingBehavior = require("../behaviors/occupancy-sensing/OccupancySensingBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var MountedOnOffControlRequirements;
((MountedOnOffControlRequirements2) => {
  MountedOnOffControlRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer.alter({ commands: { triggerEffect: { optional: false } } });
  MountedOnOffControlRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  MountedOnOffControlRequirements2.OnOffServer = import_OnOffServer.OnOffServer.with("Lighting");
  MountedOnOffControlRequirements2.ScenesManagementServer = import_ScenesManagementServer.ScenesManagementServer;
  MountedOnOffControlRequirements2.LevelControlServer = import_LevelControlServer.LevelControlServer.with("OnOff", "Lighting").alter({
    attributes: {
      currentLevel: { min: 1, max: 254 },
      minLevel: { default: 1, min: 1, max: 2 },
      maxLevel: { default: 254, min: 254, max: 255 }
    }
  });
  MountedOnOffControlRequirements2.OccupancySensingBehavior = import_OccupancySensingBehavior.OccupancySensingBehavior;
  MountedOnOffControlRequirements2.server = {
    mandatory: { Identify: MountedOnOffControlRequirements2.IdentifyServer, Groups: MountedOnOffControlRequirements2.GroupsServer, OnOff: MountedOnOffControlRequirements2.OnOffServer },
    optional: { ScenesManagement: MountedOnOffControlRequirements2.ScenesManagementServer, LevelControl: MountedOnOffControlRequirements2.LevelControlServer }
  };
  MountedOnOffControlRequirements2.client = { optional: { OccupancySensing: MountedOnOffControlRequirements2.OccupancySensingBehavior }, mandatory: {} };
})(MountedOnOffControlRequirements || (MountedOnOffControlRequirements = {}));
const MountedOnOffControlDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "MountedOnOffControl",
  deviceType: 271,
  deviceRevision: 1,
  requirements: MountedOnOffControlRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    MountedOnOffControlRequirements.server.mandatory.Identify,
    MountedOnOffControlRequirements.server.mandatory.Groups,
    MountedOnOffControlRequirements.server.mandatory.OnOff
  )
});
const MountedOnOffControlDevice = MountedOnOffControlDeviceDefinition;
