// @matter/node/dist/cjs/devices/occupancy-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var occupancy_sensor_exports = {};
__export(occupancy_sensor_exports, {
  OccupancySensorDevice: () => OccupancySensorDevice,
  OccupancySensorDeviceDefinition: () => OccupancySensorDeviceDefinition,
  OccupancySensorRequirements: () => OccupancySensorRequirements
});
module.exports = __toCommonJS(occupancy_sensor_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_OccupancySensingServer = require("../behaviors/occupancy-sensing/OccupancySensingServer.js");
var import_BooleanStateConfigurationServer = require("../behaviors/boolean-state-configuration/BooleanStateConfigurationServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var OccupancySensorRequirements;
((OccupancySensorRequirements2) => {
  OccupancySensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  OccupancySensorRequirements2.OccupancySensingServer = import_OccupancySensingServer.OccupancySensingServer;
  OccupancySensorRequirements2.BooleanStateConfigurationServer = import_BooleanStateConfigurationServer.BooleanStateConfigurationServer;
  OccupancySensorRequirements2.server = {
    mandatory: { Identify: OccupancySensorRequirements2.IdentifyServer, OccupancySensing: OccupancySensorRequirements2.OccupancySensingServer },
    optional: { BooleanStateConfiguration: OccupancySensorRequirements2.BooleanStateConfigurationServer }
  };
})(OccupancySensorRequirements || (OccupancySensorRequirements = {}));
const OccupancySensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "OccupancySensor",
  deviceType: 263,
  deviceRevision: 4,
  requirements: OccupancySensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(OccupancySensorRequirements.server.mandatory.Identify)
});
const OccupancySensorDevice = OccupancySensorDeviceDefinition;
