// @matter/node/dist/cjs/devices/pressure-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pressure_sensor_exports = {};
__export(pressure_sensor_exports, {
  PressureSensorDevice: () => PressureSensorDevice,
  PressureSensorDeviceDefinition: () => PressureSensorDeviceDefinition,
  PressureSensorRequirements: () => PressureSensorRequirements
});
module.exports = __toCommonJS(pressure_sensor_exports);
var import_PressureMeasurementServer = require("../behaviors/pressure-measurement/PressureMeasurementServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var PressureSensorRequirements;
((PressureSensorRequirements2) => {
  PressureSensorRequirements2.PressureMeasurementServer = import_PressureMeasurementServer.PressureMeasurementServer;
  PressureSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  PressureSensorRequirements2.server = { mandatory: { PressureMeasurement: PressureSensorRequirements2.PressureMeasurementServer, Identify: PressureSensorRequirements2.IdentifyServer } };
})(PressureSensorRequirements || (PressureSensorRequirements = {}));
const PressureSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "PressureSensor",
  deviceType: 773,
  deviceRevision: 2,
  requirements: PressureSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    PressureSensorRequirements.server.mandatory.PressureMeasurement,
    PressureSensorRequirements.server.mandatory.Identify
  )
});
const PressureSensorDevice = PressureSensorDeviceDefinition;
