// @matter/node/dist/cjs/devices/pump-controller.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pump_controller_exports = {};
__export(pump_controller_exports, {
  PumpControllerDevice: () => PumpControllerDevice,
  PumpControllerDeviceDefinition: () => PumpControllerDeviceDefinition,
  PumpControllerRequirements: () => PumpControllerRequirements
});
module.exports = __toCommonJS(pump_controller_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_OnOffBehavior = require("../behaviors/on-off/OnOffBehavior.js");
var import_PumpConfigurationAndControlBehavior = require("../behaviors/pump-configuration-and-control/PumpConfigurationAndControlBehavior.js");
var import_IdentifyBehavior = require("../behaviors/identify/IdentifyBehavior.js");
var import_GroupsBehavior = require("../behaviors/groups/GroupsBehavior.js");
var import_LevelControlBehavior = require("../behaviors/level-control/LevelControlBehavior.js");
var import_ScenesManagementBehavior = require("../behaviors/scenes-management/ScenesManagementBehavior.js");
var import_TemperatureMeasurementBehavior = require("../behaviors/temperature-measurement/TemperatureMeasurementBehavior.js");
var import_PressureMeasurementBehavior = require("../behaviors/pressure-measurement/PressureMeasurementBehavior.js");
var import_FlowMeasurementBehavior = require("../behaviors/flow-measurement/FlowMeasurementBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var PumpControllerRequirements;
((PumpControllerRequirements2) => {
  PumpControllerRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  PumpControllerRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  PumpControllerRequirements2.PumpConfigurationAndControlBehavior = import_PumpConfigurationAndControlBehavior.PumpConfigurationAndControlBehavior;
  PumpControllerRequirements2.IdentifyBehavior = import_IdentifyBehavior.IdentifyBehavior;
  PumpControllerRequirements2.GroupsBehavior = import_GroupsBehavior.GroupsBehavior;
  PumpControllerRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  PumpControllerRequirements2.ScenesManagementBehavior = import_ScenesManagementBehavior.ScenesManagementBehavior;
  PumpControllerRequirements2.TemperatureMeasurementBehavior = import_TemperatureMeasurementBehavior.TemperatureMeasurementBehavior;
  PumpControllerRequirements2.PressureMeasurementBehavior = import_PressureMeasurementBehavior.PressureMeasurementBehavior;
  PumpControllerRequirements2.FlowMeasurementBehavior = import_FlowMeasurementBehavior.FlowMeasurementBehavior;
  PumpControllerRequirements2.server = { mandatory: { Identify: PumpControllerRequirements2.IdentifyServer } };
  PumpControllerRequirements2.client = {
    mandatory: { OnOff: PumpControllerRequirements2.OnOffBehavior, PumpConfigurationAndControl: PumpControllerRequirements2.PumpConfigurationAndControlBehavior },
    optional: {
      Identify: PumpControllerRequirements2.IdentifyBehavior,
      Groups: PumpControllerRequirements2.GroupsBehavior,
      LevelControl: PumpControllerRequirements2.LevelControlBehavior,
      ScenesManagement: PumpControllerRequirements2.ScenesManagementBehavior,
      TemperatureMeasurement: PumpControllerRequirements2.TemperatureMeasurementBehavior,
      PressureMeasurement: PumpControllerRequirements2.PressureMeasurementBehavior,
      FlowMeasurement: PumpControllerRequirements2.FlowMeasurementBehavior
    }
  };
})(PumpControllerRequirements || (PumpControllerRequirements = {}));
const PumpControllerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "PumpController",
  deviceType: 772,
  deviceRevision: 4,
  requirements: PumpControllerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(PumpControllerRequirements.server.mandatory.Identify)
});
const PumpControllerDevice = PumpControllerDeviceDefinition;
