// @matter/node/dist/cjs/devices/rain-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rain_sensor_exports = {};
__export(rain_sensor_exports, {
  RainSensorDevice: () => RainSensorDevice,
  RainSensorDeviceDefinition: () => RainSensorDeviceDefinition,
  RainSensorRequirements: () => RainSensorRequirements
});
module.exports = __toCommonJS(rain_sensor_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_BooleanStateServer = require("../behaviors/boolean-state/BooleanStateServer.js");
var import_BooleanStateConfigurationServer = require("../behaviors/boolean-state-configuration/BooleanStateConfigurationServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var RainSensorRequirements;
((RainSensorRequirements2) => {
  RainSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  RainSensorRequirements2.BooleanStateServer = import_BooleanStateServer.BooleanStateServer.alter({ events: { stateChange: { optional: false } } });
  RainSensorRequirements2.BooleanStateConfigurationServer = import_BooleanStateConfigurationServer.BooleanStateConfigurationServer;
  RainSensorRequirements2.server = {
    mandatory: { Identify: RainSensorRequirements2.IdentifyServer, BooleanState: RainSensorRequirements2.BooleanStateServer },
    optional: { BooleanStateConfiguration: RainSensorRequirements2.BooleanStateConfigurationServer }
  };
})(RainSensorRequirements || (RainSensorRequirements = {}));
const RainSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "RainSensor",
  deviceType: 68,
  deviceRevision: 1,
  requirements: RainSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    RainSensorRequirements.server.mandatory.Identify,
    RainSensorRequirements.server.mandatory.BooleanState
  )
});
const RainSensorDevice = RainSensorDeviceDefinition;
