// @matter/node/dist/cjs/devices/refrigerator.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var refrigerator_exports = {};
__export(refrigerator_exports, {
  RefrigeratorDevice: () => RefrigeratorDevice,
  RefrigeratorDeviceDefinition: () => RefrigeratorDeviceDefinition,
  RefrigeratorRequirements: () => RefrigeratorRequirements
});
module.exports = __toCommonJS(refrigerator_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_RefrigeratorAndTemperatureControlledCabinetModeServer = require("../behaviors/refrigerator-and-temperature-controlled-cabinet-mode/RefrigeratorAndTemperatureControlledCabinetModeServer.js");
var import_RefrigeratorAlarmServer = require("../behaviors/refrigerator-alarm/RefrigeratorAlarmServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var RefrigeratorRequirements;
((RefrigeratorRequirements2) => {
  RefrigeratorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  RefrigeratorRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer = import_RefrigeratorAndTemperatureControlledCabinetModeServer.RefrigeratorAndTemperatureControlledCabinetModeServer;
  RefrigeratorRequirements2.RefrigeratorAlarmServer = import_RefrigeratorAlarmServer.RefrigeratorAlarmServer;
  RefrigeratorRequirements2.server = {
    optional: {
      Identify: RefrigeratorRequirements2.IdentifyServer,
      RefrigeratorAndTemperatureControlledCabinetMode: RefrigeratorRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer,
      RefrigeratorAlarm: RefrigeratorRequirements2.RefrigeratorAlarmServer
    },
    mandatory: {}
  };
})(RefrigeratorRequirements || (RefrigeratorRequirements = {}));
const RefrigeratorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "Refrigerator",
  deviceType: 112,
  deviceRevision: 2,
  requirements: RefrigeratorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const RefrigeratorDevice = RefrigeratorDeviceDefinition;
