// @matter/node/dist/cjs/devices/solar-power.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var solar_power_exports = {};
__export(solar_power_exports, {
  SolarPowerDevice: () => SolarPowerDevice,
  SolarPowerDeviceDefinition: () => SolarPowerDeviceDefinition,
  SolarPowerRequirements: () => SolarPowerRequirements
});
module.exports = __toCommonJS(solar_power_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var SolarPowerRequirements;
((SolarPowerRequirements2) => {
  SolarPowerRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  SolarPowerRequirements2.server = { optional: { Identify: SolarPowerRequirements2.IdentifyServer }, mandatory: {} };
})(SolarPowerRequirements || (SolarPowerRequirements = {}));
const SolarPowerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "SolarPower",
  deviceType: 23,
  deviceRevision: 1,
  requirements: SolarPowerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const SolarPowerDevice = SolarPowerDeviceDefinition;
