// @matter/node/dist/cjs/devices/speaker.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var speaker_exports = {};
__export(speaker_exports, {
  SpeakerDevice: () => SpeakerDevice,
  SpeakerDeviceDefinition: () => SpeakerDeviceDefinition,
  SpeakerRequirements: () => SpeakerRequirements
});
module.exports = __toCommonJS(speaker_exports);
var import_OnOffServer = require("../behaviors/on-off/OnOffServer.js");
var import_LevelControlServer = require("../behaviors/level-control/LevelControlServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var SpeakerRequirements;
((SpeakerRequirements2) => {
  SpeakerRequirements2.OnOffServer = import_OnOffServer.OnOffServer;
  SpeakerRequirements2.LevelControlServer = import_LevelControlServer.LevelControlServer;
  SpeakerRequirements2.server = { mandatory: { OnOff: SpeakerRequirements2.OnOffServer, LevelControl: SpeakerRequirements2.LevelControlServer } };
})(SpeakerRequirements || (SpeakerRequirements = {}));
const SpeakerDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "Speaker",
  deviceType: 34,
  deviceRevision: 1,
  requirements: SpeakerRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    SpeakerRequirements.server.mandatory.OnOff,
    SpeakerRequirements.server.mandatory.LevelControl
  )
});
const SpeakerDevice = SpeakerDeviceDefinition;
