// @matter/node/dist/cjs/devices/temperature-sensor.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var temperature_sensor_exports = {};
__export(temperature_sensor_exports, {
  TemperatureSensorDevice: () => TemperatureSensorDevice,
  TemperatureSensorDeviceDefinition: () => TemperatureSensorDeviceDefinition,
  TemperatureSensorRequirements: () => TemperatureSensorRequirements
});
module.exports = __toCommonJS(temperature_sensor_exports);
var import_TemperatureMeasurementServer = require("../behaviors/temperature-measurement/TemperatureMeasurementServer.js");
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var TemperatureSensorRequirements;
((TemperatureSensorRequirements2) => {
  TemperatureSensorRequirements2.TemperatureMeasurementServer = import_TemperatureMeasurementServer.TemperatureMeasurementServer;
  TemperatureSensorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  TemperatureSensorRequirements2.server = { mandatory: { TemperatureMeasurement: TemperatureSensorRequirements2.TemperatureMeasurementServer, Identify: TemperatureSensorRequirements2.IdentifyServer } };
})(TemperatureSensorRequirements || (TemperatureSensorRequirements = {}));
const TemperatureSensorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "TemperatureSensor",
  deviceType: 770,
  deviceRevision: 2,
  requirements: TemperatureSensorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    TemperatureSensorRequirements.server.mandatory.TemperatureMeasurement,
    TemperatureSensorRequirements.server.mandatory.Identify
  )
});
const TemperatureSensorDevice = TemperatureSensorDeviceDefinition;
