// @matter/node/dist/cjs/devices/temperature-controlled-cabinet.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var temperature_controlled_cabinet_exports = {};
__export(temperature_controlled_cabinet_exports, {
  TemperatureControlledCabinetDevice: () => TemperatureControlledCabinetDevice,
  TemperatureControlledCabinetDeviceDefinition: () => TemperatureControlledCabinetDeviceDefinition,
  TemperatureControlledCabinetRequirements: () => TemperatureControlledCabinetRequirements
});
module.exports = __toCommonJS(temperature_controlled_cabinet_exports);
var import_TemperatureControlServer = require("../behaviors/temperature-control/TemperatureControlServer.js");
var import_TemperatureMeasurementServer = require("../behaviors/temperature-measurement/TemperatureMeasurementServer.js");
var import_RefrigeratorAndTemperatureControlledCabinetModeServer = require("../behaviors/refrigerator-and-temperature-controlled-cabinet-mode/RefrigeratorAndTemperatureControlledCabinetModeServer.js");
var import_OvenModeServer = require("../behaviors/oven-mode/OvenModeServer.js");
var import_OvenCavityOperationalStateServer = require("../behaviors/oven-cavity-operational-state/OvenCavityOperationalStateServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var TemperatureControlledCabinetRequirements;
((TemperatureControlledCabinetRequirements2) => {
  TemperatureControlledCabinetRequirements2.TemperatureControlServer = import_TemperatureControlServer.TemperatureControlServer;
  TemperatureControlledCabinetRequirements2.TemperatureMeasurementServer = import_TemperatureMeasurementServer.TemperatureMeasurementServer;
  TemperatureControlledCabinetRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer = import_RefrigeratorAndTemperatureControlledCabinetModeServer.RefrigeratorAndTemperatureControlledCabinetModeServer;
  TemperatureControlledCabinetRequirements2.OvenModeServer = import_OvenModeServer.OvenModeServer;
  TemperatureControlledCabinetRequirements2.OvenCavityOperationalStateServer = import_OvenCavityOperationalStateServer.OvenCavityOperationalStateServer;
  TemperatureControlledCabinetRequirements2.server = {
    mandatory: { TemperatureControl: TemperatureControlledCabinetRequirements2.TemperatureControlServer },
    optional: {
      TemperatureMeasurement: TemperatureControlledCabinetRequirements2.TemperatureMeasurementServer,
      RefrigeratorAndTemperatureControlledCabinetMode: TemperatureControlledCabinetRequirements2.RefrigeratorAndTemperatureControlledCabinetModeServer,
      OvenMode: TemperatureControlledCabinetRequirements2.OvenModeServer,
      OvenCavityOperationalState: TemperatureControlledCabinetRequirements2.OvenCavityOperationalStateServer
    }
  };
})(TemperatureControlledCabinetRequirements || (TemperatureControlledCabinetRequirements = {}));
const TemperatureControlledCabinetDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "TemperatureControlledCabinet",
  deviceType: 113,
  deviceRevision: 3,
  requirements: TemperatureControlledCabinetRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const TemperatureControlledCabinetDevice = TemperatureControlledCabinetDeviceDefinition;
