// @matter/node/dist/cjs/devices/video-remote-control.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var video_remote_control_exports = {};
__export(video_remote_control_exports, {
  VideoRemoteControlDevice: () => VideoRemoteControlDevice,
  VideoRemoteControlDeviceDefinition: () => VideoRemoteControlDeviceDefinition,
  VideoRemoteControlRequirements: () => VideoRemoteControlRequirements
});
module.exports = __toCommonJS(video_remote_control_exports);
var import_OnOffBehavior = require("../behaviors/on-off/OnOffBehavior.js");
var import_MediaPlaybackBehavior = require("../behaviors/media-playback/MediaPlaybackBehavior.js");
var import_KeypadInputBehavior = require("../behaviors/keypad-input/KeypadInputBehavior.js");
var import_LevelControlBehavior = require("../behaviors/level-control/LevelControlBehavior.js");
var import_WakeOnLanBehavior = require("../behaviors/wake-on-lan/WakeOnLanBehavior.js");
var import_ChannelBehavior = require("../behaviors/channel/ChannelBehavior.js");
var import_TargetNavigatorBehavior = require("../behaviors/target-navigator/TargetNavigatorBehavior.js");
var import_MediaInputBehavior = require("../behaviors/media-input/MediaInputBehavior.js");
var import_LowPowerBehavior = require("../behaviors/low-power/LowPowerBehavior.js");
var import_ContentLauncherBehavior = require("../behaviors/content-launcher/ContentLauncherBehavior.js");
var import_AudioOutputBehavior = require("../behaviors/audio-output/AudioOutputBehavior.js");
var import_ApplicationLauncherBehavior = require("../behaviors/application-launcher/ApplicationLauncherBehavior.js");
var import_AccountLoginBehavior = require("../behaviors/account-login/AccountLoginBehavior.js");
var import_ContentControlBehavior = require("../behaviors/content-control/ContentControlBehavior.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var VideoRemoteControlRequirements;
((VideoRemoteControlRequirements2) => {
  VideoRemoteControlRequirements2.OnOffBehavior = import_OnOffBehavior.OnOffBehavior;
  VideoRemoteControlRequirements2.MediaPlaybackBehavior = import_MediaPlaybackBehavior.MediaPlaybackBehavior;
  VideoRemoteControlRequirements2.KeypadInputBehavior = import_KeypadInputBehavior.KeypadInputBehavior;
  VideoRemoteControlRequirements2.LevelControlBehavior = import_LevelControlBehavior.LevelControlBehavior;
  VideoRemoteControlRequirements2.WakeOnLanBehavior = import_WakeOnLanBehavior.WakeOnLanBehavior;
  VideoRemoteControlRequirements2.ChannelBehavior = import_ChannelBehavior.ChannelBehavior;
  VideoRemoteControlRequirements2.TargetNavigatorBehavior = import_TargetNavigatorBehavior.TargetNavigatorBehavior;
  VideoRemoteControlRequirements2.MediaInputBehavior = import_MediaInputBehavior.MediaInputBehavior;
  VideoRemoteControlRequirements2.LowPowerBehavior = import_LowPowerBehavior.LowPowerBehavior;
  VideoRemoteControlRequirements2.ContentLauncherBehavior = import_ContentLauncherBehavior.ContentLauncherBehavior;
  VideoRemoteControlRequirements2.AudioOutputBehavior = import_AudioOutputBehavior.AudioOutputBehavior;
  VideoRemoteControlRequirements2.ApplicationLauncherBehavior = import_ApplicationLauncherBehavior.ApplicationLauncherBehavior;
  VideoRemoteControlRequirements2.AccountLoginBehavior = import_AccountLoginBehavior.AccountLoginBehavior;
  VideoRemoteControlRequirements2.ContentControlBehavior = import_ContentControlBehavior.ContentControlBehavior;
  VideoRemoteControlRequirements2.client = {
    mandatory: { OnOff: VideoRemoteControlRequirements2.OnOffBehavior, MediaPlayback: VideoRemoteControlRequirements2.MediaPlaybackBehavior, KeypadInput: VideoRemoteControlRequirements2.KeypadInputBehavior },
    optional: {
      LevelControl: VideoRemoteControlRequirements2.LevelControlBehavior,
      WakeOnLan: VideoRemoteControlRequirements2.WakeOnLanBehavior,
      Channel: VideoRemoteControlRequirements2.ChannelBehavior,
      TargetNavigator: VideoRemoteControlRequirements2.TargetNavigatorBehavior,
      MediaInput: VideoRemoteControlRequirements2.MediaInputBehavior,
      LowPower: VideoRemoteControlRequirements2.LowPowerBehavior,
      ContentLauncher: VideoRemoteControlRequirements2.ContentLauncherBehavior,
      AudioOutput: VideoRemoteControlRequirements2.AudioOutputBehavior,
      ApplicationLauncher: VideoRemoteControlRequirements2.ApplicationLauncherBehavior,
      AccountLogin: VideoRemoteControlRequirements2.AccountLoginBehavior,
      ContentControl: VideoRemoteControlRequirements2.ContentControlBehavior
    }
  };
})(VideoRemoteControlRequirements || (VideoRemoteControlRequirements = {}));
const VideoRemoteControlDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "VideoRemoteControl",
  deviceType: 42,
  deviceRevision: 2,
  requirements: VideoRemoteControlRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)()
});
const VideoRemoteControlDevice = VideoRemoteControlDeviceDefinition;
