// @matter/node/dist/cjs/devices/water-freeze-detector.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var water_freeze_detector_exports = {};
__export(water_freeze_detector_exports, {
  WaterFreezeDetectorDevice: () => WaterFreezeDetectorDevice,
  WaterFreezeDetectorDeviceDefinition: () => WaterFreezeDetectorDeviceDefinition,
  WaterFreezeDetectorRequirements: () => WaterFreezeDetectorRequirements
});
module.exports = __toCommonJS(water_freeze_detector_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_BooleanStateServer = require("../behaviors/boolean-state/BooleanStateServer.js");
var import_BooleanStateConfigurationServer = require("../behaviors/boolean-state-configuration/BooleanStateConfigurationServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var WaterFreezeDetectorRequirements;
((WaterFreezeDetectorRequirements2) => {
  WaterFreezeDetectorRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  WaterFreezeDetectorRequirements2.BooleanStateServer = import_BooleanStateServer.BooleanStateServer.alter({ events: { stateChange: { optional: false } } });
  WaterFreezeDetectorRequirements2.BooleanStateConfigurationServer = import_BooleanStateConfigurationServer.BooleanStateConfigurationServer;
  WaterFreezeDetectorRequirements2.server = {
    mandatory: { Identify: WaterFreezeDetectorRequirements2.IdentifyServer, BooleanState: WaterFreezeDetectorRequirements2.BooleanStateServer },
    optional: { BooleanStateConfiguration: WaterFreezeDetectorRequirements2.BooleanStateConfigurationServer }
  };
})(WaterFreezeDetectorRequirements || (WaterFreezeDetectorRequirements = {}));
const WaterFreezeDetectorDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "WaterFreezeDetector",
  deviceType: 65,
  deviceRevision: 1,
  requirements: WaterFreezeDetectorRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(
    WaterFreezeDetectorRequirements.server.mandatory.Identify,
    WaterFreezeDetectorRequirements.server.mandatory.BooleanState
  )
});
const WaterFreezeDetectorDevice = WaterFreezeDetectorDeviceDefinition;
