// @matter/node/dist/cjs/devices/window-covering.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var window_covering_exports = {};
__export(window_covering_exports, {
  WindowCoveringDevice: () => WindowCoveringDevice,
  WindowCoveringDeviceDefinition: () => WindowCoveringDeviceDefinition,
  WindowCoveringRequirements: () => WindowCoveringRequirements
});
module.exports = __toCommonJS(window_covering_exports);
var import_IdentifyServer = require("../behaviors/identify/IdentifyServer.js");
var import_WindowCoveringServer = require("../behaviors/window-covering/WindowCoveringServer.js");
var import_GroupsServer = require("../behaviors/groups/GroupsServer.js");
var import_MutableEndpoint = require("../endpoint/type/MutableEndpoint.js");
var import_SupportedBehaviors = require("../endpoint/properties/SupportedBehaviors.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var WindowCoveringRequirements;
((WindowCoveringRequirements2) => {
  WindowCoveringRequirements2.IdentifyServer = import_IdentifyServer.IdentifyServer;
  WindowCoveringRequirements2.WindowCoveringServer = import_WindowCoveringServer.WindowCoveringServer;
  WindowCoveringRequirements2.GroupsServer = import_GroupsServer.GroupsServer;
  WindowCoveringRequirements2.server = {
    mandatory: { Identify: WindowCoveringRequirements2.IdentifyServer, WindowCovering: WindowCoveringRequirements2.WindowCoveringServer },
    optional: { Groups: WindowCoveringRequirements2.GroupsServer }
  };
})(WindowCoveringRequirements || (WindowCoveringRequirements = {}));
const WindowCoveringDeviceDefinition = (0, import_MutableEndpoint.MutableEndpoint)({
  name: "WindowCovering",
  deviceType: 514,
  deviceRevision: 3,
  requirements: WindowCoveringRequirements,
  behaviors: (0, import_SupportedBehaviors.SupportedBehaviors)(WindowCoveringRequirements.server.mandatory.Identify)
});
const WindowCoveringDevice = WindowCoveringDeviceDefinition;
