// @matter/types/dist/cjs/globals/MeasurementAccuracyRange.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MeasurementAccuracyRange_exports = {};
__export(MeasurementAccuracyRange_exports, {
  TlvMeasurementAccuracyRange: () => TlvMeasurementAccuracyRange
});
module.exports = __toCommonJS(MeasurementAccuracyRange_exports);
var import_TlvObject = require("../tlv/TlvObject.js");
var import_TlvNumber = require("../tlv/TlvNumber.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvMeasurementAccuracyRange = (0, import_TlvObject.TlvObject)({
  /**
   * This field shall indicate the minimum measurement value for the specified level of accuracy.
   *
   * The value of this field shall be greater than or equal to the value of the MinMeasuredValue field on the
   * encompassing MeasurementAccuracyStruct.
   *
   * The value of this field shall be less than or equal to the value of the MaxMeasuredValue field on the
   * encompassing MeasurementAccuracyStruct.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.1
   */
  rangeMin: (0, import_TlvObject.TlvField)(0, import_TlvNumber.TlvInt64),
  /**
   * This field shall indicate the maximum measurement value for the specified level of accuracy. The value of this
   * field shall be greater than the value of the RangeMin field.
   *
   * The value of this field shall be greater than or equal to the value of the MinMeasuredValue field on the
   * encompassing MeasurementAccuracyStruct.
   *
   * The value of this field shall be less than or equal to the value of the MaxMeasuredValue field on the
   * encompassing MeasurementAccuracyStruct.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.2
   */
  rangeMax: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvInt64),
  /**
   * This field shall indicate the maximum +/- percentage accuracy for the associated measurement.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.3
   */
  percentMax: (0, import_TlvObject.TlvOptionalField)(2, import_TlvNumber.TlvPercent100ths),
  /**
   * This field shall indicate the minimum +/- percentage accuracy for the associated measurement.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.4
   */
  percentMin: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvPercent100ths),
  /**
   * This field shall indicate the typical +/- percentage accuracy for the associated measurement.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.5
   */
  percentTypical: (0, import_TlvObject.TlvOptionalField)(4, import_TlvNumber.TlvPercent100ths),
  /**
   * This field shall indicate the maximum +/- fixed accuracy for the associated measurement, in the unit indicated by
   * MeasurementType.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.6
   */
  fixedMax: (0, import_TlvObject.TlvOptionalField)(5, import_TlvNumber.TlvUInt64),
  /**
   * This field shall indicate the minimum +/- fixed accuracy for the associated measurement, in the unit indicated by
   * MeasurementType.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.7
   */
  fixedMin: (0, import_TlvObject.TlvOptionalField)(6, import_TlvNumber.TlvUInt64),
  /**
   * This field shall indicate the typical +/- fixed accuracy for the associated measurement, in the unit indicated by
   * MeasurementType.
   *
   * @see {@link MatterSpecification.v14.Cluster} § 2.1.4.3.8
   */
  fixedTypical: (0, import_TlvObject.TlvOptionalField)(7, import_TlvNumber.TlvUInt64)
});
