// @matter/model/dist/esm/standard/elements/AcceptedCommandList.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { AttributeElement as Attribute, FieldElement as Field } from "../../elements/index.js";
const AcceptedCommandList = Attribute(
  {
    name: "AcceptedCommandList",
    id: 65529,
    type: "list",
    access: "R V",
    conformance: "M",
    isSeed: true,
    quality: "F",
    details: "This attribute is a list of client generated commands which are supported by this cluster server instance.\nEach instance of a cluster shall support this attribute.\nThis attribute shall be a list of the command IDs for client generated commands that are supported and processed by the server.\nFor each client request command in this list that mandates a response from the server, the\nresponse command shall be indicated in the GeneratedCommandList attribute.\nIf any attribute on a server supports atomic writes, this attribute shall contain the command ID for AtomicRequest.",
    xref: { document: "core", section: "7.13.4" }
  },
  Field({ name: "entry", type: "command-id" })
);
MatterDefinition.children.push(AcceptedCommandList);
export {
  AcceptedCommandList
};
