// @matter/model/dist/esm/standard/elements/AirPurifierDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const AirPurifierDt = DeviceType(
  {
    name: "AirPurifier",
    id: 45,
    category: "HVAC",
    classification: "simple",
    details: "An Air Purifier is a standalone device that is designed to clean the air in a room.\nIt is a device that has a fan to control the air speed while it is operating. Optionally, it can report on the condition of its filters.",
    xref: { document: "device", section: "9.3" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 45, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  Requirement({
    name: "OnOff",
    id: 6,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  Requirement({
    name: "FanControl",
    id: 514,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  Requirement({
    name: "HepaFilterMonitoring",
    id: 113,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  }),
  Requirement({
    name: "ActivatedCarbonFilterMonitoring",
    id: 114,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.3.5" }
  })
);
MatterDefinition.children.push(AirPurifierDt);
export {
  AirPurifierDt
};
