// @matter/model/dist/esm/standard/elements/AirQualitySensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const AirQualitySensorDt = DeviceType(
  {
    name: "AirQualitySensor",
    id: 44,
    category: "Sensor",
    classification: "simple",
    details: "This defines conformance for the Air Quality Sensor device type.\nAn air quality sensor is a device designed to monitor and measure various parameters related to the quality of ambient air in indoor or outdoor environments.",
    xref: { document: "device", section: "7.10" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 44, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "AirQuality",
    id: 91,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "RelativeHumidityMeasurement",
    id: 1029,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "CarbonMonoxideConcentrationMeasurement",
    id: 1036,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "CarbonDioxideConcentrationMeasurement",
    id: 1037,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "NitrogenDioxideConcentrationMeasurement",
    id: 1043,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "OzoneConcentrationMeasurement",
    id: 1045,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "FormaldehydeConcentrationMeasurement",
    id: 1067,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "Pm1ConcentrationMeasurement",
    id: 1068,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "Pm25ConcentrationMeasurement",
    id: 1066,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "Pm10ConcentrationMeasurement",
    id: 1069,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "RadonConcentrationMeasurement",
    id: 1071,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  }),
  Requirement({
    name: "TotalVolatileOrganicCompoundsConcentrationMeasurement",
    id: 1070,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.10.4" }
  })
);
MatterDefinition.children.push(AirQualitySensorDt);
export {
  AirQualitySensorDt
};
