// @matter/model/dist/esm/standard/elements/AreaNamespaceNS.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  SemanticNamespaceElement as SemanticNamespace,
  SemanticTagElement as SemanticTag
} from "../../elements/index.js";
const AreaNamespaceNs = SemanticNamespace(
  {
    name: "AreaNamespace",
    id: 16,
    details: "The tags contained in this namespace may be used in any domain or context, to indicate an association with an indoor or outdoor area of a home.",
    xref: { document: "namespace", section: "13" }
  },
  SemanticTag({ name: "Aisle", id: 0 }),
  SemanticTag({ name: "Attic", id: 1 }),
  SemanticTag({ name: "Back Door", id: 2 }),
  SemanticTag({ name: "Back Yard", id: 3 }),
  SemanticTag({ name: "Balcony", id: 4 }),
  SemanticTag({ name: "Ballroom", id: 5 }),
  SemanticTag({ name: "Bathroom", id: 6, description: "Also known as Restroom" }),
  SemanticTag({ name: "Bedroom", id: 7 }),
  SemanticTag({ name: "Border", id: 8 }),
  SemanticTag({ name: "Boxroom", id: 9, description: "A small room typically used for storage" }),
  SemanticTag({ name: "Breakfast Room", id: 10 }),
  SemanticTag({ name: "Carport", id: 11 }),
  SemanticTag({ name: "Cellar", id: 12 }),
  SemanticTag({ name: "Cloakroom", id: 13 }),
  SemanticTag({ name: "Closet", id: 14 }),
  SemanticTag({ name: "Conservatory", id: 15 }),
  SemanticTag({ name: "Corridor", id: 16 }),
  SemanticTag({ name: "Craft Room", id: 17 }),
  SemanticTag({ name: "Cupboard", id: 18 }),
  SemanticTag({ name: "Deck", id: 19 }),
  SemanticTag({
    name: "Den",
    id: 20,
    description: "A small, comfortable room for individual activities such as work or hobbies"
  }),
  SemanticTag({ name: "Dining", id: 21 }),
  SemanticTag({ name: "Drawing Room", id: 22 }),
  SemanticTag({ name: "Dressing Room", id: 23 }),
  SemanticTag({ name: "Driveway", id: 24 }),
  SemanticTag({ name: "Elevator", id: 25 }),
  SemanticTag({ name: "Ensuite", id: 26, description: "A bathroom directly accessible from a bedroom" }),
  SemanticTag({ name: "Entrance", id: 27 }),
  SemanticTag({ name: "Entryway", id: 28 }),
  SemanticTag({ name: "Family Room", id: 29 }),
  SemanticTag({ name: "Foyer", id: 30 }),
  SemanticTag({ name: "Front Door", id: 31 }),
  SemanticTag({ name: "Front Yard", id: 32 }),
  SemanticTag({ name: "Game Room", id: 33 }),
  SemanticTag({ name: "Garage", id: 34 }),
  SemanticTag({ name: "Garage Door", id: 35 }),
  SemanticTag({ name: "Garden", id: 36 }),
  SemanticTag({ name: "Garden Door", id: 37 }),
  SemanticTag({ name: "Guest Bathroom", id: 38, description: "Also known as Guest Restroom" }),
  SemanticTag({ name: "Guest Bedroom", id: 39 }),
  SemanticTag({ name: "Guest Room", id: 40 }),
  SemanticTag({ name: "Gym", id: 41 }),
  SemanticTag({ name: "Hallway", id: 42 }),
  SemanticTag(
    { name: "Hearth Room", id: 43, description: "A cozy room containing a fireplace or other point heat source" }
  ),
  SemanticTag({ name: "Kids Room", id: 44 }),
  SemanticTag({ name: "Kids Bedroom", id: 45 }),
  SemanticTag({ name: "Kitchen", id: 46 }),
  SemanticTag({ name: "Laundry Room", id: 47 }),
  SemanticTag({ name: "Lawn", id: 48 }),
  SemanticTag({ name: "Library", id: 49 }),
  SemanticTag({ name: "Living Room", id: 50 }),
  SemanticTag({ name: "Lounge", id: 51 }),
  SemanticTag({ name: "Media/TV Room", id: 52 }),
  SemanticTag({
    name: "Mud Room",
    id: 53,
    description: "A space used to remove soiled garments prior to entering the domicile proper"
  }),
  SemanticTag({ name: "Music Room", id: 54 }),
  SemanticTag({ name: "Nursery", id: 55 }),
  SemanticTag({ name: "Office", id: 56 }),
  SemanticTag({ name: "Outdoor Kitchen", id: 57 }),
  SemanticTag({ name: "Outside", id: 58 }),
  SemanticTag({ name: "Pantry", id: 59, description: "AKA a larder, a place where food is stored" }),
  SemanticTag({ name: "Parking Lot", id: 60 }),
  SemanticTag({ name: "Parlor", id: 61 }),
  SemanticTag({ name: "Patio", id: 62 }),
  SemanticTag({ name: "Play Room", id: 63 }),
  SemanticTag({ name: "Pool Room", id: 64 }),
  SemanticTag({ name: "Porch", id: 65 }),
  SemanticTag({ name: "Primary Bathroom", id: 66 }),
  SemanticTag({ name: "Primary Bedroom", id: 67 }),
  SemanticTag({ name: "Ramp", id: 68 }),
  SemanticTag({ name: "Reception Room", id: 69 }),
  SemanticTag({ name: "Recreation Room", id: 70 }),
  SemanticTag({ name: "Roof", id: 71 }),
  SemanticTag({ name: "Sauna", id: 72 }),
  SemanticTag({ name: "Scullery", id: 73, description: "A utility space for cleaning dishes and laundry" }),
  SemanticTag({ name: "Sewing Room", id: 74 }),
  SemanticTag({ name: "Shed", id: 75 }),
  SemanticTag({ name: "Side Door", id: 76 }),
  SemanticTag({ name: "Side Yard", id: 77 }),
  SemanticTag({ name: "Sitting Room", id: 78 }),
  SemanticTag({
    name: "Snug",
    id: 79,
    description: "An informal space meant to be 'cozy', 'snug', relaxed, meant to share with family or friends"
  }),
  SemanticTag({ name: "Spa", id: 80 }),
  SemanticTag({ name: "Staircase", id: 81 }),
  SemanticTag({ name: "Steam Room", id: 82 }),
  SemanticTag({ name: "Storage Room", id: 83 }),
  SemanticTag({ name: "Studio", id: 84 }),
  SemanticTag({ name: "Study", id: 85 }),
  SemanticTag({ name: "Sun Room", id: 86 }),
  SemanticTag({ name: "Swimming Pool", id: 87 }),
  SemanticTag({ name: "Terrace", id: 88 }),
  SemanticTag({ name: "Toilet", id: 89, description: "A room dedicated to a toilet; a water closet / WC" }),
  SemanticTag({ name: "Utility Room", id: 90 }),
  SemanticTag({ name: "Ward", id: 91 }),
  SemanticTag({ name: "Workshop", id: 92 })
);
MatterDefinition.children.push(AreaNamespaceNs);
export {
  AreaNamespaceNs
};
