// @matter/model/dist/esm/standard/elements/CastingVideoClientDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const CastingVideoClientDt = DeviceType(
  {
    name: "CastingVideoClient",
    id: 41,
    category: "Media",
    classification: "simple",
    details: "This defines conformance to the Casting Video Client device type.\nA Casting Video Client is a client that can launch content on a Casting Video Player, for example, a Smart Speaker or a Content Provider phone app.",
    xref: { document: "device", section: "10.6" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 41, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "LevelControl",
    id: 8,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "WakeOnLan",
    id: 1283,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "Channel",
    id: 1284,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "TargetNavigator",
    id: 1285,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "MediaPlayback",
    id: 1286,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "MediaInput",
    id: 1287,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "LowPower",
    id: 1288,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "KeypadInput",
    id: 1289,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "ContentLauncher",
    id: 1290,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "AudioOutput",
    id: 1291,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "ApplicationLauncher",
    id: 1292,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "ApplicationBasic",
    id: 1293,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "AccountLogin",
    id: 1294,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "ContentControl",
    id: 1295,
    conformance: "P, O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "ContentAppObserver",
    id: 1296,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.6.4" }
  }),
  Requirement({
    name: "Messages",
    id: 151,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "10.6.4" }
  })
);
MatterDefinition.children.push(CastingVideoClientDt);
export {
  CastingVideoClientDt
};
