// @matter/model/dist/esm/standard/elements/ContentAppDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  DeviceTypeElement as DeviceType,
  RequirementElement as Requirement,
  FieldElement as Field
} from "../../elements/index.js";
const ContentAppDt = DeviceType(
  {
    name: "ContentApp",
    id: 36,
    category: "Media",
    classification: "simple",
    details: "This defines conformance to the Content App device type.\nA Content App is usually an application built by a Content Provider. A Casting Video Player with a Content App Platform is able to launch Content Apps and represent these apps as separate endpoints.",
    xref: { document: "device", section: "10.5" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 36, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "Binding",
    id: 30,
    conformance: "ObserverClient",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "Channel",
    id: 1284,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "TargetNavigator",
    id: 1285,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "MediaPlayback",
    id: 1286,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "KeypadInput",
    id: 1289,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "ContentLauncher",
    id: 1290,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement(
    {
      name: "ApplicationLauncher",
      id: 1292,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "10.5.4" }
    },
    Requirement({ name: "APPLICATIONPLATFORM", conformance: "X", element: "feature" })
  ),
  Requirement({
    name: "ApplicationBasic",
    id: 1293,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "AccountLogin",
    id: 1294,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Requirement({
    name: "ContentAppObserver",
    id: 1296,
    conformance: "ObserverClient",
    element: "clientCluster",
    xref: { document: "device", section: "10.5.4" }
  }),
  Field(
    { name: "conditions", type: "enum8" },
    Field({
      name: "ObserverClient",
      description: "The node is a client for ContentAppObservers.",
      xref: { document: "device", section: "10.5.3" }
    })
  )
);
MatterDefinition.children.push(ContentAppDt);
export {
  ContentAppDt
};
