// @matter/model/dist/esm/standard/elements/ContactSensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const ContactSensorDt = DeviceType(
  {
    name: "ContactSensor",
    id: 21,
    category: "Sensor",
    classification: "simple",
    details: "This defines conformance to the Contact Sensor device type.",
    xref: { document: "device", section: "7.1" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 21, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.1.4" }
  }),
  Requirement({
    name: "BooleanState",
    id: 69,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.1.4" }
  }),
  Requirement({
    name: "BooleanStateConfiguration",
    id: 128,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.1.4" }
  })
);
MatterDefinition.children.push(ContactSensorDt);
export {
  ContactSensorDt
};
