// @matter/model/dist/esm/standard/elements/DeviceEnergyManagementDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  DeviceTypeElement as DeviceType,
  RequirementElement as Requirement,
  FieldElement as Field
} from "../../elements/index.js";
const DeviceEnergyManagementDt = DeviceType(
  {
    name: "DeviceEnergyManagement",
    id: 1293,
    category: "Utility",
    classification: "utility",
    details: "A Device Energy Management device provides reporting and optionally adjustment of the electrical power planned on being consumed or produced by the device.",
    xref: { document: "device", section: "2.7" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 1293, revision: 2 }], element: "attribute" })
  ),
  Requirement(
    {
      name: "DeviceEnergyManagement",
      id: 152,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "2.7.4" }
    },
    Requirement({ name: "POWERADJUSTMENT", conformance: "ControllableESA.a+", element: "feature" }),
    Requirement({ name: "STARTTIMEADJUSTMENT", conformance: "ControllableESA.a+", element: "feature" }),
    Requirement({ name: "PAUSABLE", conformance: "ControllableESA.a+", element: "feature" }),
    Requirement({ name: "FORECASTADJUSTMENT", conformance: "ControllableESA.a+", element: "feature" }),
    Requirement({ name: "CONSTRAINTBASEDADJUSTMENT", conformance: "ControllableESA.a+", element: "feature" })
  ),
  Requirement({
    name: "DeviceEnergyManagementMode",
    id: 159,
    conformance: "ControllableESA, O",
    element: "serverCluster",
    xref: { document: "device", section: "2.7.4" }
  }),
  Field(
    { name: "conditions", type: "enum8" },
    Field({
      name: "ControllableEsa",
      description: "The DEM cluster on this endpoint accepts commands to adjust its energy operation.",
      xref: { document: "device", section: "2.7.3" }
    })
  )
);
MatterDefinition.children.push(DeviceEnergyManagementDt);
export {
  DeviceEnergyManagementDt
};
