// @matter/model/dist/esm/standard/elements/DoorLockDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const DoorLockDt = DeviceType(
  {
    name: "DoorLock",
    id: 10,
    category: "Closure",
    classification: "simple",
    details: "A Door Lock is a device used to secure a door. It is possible to actuate a door lock either by means of a manual or a remote method.",
    xref: { document: "device", section: "8.1" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 10, revision: 3 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "8.1.4" }
  }),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "X",
    element: "serverCluster",
    xref: { document: "device", section: "8.1.4" }
  }),
  Requirement({
    name: "ScenesManagement",
    id: 98,
    conformance: "X",
    element: "serverCluster",
    xref: { document: "device", section: "8.1.4" }
  }),
  Requirement(
    {
      name: "DoorLock",
      id: 257,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "8.1.4" }
    },
    Requirement({ name: "USER", conformance: "Matter & (PIN | RID | FPG | FACE | ALIRO)", element: "feature" }),
    Requirement({ name: "RFIDCREDENTIAL", conformance: "P, O", element: "feature" }),
    Requirement({ name: "AlarmMask", conformance: "[Alarms]", element: "attribute" })
  )
);
MatterDefinition.children.push(DoorLockDt);
export {
  DoorLockDt
};
