// @matter/model/dist/esm/standard/elements/EnergyEvseDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const EnergyEvseDt = DeviceType(
  {
    name: "EnergyEvse",
    id: 1292,
    category: "Energy",
    classification: "simple",
    details: "An EVSE (Electric Vehicle Supply Equipment) is a device that allows an EV (Electric Vehicle) to be connected to the mains electricity supply to allow it to be charged (or discharged in case of Vehicle to Grid / Vehicle to Home applications).",
    xref: { document: "device", section: "14.1" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 1292, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  }),
  Requirement({
    name: "EnergyEvse",
    id: 153,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  }),
  Requirement({
    name: "EnergyEvseMode",
    id: 157,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  }),
  Requirement({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.1.6" }
  })
);
MatterDefinition.children.push(EnergyEvseDt);
export {
  EnergyEvseDt
};
