// @matter/model/dist/esm/standard/elements/ExtractorHoodDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const ExtractorHoodDt = DeviceType(
  {
    name: "ExtractorHood",
    id: 122,
    category: "Appliances",
    classification: "simple",
    details: "An Extractor Hood is a device that is generally installed above a cooking surface in residential kitchens. An Extractor Hood\u2019s primary purpose is to reduce odors that arise during the cooking process by either extracting the air above the cooking surface or by recirculating and filtering it. It may also contain a light for illuminating the cooking surface.\nExtractor Hoods may also be known by the following names:\n  \u2022 Hoods\n  \u2022 Extractor Fans\n  \u2022 Extractors\n  \u2022 Range Hoods\n  \u2022 Telescoping Hoods\n  \u2022 Telescoping Extractors",
    xref: { document: "device", section: "13.10" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 122, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.10.5" }
  }),
  Requirement({
    name: "HepaFilterMonitoring",
    id: 113,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.10.5" }
  }),
  Requirement({
    name: "ActivatedCarbonFilterMonitoring",
    id: 114,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.10.5" }
  }),
  Requirement(
    {
      name: "FanControl",
      id: 514,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.10.5" }
    },
    Requirement({ name: "ROCKING", conformance: "X", element: "feature" }),
    Requirement({ name: "WIND", conformance: "X", element: "feature" }),
    Requirement({ name: "AIRFLOWDIRECTION", conformance: "X", element: "feature" })
  )
);
MatterDefinition.children.push(ExtractorHoodDt);
export {
  ExtractorHoodDt
};
