// @matter/model/dist/esm/standard/elements/FanDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const FanDt = DeviceType(
  {
    name: "Fan",
    id: 43,
    category: "HVAC",
    classification: "simple",
    details: "A Fan device is typically standalone or mounted on a ceiling or wall and is used to circulate air in a room.",
    xref: { document: "device", section: "9.2" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 43, revision: 3 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.2.5" }
  }),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "9.2.5" }
  }),
  Requirement({
    name: "OnOff",
    id: 6,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "9.2.5" }
  }),
  Requirement(
    {
      name: "FanControl",
      id: 514,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "9.2.5" }
    },
    Requirement({ name: "FanModeSequence", access: "R V", conformance: "Matter", element: "attribute" })
  )
);
MatterDefinition.children.push(FanDt);
export {
  FanDt
};
