// @matter/model/dist/esm/standard/elements/FlowSensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const FlowSensorDt = DeviceType(
  {
    name: "FlowSensor",
    id: 774,
    category: "Sensor",
    classification: "simple",
    details: "A Flow Sensor device measures and reports the flow rate of a fluid.",
    xref: { document: "device", section: "7.6" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 774, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.6.4" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.6.4" }
  })
);
MatterDefinition.children.push(FlowSensorDt);
export {
  FlowSensorDt
};
