// @matter/model/dist/esm/standard/elements/HeatPumpDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const HeatPumpDt = DeviceType(
  {
    name: "HeatPump",
    id: 777,
    category: "Energy",
    classification: "simple",
    details: "A Heat Pump device is a device that uses electrical energy to heat either spaces or water tanks using ground, water or air as the heat source. These typically can heat the air or can pump water via central heating radiators or underfloor heating systems. It is typical to also heat hot water and store the heat in a hot water tank.\nNote that the Water Heater device type can also be heated by a heat pump and has similar requirements, but that cannot be used for space heating.",
    xref: { document: "device", section: "14.5" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 777, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.5.6" }
  }),
  Requirement({
    name: "Thermostat",
    id: 513,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "14.5.6" }
  })
);
MatterDefinition.children.push(HeatPumpDt);
export {
  HeatPumpDt
};
