// @matter/model/dist/esm/standard/elements/HumiditySensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const HumiditySensorDt = DeviceType(
  {
    name: "HumiditySensor",
    id: 775,
    category: "Sensor",
    classification: "simple",
    details: "A humidity sensor (in most cases a Relative humidity sensor) reports humidity measurements.",
    xref: { document: "device", section: "7.7" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 775, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.7.4" }
  }),
  Requirement({
    name: "RelativeHumidityMeasurement",
    id: 1029,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.7.4" }
  })
);
MatterDefinition.children.push(HumiditySensorDt);
export {
  HumiditySensorDt
};
