// @matter/model/dist/esm/standard/elements/LaundryDryerDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const LaundryDryerDt = DeviceType(
  {
    name: "LaundryDryer",
    id: 124,
    category: "Appliances",
    classification: "simple",
    details: "A Laundry Dryer represents a device that is capable of drying laundry items.",
    xref: { document: "device", section: "13.6" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 124, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  }),
  Requirement(
    {
      name: "LaundryWasherMode",
      id: 81,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.6.4" }
    },
    Requirement({ name: "StartUpMode", conformance: "X", element: "attribute" }),
    Requirement({ name: "ONOFF", conformance: "X", element: "feature" })
  ),
  Requirement(
    {
      name: "OnOff",
      id: 6,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.6.4" }
    },
    Requirement({ name: "DEADFRONTBEHAVIOR", conformance: "M", element: "feature" })
  ),
  Requirement({
    name: "LaundryDryerControls",
    id: 74,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  }),
  Requirement({
    name: "TemperatureControl",
    id: 86,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  }),
  Requirement({
    name: "OperationalState",
    id: 96,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.6.4" }
  })
);
MatterDefinition.children.push(LaundryDryerDt);
export {
  LaundryDryerDt
};
