// @matter/model/dist/esm/standard/elements/MicrowaveOvenDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const MicrowaveOvenDt = DeviceType(
  {
    name: "MicrowaveOven",
    id: 121,
    category: "Appliances",
    classification: "simple",
    details: "This defines conformance to the Microwave Oven device type.\nA Microwave Oven is a device with the primary function of heating foods and beverages using a magnetron.",
    xref: { document: "device", section: "13.11" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 121, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "13.11.6" }
  }),
  Requirement(
    {
      name: "OperationalState",
      id: 96,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "13.11.6" }
    },
    Requirement({ name: "CountdownTime", conformance: "M", element: "attribute" })
  ),
  Requirement(
    {
      name: "FanControl",
      id: 514,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "13.11.6" }
    },
    Requirement({ name: "WIND", conformance: "X", element: "feature" }),
    Requirement({ name: "AIRFLOWDIRECTION", conformance: "X", element: "feature" })
  ),
  Requirement({
    name: "MicrowaveOvenMode",
    id: 94,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.11.6" }
  }),
  Requirement({
    name: "MicrowaveOvenControl",
    id: 95,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "13.11.6" }
  })
);
MatterDefinition.children.push(MicrowaveOvenDt);
export {
  MicrowaveOvenDt
};
