// @matter/model/dist/esm/standard/elements/OnOffPlugInUnitDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const OnOffPlugInUnitDt = DeviceType(
  {
    name: "OnOffPlugInUnit",
    id: 266,
    category: "Smart Plugs/Outlets and other Actuators",
    classification: "simple",
    details: "An On/Off Plug-in Unit is a device that provides power to another device that is plugged into it, and is capable of switching that provided power on or off.",
    xref: { document: "device", section: "5.1" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 266, revision: 3 }], element: "attribute" })
  ),
  Requirement(
    {
      name: "Identify",
      id: 3,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "5.1.4" }
    },
    Requirement({ name: "TriggerEffect", conformance: "M", element: "command" })
  ),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.1.4" }
  }),
  Requirement(
    {
      name: "ScenesManagement",
      id: 98,
      conformance: "P, M",
      element: "serverCluster",
      xref: { document: "device", section: "5.1.4" }
    },
    Requirement({ name: "CopyScene", conformance: "P, M", element: "command" })
  ),
  Requirement(
    {
      name: "OnOff",
      id: 6,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "5.1.4" }
    },
    Requirement({ name: "LIGHTING", conformance: "M", element: "feature" })
  ),
  Requirement(
    {
      name: "LevelControl",
      id: 8,
      conformance: "O",
      element: "serverCluster",
      xref: { document: "device", section: "5.1.4" }
    },
    Requirement({ name: "ONOFF", conformance: "M", element: "feature" }),
    Requirement({ name: "LIGHTING", conformance: "M", element: "feature" }),
    Requirement({ name: "CurrentLevel", constraint: "1 to 254", element: "attribute" }),
    Requirement({ name: "MinLevel", constraint: "1", element: "attribute" }),
    Requirement({ name: "MaxLevel", constraint: "254", element: "attribute" })
  ),
  Requirement({
    name: "OccupancySensing",
    id: 1030,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.1.4" }
  })
);
MatterDefinition.children.push(OnOffPlugInUnitDt);
export {
  OnOffPlugInUnitDt
};
