// @matter/model/dist/esm/standard/elements/OtaRequestorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const OtaRequestorDt = DeviceType(
  {
    name: "OtaRequestor",
    id: 18,
    category: "Utility",
    classification: "utility",
    details: "An OTA Requestor is a device that is capable of receiving an OTA software update.",
    xref: { document: "device", section: "2.3" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 18, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "OtaSoftwareUpdateRequestor",
    id: 42,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "2.3.3" }
  }),
  Requirement({
    name: "OtaSoftwareUpdateProvider",
    id: 41,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "2.3.3" }
  })
);
MatterDefinition.children.push(OtaRequestorDt);
export {
  OtaRequestorDt
};
