// @matter/model/dist/esm/standard/elements/OvenMode.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field,
  DatatypeElement as Datatype
} from "../../elements/index.js";
const OvenMode = Cluster(
  {
    name: "OvenMode",
    id: 73,
    type: "ModeBase",
    classification: "application",
    pics: "OTCCM",
    details: "This cluster is derived from the Mode Base cluster and defines additional mode tags and namespaced enumerated values for oven devices.",
    xref: { document: "cluster", section: "8.11" }
  },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 2 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap", xref: { document: "cluster", section: "8.11.4" } },
    Field({
      name: "DEPONOFF",
      conformance: "X",
      constraint: "0",
      description: "OnOff",
      details: "Dependency with the OnOff cluster"
    })
  ),
  Attribute({
    name: "SupportedModes",
    id: 0,
    details: "At least one entry in the SupportedModes attribute shall include the Bake mode tag in the ModeTags field list.",
    xref: { document: "cluster", section: "8.11.6.1" }
  }),
  Attribute({ name: "CurrentMode", id: 1, xref: { document: "cluster", section: "8.11.6" } }),
  Attribute({ name: "StartUpMode", id: 2, conformance: "X", xref: { document: "cluster", section: "8.11.6" } }),
  Attribute({ name: "OnMode", id: 3, conformance: "X", xref: { document: "cluster", section: "8.11.6" } }),
  Datatype({
    name: "ModeOptionStruct",
    type: "ModeOptionStruct",
    details: "The table below lists the changes relative to the Mode Base cluster for the fields of the ModeOptionStruct type. A blank field indicates no change.",
    xref: { document: "cluster", section: "8.11.5.1" }
  }),
  Datatype(
    { name: "ModeTag", type: "enum16" },
    Field({ name: "Auto", id: 0, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Quick", id: 1, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Quiet", id: 2, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "LowNoise", id: 3, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "LowEnergy", id: 4, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Vacation", id: 5, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Min", id: 6, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Max", id: 7, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Night", id: 8, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "Day", id: 9, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({
      name: "Bake",
      id: 16384,
      details: "This mode sets the device into baking mode for baking food items.",
      xref: { document: "cluster", section: "8.11.7.1.1" }
    }),
    Field({
      name: "Convection",
      id: 16385,
      details: "This mode sets the device into convection mode which creates an airflow within the device during the cooking duration.",
      xref: { document: "cluster", section: "8.11.7.1.2" }
    }),
    Field({
      name: "Grill",
      id: 16386,
      details: "This mode sets the device into grill mode for grilling food items. This is the same as Broil for many regions.",
      xref: { document: "cluster", section: "8.11.7.1.3" }
    }),
    Field({
      name: "Roast",
      id: 16387,
      details: "This mode sets the device into roast mode for roasting food items.",
      xref: { document: "cluster", section: "8.11.7.1.4" }
    }),
    Field({
      name: "Clean",
      id: 16388,
      details: "This mode sets the device into cleaning mode to clean the internal components of the appliance.",
      xref: { document: "cluster", section: "8.11.7.1.5" }
    }),
    Field({ name: "ConvectionBake", id: 16389, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({ name: "ConvectionRoast", id: 16390, xref: { document: "cluster", section: "8.11.7.1" } }),
    Field({
      name: "Warming",
      id: 16391,
      details: "This mode sets the device into a warming mode which begins warming the cavity.",
      xref: { document: "cluster", section: "8.11.7.1.8" }
    }),
    Field({
      name: "Proofing",
      id: 16392,
      details: "This mode sets the device into proofing mode which creates an environment ready for proofing.",
      xref: { document: "cluster", section: "8.11.7.1.9" }
    }),
    Field({ name: "Steam", id: 16393, xref: { document: "cluster", section: "8.11.7.1" } })
  )
);
MatterDefinition.children.push(OvenMode);
export {
  OvenMode
};
