// @matter/model/dist/esm/standard/elements/PowerTopology.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  ClusterElement as Cluster,
  AttributeElement as Attribute,
  FieldElement as Field
} from "../../elements/index.js";
const PowerTopology = Cluster(
  {
    name: "PowerTopology",
    id: 156,
    classification: "application",
    pics: "PWRTL",
    details: "The Power Topology Cluster provides a mechanism for expressing how power is flowing between endpoints.",
    xref: { document: "core", section: "11.8" }
  },
  Attribute({ name: "ClusterRevision", id: 65533, type: "ClusterRevision", default: 1 }),
  Attribute(
    { name: "FeatureMap", id: 65532, type: "FeatureMap", xref: { document: "core", section: "11.8.4" } },
    Field({
      name: "NODE",
      conformance: "O.a",
      constraint: "0",
      description: "NodeTopology",
      details: "This endpoint provides or consumes power to/from the entire node"
    }),
    Field({
      name: "TREE",
      conformance: "O.a",
      constraint: "1",
      description: "TreeTopology",
      details: "This endpoint provides or consumes power to/from itself and its child endpoints"
    }),
    Field({
      name: "SET",
      conformance: "O.a",
      constraint: "2",
      description: "SetTopology",
      details: "This endpoint provides or consumes power to/from a specified set of endpoints"
    }),
    Field({
      name: "DYPF",
      conformance: "[SET]",
      constraint: "3",
      description: "DynamicPowerFlow",
      details: "The specified set of endpoints may change"
    })
  ),
  Attribute(
    {
      name: "AvailableEndpoints",
      id: 0,
      type: "list",
      access: "R V",
      conformance: "SET",
      constraint: "max 20",
      quality: "F",
      details: "Indicates the list of endpoints capable of providing power to and/or consuming power from the endpoint hosting this server.",
      xref: { document: "core", section: "11.8.5.1" }
    },
    Field({ name: "entry", type: "endpoint-no" })
  ),
  Attribute(
    {
      name: "ActiveEndpoints",
      id: 1,
      type: "list",
      access: "R V",
      conformance: "DYPF",
      constraint: "max 20",
      quality: "N",
      details: "Indicates the current list of endpoints currently providing or consuming power to or from the endpoint hosting this server. This list shall be a subset of the value of the AvailableEndpoints attribute.",
      xref: { document: "core", section: "11.8.5.2" }
    },
    Field({ name: "entry", type: "endpoint-no" })
  )
);
MatterDefinition.children.push(PowerTopology);
export {
  PowerTopology
};
