// @matter/model/dist/esm/standard/elements/PowerSourceNS.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import {
  SemanticNamespaceElement as SemanticNamespace,
  SemanticTagElement as SemanticTag
} from "../../elements/index.js";
const PowerSourceNs = SemanticNamespace(
  {
    name: "PowerSource",
    id: 15,
    details: "The tags contained in this namespace are restricted for use in the power source domain and shall NOT be used in any other domain or context.",
    xref: { document: "namespace", section: "15" }
  },
  SemanticTag({
    name: "Unknown",
    id: 0,
    description: "The Power Source cluster is related to power provided from an unknown source"
  }),
  SemanticTag({
    name: "Grid",
    id: 1,
    description: "The Power Source cluster is related to power provided from the electrical grid",
    details: "Power Source clusters with this tag shall implement the WIRED feature.",
    xref: { document: "namespace", section: "15.1" }
  }),
  SemanticTag({
    name: "Solar",
    id: 2,
    description: "The Power Source cluster is related to power provided from a solar panel array",
    details: "Power Source clusters with this tag shall implement the WIRED feature.",
    xref: { document: "namespace", section: "15.2" }
  }),
  SemanticTag({
    name: "Battery",
    id: 3,
    description: "The Power Source cluster is related to power provided from a battery",
    details: "Power Source clusters with this tag shall implement the BAT feature.",
    xref: { document: "namespace", section: "15.3" }
  }),
  SemanticTag({
    name: "EV",
    id: 4,
    description: "The Power Source cluster is related to power provided from an electric vehicle",
    details: "Power Source clusters with this tag shall implement the BAT feature.",
    xref: { document: "namespace", section: "15.4" }
  })
);
MatterDefinition.children.push(PowerSourceNs);
export {
  PowerSourceNs
};
