// @matter/model/dist/esm/standard/elements/PressureSensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const PressureSensorDt = DeviceType(
  {
    name: "PressureSensor",
    id: 773,
    category: "Sensor",
    classification: "simple",
    details: "A Pressure Sensor device measures and reports the pressure of a fluid.",
    xref: { document: "device", section: "7.5" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 773, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.5.4" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.5.4" }
  })
);
MatterDefinition.children.push(PressureSensorDt);
export {
  PressureSensorDt
};
