// @matter/model/dist/esm/standard/elements/PumpControllerDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const PumpControllerDt = DeviceType(
  {
    name: "PumpController",
    id: 772,
    category: "Switches and Controls",
    classification: "simple",
    details: "A Pump Controller device is capable of configuring and controlling a Pump device.",
    xref: { document: "device", section: "6.5" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 772, revision: 4 }], element: "attribute" })
  ),
  Requirement({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "PumpConfigurationAndControl",
    id: 512,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "LevelControl",
    id: 8,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "ScenesManagement",
    id: 98,
    conformance: "P, O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  }),
  Requirement({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "6.5.3" }
  })
);
MatterDefinition.children.push(PumpControllerDt);
export {
  PumpControllerDt
};
