// @matter/model/dist/esm/standard/elements/PumpDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const PumpDt = DeviceType(
  {
    name: "Pump",
    id: 771,
    category: "Smart Plugs/Outlets and other Actuators",
    classification: "simple",
    details: "A Pump device is a pump that may have variable speed. It may have optional built-in sensors and a regulation mechanism. It is typically used for pumping fluids like water.",
    xref: { document: "device", section: "5.5" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 771, revision: 3 }], element: "attribute" })
  ),
  Requirement({
    name: "OnOff",
    id: 6,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "PumpConfigurationAndControl",
    id: 512,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "LevelControl",
    id: 8,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "ScenesManagement",
    id: 98,
    conformance: "P, O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "PressureMeasurement",
    id: 1027,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  }),
  Requirement({
    name: "OccupancySensing",
    id: 1030,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.5.4" }
  })
);
MatterDefinition.children.push(PumpDt);
export {
  PumpDt
};
