// @matter/model/dist/esm/standard/elements/RainSensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const RainSensorDt = DeviceType(
  {
    name: "RainSensor",
    id: 68,
    category: "Sensor",
    classification: "simple",
    details: "This defines conformance to the Rain Sensor device type.",
    xref: { document: "device", section: "7.13" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 68, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.13.4" }
  }),
  Requirement(
    {
      name: "BooleanState",
      id: 69,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "7.13.4" }
    },
    Requirement({ name: "StateChange", conformance: "M", element: "event" })
  ),
  Requirement({
    name: "BooleanStateConfiguration",
    id: 128,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "7.13.4" }
  })
);
MatterDefinition.children.push(RainSensorDt);
export {
  RainSensorDt
};
