// @matter/model/dist/esm/standard/elements/RoboticVacuumCleanerDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const RoboticVacuumCleanerDt = DeviceType(
  {
    name: "RoboticVacuumCleaner",
    id: 116,
    category: "Robotic",
    classification: "simple",
    details: "This defines conformance for the Robotic Vacuum Cleaner device type.",
    xref: { document: "device", section: "12.1" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 116, revision: 3 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  Requirement({
    name: "RvcRunMode",
    id: 84,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  Requirement({
    name: "RvcCleanMode",
    id: 85,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  Requirement({
    name: "RvcOperationalState",
    id: 97,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  }),
  Requirement({
    name: "ServiceArea",
    id: 336,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "12.1.4" }
  })
);
MatterDefinition.children.push(RoboticVacuumCleanerDt);
export {
  RoboticVacuumCleanerDt
};
