// @matter/model/dist/esm/standard/elements/TemperatureSensorDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const TemperatureSensorDt = DeviceType(
  {
    name: "TemperatureSensor",
    id: 770,
    category: "Sensor",
    classification: "simple",
    details: "A Temperature Sensor device reports measurements of temperature.",
    xref: { document: "device", section: "7.4" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 770, revision: 2 }], element: "attribute" })
  ),
  Requirement({
    name: "TemperatureMeasurement",
    id: 1026,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.4.4" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "7.4.4" }
  })
);
MatterDefinition.children.push(TemperatureSensorDt);
export {
  TemperatureSensorDt
};
