// @matter/model/dist/esm/standard/elements/WaterValveDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const WaterValveDt = DeviceType(
  {
    name: "WaterValve",
    id: 66,
    category: "Smart Plugs/Outlets and other Actuators",
    classification: "simple",
    details: "This defines conformance to the Water Valve device type.",
    xref: { document: "device", section: "5.6" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 66, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.6.4" }
  }),
  Requirement({
    name: "ValveConfigurationAndControl",
    id: 129,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "5.6.4" }
  }),
  Requirement({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "5.6.4" }
  }),
  Requirement({
    name: "FlowMeasurement",
    id: 1028,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "5.6.4" }
  })
);
MatterDefinition.children.push(WaterValveDt);
export {
  WaterValveDt
};
