// @matter/model/dist/esm/standard/elements/WaterHeaterDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const WaterHeaterDt = DeviceType(
  {
    name: "WaterHeater",
    id: 1295,
    category: "Energy",
    classification: "simple",
    details: "A water heater is a device that is generally installed in properties to heat water for showers, baths etc.",
    xref: { document: "device", section: "14.2" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 1295, revision: 1 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "14.2.6" }
  }),
  Requirement(
    {
      name: "Thermostat",
      id: 513,
      conformance: "M",
      element: "serverCluster",
      xref: { document: "device", section: "14.2.6" }
    },
    Requirement({ name: "HEATING", conformance: "M", element: "feature" })
  ),
  Requirement({
    name: "WaterHeaterManagement",
    id: 148,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.2.6" }
  }),
  Requirement({
    name: "WaterHeaterMode",
    id: 158,
    conformance: "M",
    element: "serverCluster",
    xref: { document: "device", section: "14.2.6" }
  })
);
MatterDefinition.children.push(WaterHeaterDt);
export {
  WaterHeaterDt
};
