// @matter/model/dist/esm/standard/elements/WindowCoveringControllerDT.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MatterDefinition } from "../MatterDefinition.js";
import { DeviceTypeElement as DeviceType, RequirementElement as Requirement } from "../../elements/index.js";
const WindowCoveringControllerDt = DeviceType(
  {
    name: "WindowCoveringController",
    id: 515,
    category: "Closure",
    classification: "simple",
    details: "A Window Covering Controller is a device that controls an automatic window covering.",
    xref: { document: "device", section: "8.4" }
  },
  Requirement(
    { name: "Descriptor", id: 29, element: "serverCluster" },
    Requirement({ name: "DeviceTypeList", default: [{ deviceType: 515, revision: 3 }], element: "attribute" })
  ),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "serverCluster",
    xref: { document: "device", section: "8.4.4" }
  }),
  Requirement({
    name: "Identify",
    id: 3,
    conformance: "O",
    element: "clientCluster",
    xref: { document: "device", section: "8.4.4" }
  }),
  Requirement({
    name: "Groups",
    id: 4,
    conformance: "Active, O",
    element: "clientCluster",
    xref: { document: "device", section: "8.4.4" }
  }),
  Requirement({
    name: "WindowCovering",
    id: 258,
    conformance: "M",
    element: "clientCluster",
    xref: { document: "device", section: "8.4.4" }
  })
);
MatterDefinition.children.push(WindowCoveringControllerDt);
export {
  WindowCoveringControllerDt
};
