// @matter/model/dist/esm/common/Datatype.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Datatype = /* @__PURE__ */ ((Datatype2) => {
  Datatype2["bool"] = "bool";
  Datatype2["map8"] = "map8";
  Datatype2["map16"] = "map16";
  Datatype2["map32"] = "map32";
  Datatype2["map64"] = "map64";
  Datatype2["uint8"] = "uint8";
  Datatype2["uint16"] = "uint16";
  Datatype2["uint24"] = "uint24";
  Datatype2["uint32"] = "uint32";
  Datatype2["uint40"] = "uint40";
  Datatype2["uint48"] = "uint48";
  Datatype2["uint56"] = "uint56";
  Datatype2["uint64"] = "uint64";
  Datatype2["int8"] = "int8";
  Datatype2["int16"] = "int16";
  Datatype2["int24"] = "int24";
  Datatype2["int32"] = "int32";
  Datatype2["int40"] = "int40";
  Datatype2["int48"] = "int48";
  Datatype2["int56"] = "int56";
  Datatype2["int64"] = "int64";
  Datatype2["single"] = "single";
  Datatype2["double"] = "double";
  Datatype2["octstr"] = "octstr";
  Datatype2["list"] = "list";
  Datatype2["struct"] = "struct";
  Datatype2["string"] = "string";
  Datatype2["enum8"] = "enum8";
  Datatype2["enum16"] = "enum16";
  Datatype2["any"] = "any";
  return Datatype2;
})(Datatype || {});
export {
  Datatype
};
