// @matter/node/dist/esm/endpoints/ota-requestor.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  OtaSoftwareUpdateRequestorServer as BaseOtaSoftwareUpdateRequestorServer
} from "../behaviors/ota-software-update-requestor/OtaSoftwareUpdateRequestorServer.js";
import {
  OtaSoftwareUpdateProviderBehavior as BaseOtaSoftwareUpdateProviderBehavior
} from "../behaviors/ota-software-update-provider/OtaSoftwareUpdateProviderBehavior.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { DeviceClassification } from "#model";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var OtaRequestorRequirements;
((OtaRequestorRequirements2) => {
  OtaRequestorRequirements2.OtaSoftwareUpdateRequestorServer = BaseOtaSoftwareUpdateRequestorServer;
  OtaRequestorRequirements2.OtaSoftwareUpdateProviderBehavior = BaseOtaSoftwareUpdateProviderBehavior;
  OtaRequestorRequirements2.server = { mandatory: { OtaSoftwareUpdateRequestor: OtaRequestorRequirements2.OtaSoftwareUpdateRequestorServer } };
  OtaRequestorRequirements2.client = { mandatory: { OtaSoftwareUpdateProvider: OtaRequestorRequirements2.OtaSoftwareUpdateProviderBehavior } };
})(OtaRequestorRequirements || (OtaRequestorRequirements = {}));
const OtaRequestorEndpointDefinition = MutableEndpoint({
  name: "OtaRequestor",
  deviceType: 18,
  deviceRevision: 1,
  deviceClass: DeviceClassification.Utility,
  requirements: OtaRequestorRequirements,
  behaviors: SupportedBehaviors(OtaRequestorRequirements.server.mandatory.OtaSoftwareUpdateRequestor)
});
const OtaRequestorEndpoint = OtaRequestorEndpointDefinition;
export {
  OtaRequestorEndpoint,
  OtaRequestorEndpointDefinition,
  OtaRequestorRequirements
};
