// @matter/node/dist/esm/endpoints/power-source.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { PowerSourceServer as BasePowerSourceServer } from "../behaviors/power-source/PowerSourceServer.js";
import { MutableEndpoint } from "../endpoint/type/MutableEndpoint.js";
import { DeviceClassification } from "#model";
import { SupportedBehaviors } from "../endpoint/properties/SupportedBehaviors.js";
var PowerSourceRequirements;
((PowerSourceRequirements2) => {
  PowerSourceRequirements2.PowerSourceServer = BasePowerSourceServer;
  PowerSourceRequirements2.server = { mandatory: { PowerSource: PowerSourceRequirements2.PowerSourceServer } };
})(PowerSourceRequirements || (PowerSourceRequirements = {}));
const PowerSourceEndpointDefinition = MutableEndpoint({
  name: "PowerSource",
  deviceType: 17,
  deviceRevision: 1,
  deviceClass: DeviceClassification.Utility,
  requirements: PowerSourceRequirements,
  behaviors: SupportedBehaviors()
});
const PowerSourceEndpoint = PowerSourceEndpointDefinition;
export {
  PowerSourceEndpoint,
  PowerSourceEndpointDefinition,
  PowerSourceRequirements
};
