// @matter/node/dist/esm/node/storage/NodeStore.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { asyncNew, Construction, ImplementationError } from "#general";
import { EndpointStoreFactory } from "./EndpointStoreService.js";
class NodeStore {
  #factory;
  #rootStore;
  #construction;
  get construction() {
    return this.#construction;
  }
  constructor(factory) {
    this.#factory = factory;
    this.#construction = Construction(this);
  }
  toString() {
    return "node store";
  }
  [Construction.construct]() {
    return this.initializeStorage();
  }
  async erase() {
    await this.#rootStore?.erase();
  }
  get endpointStores() {
    if (this.#rootStore === void 0) {
      throw new ImplementationError("Endpoint storage accessed prior to initialization");
    }
    return this.#rootStore;
  }
  async initializeStorage() {
    this.#rootStore = await asyncNew(EndpointStoreFactory, {
      storage: this.factory.createContext("root")
    });
  }
  get factory() {
    return this.#factory;
  }
}
export {
  NodeStore
};
