// @matter/types/dist/cjs/protocol/messages/TlvDataReportForSend.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvDataReportForSend_exports = {};
__export(TlvDataReportForSend_exports, {
  TlvDataReportForSend: () => TlvDataReportForSend
});
module.exports = __toCommonJS(TlvDataReportForSend_exports);
var import_TlvAny = require("../../tlv/TlvAny.js");
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvDataReportForSend = (0, import_TlvObject.TlvObject)({
  /** A SubscriptionId only used in a Subscribe interaction. */
  subscriptionId: (0, import_TlvObject.TlvOptionalField)(0, import_TlvNumber.TlvUInt32),
  /** A list of zero or more attribute data reports, */
  attributeReports: (0, import_TlvObject.TlvOptionalField)(1, (0, import_TlvArray.TlvArray)(import_TlvAny.TlvAny)),
  /** A list of zero or more event reports. */
  eventReports: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvArray.TlvArray)(import_TlvAny.TlvAny)),
  moreChunkedMessages: (0, import_TlvObject.TlvOptionalField)(3, import_TlvBoolean.TlvBoolean),
  /** Do not send a response to this action. */
  suppressResponse: (0, import_TlvObject.TlvOptionalField)(4, import_TlvBoolean.TlvBoolean),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
