// @matter/types/dist/cjs/protocol/messages/TlvReadRequest.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvReadRequest_exports = {};
__export(TlvReadRequest_exports, {
  TlvReadRequest: () => TlvReadRequest
});
module.exports = __toCommonJS(TlvReadRequest_exports);
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvAttributePath = require("../types/TlvAttributePath.js");
var import_TlvDataVersionFilter = require("../types/TlvDataVersionFilter.js");
var import_TlvEventFilter = require("../types/TlvEventFilter.js");
var import_TlvEventPath = require("../types/TlvEventPath.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvReadRequest = (0, import_TlvObject.TlvObject)({
  /** A list of zero or more request paths to cluster attribute data. */
  attributeRequests: (0, import_TlvObject.TlvOptionalField)(0, (0, import_TlvArray.TlvArray)(import_TlvAttributePath.TlvAttributePath)),
  /** A list of zero or more request paths to cluster events. */
  eventRequests: (0, import_TlvObject.TlvOptionalField)(1, (0, import_TlvArray.TlvArray)(import_TlvEventPath.TlvEventPath)),
  /** A list of zero or more minimum event numbers per specific node. */
  eventFilters: (0, import_TlvObject.TlvOptionalField)(2, (0, import_TlvArray.TlvArray)(import_TlvEventFilter.TlvEventFilter)),
  /** Limits the data read within fabric-scoped lists to the accessing fabric. */
  isFabricFiltered: (0, import_TlvObject.TlvField)(3, import_TlvBoolean.TlvBoolean),
  /** A list of zero or more cluster instance data versions. */
  dataVersionFilters: (0, import_TlvObject.TlvOptionalField)(4, (0, import_TlvArray.TlvArray)(import_TlvDataVersionFilter.TlvDataVersionFilter)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
