// @matter/types/dist/cjs/protocol/messages/TlvSubscribeRequest.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvSubscribeRequest_exports = {};
__export(TlvSubscribeRequest_exports, {
  TlvSubscribeRequest: () => TlvSubscribeRequest
});
module.exports = __toCommonJS(TlvSubscribeRequest_exports);
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvAttributePath = require("../types/TlvAttributePath.js");
var import_TlvDataVersionFilter = require("../types/TlvDataVersionFilter.js");
var import_TlvEventFilter = require("../types/TlvEventFilter.js");
var import_TlvEventPath = require("../types/TlvEventPath.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvSubscribeRequest = (0, import_TlvObject.TlvObject)({
  /** False to terminate existing subscriptions from initiator. */
  keepSubscriptions: (0, import_TlvObject.TlvField)(0, import_TlvBoolean.TlvBoolean),
  /** The requested minimum interval boundary floor in seconds. */
  minIntervalFloorSeconds: (0, import_TlvObject.TlvField)(1, import_TlvNumber.TlvUInt16),
  /** The requested maximum interval boundary ceiling in seconds. */
  maxIntervalCeilingSeconds: (0, import_TlvObject.TlvField)(2, import_TlvNumber.TlvUInt16),
  /** A list of zero or more request paths to cluster attribute data. */
  attributeRequests: (0, import_TlvObject.TlvOptionalField)(3, (0, import_TlvArray.TlvArray)(import_TlvAttributePath.TlvAttributePath)),
  /** A list of zero or more request paths to cluster events. */
  eventRequests: (0, import_TlvObject.TlvOptionalField)(4, (0, import_TlvArray.TlvArray)(import_TlvEventPath.TlvEventPath)),
  /** A list of zero or more minimum event numbers per specific node. */
  eventFilters: (0, import_TlvObject.TlvOptionalField)(5, (0, import_TlvArray.TlvArray)(import_TlvEventFilter.TlvEventFilter)),
  /** Limits the data read within fabric-scoped lists to the accessing fabric. */
  isFabricFiltered: (0, import_TlvObject.TlvField)(7, import_TlvBoolean.TlvBoolean),
  /** A list of zero or more cluster instance data versions. */
  dataVersionFilters: (0, import_TlvObject.TlvOptionalField)(8, (0, import_TlvArray.TlvArray)(import_TlvDataVersionFilter.TlvDataVersionFilter)),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
