// @matter/types/dist/cjs/protocol/messages/TlvWriteRequest.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvWriteRequest_exports = {};
__export(TlvWriteRequest_exports, {
  TlvWriteRequest: () => TlvWriteRequest
});
module.exports = __toCommonJS(TlvWriteRequest_exports);
var import_TlvArray = require("../../tlv/TlvArray.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvAttributeData = require("../types/TlvAttributeData.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvWriteRequest = (0, import_TlvObject.TlvObject)({
  /** Do not send a response to this action. */
  suppressResponse: (0, import_TlvObject.TlvOptionalField)(0, import_TlvBoolean.TlvBoolean),
  /** Flag action as part of a timed write transaction. */
  timedRequest: (0, import_TlvObject.TlvField)(1, import_TlvBoolean.TlvBoolean),
  /** A list of one or more path and data tuples. */
  writeRequests: (0, import_TlvObject.TlvField)(2, (0, import_TlvArray.TlvArray)(import_TlvAttributeData.TlvAttributeData)),
  moreChunkedMessages: (0, import_TlvObject.TlvOptionalField)(3, import_TlvBoolean.TlvBoolean),
  interactionModelRevision: (0, import_TlvObject.TlvField)(255, import_TlvNumber.TlvUInt8)
});
