// @matter/node/dist/cjs/node/storage/NodeStore.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NodeStore_exports = {};
__export(NodeStore_exports, {
  NodeStore: () => NodeStore
});
module.exports = __toCommonJS(NodeStore_exports);
var import_general = require("#general");
var import_EndpointStoreService = require("./EndpointStoreService.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class NodeStore {
  #factory;
  #rootStore;
  #construction;
  get construction() {
    return this.#construction;
  }
  constructor(factory) {
    this.#factory = factory;
    this.#construction = (0, import_general.Construction)(this);
  }
  toString() {
    return "node store";
  }
  [import_general.Construction.construct]() {
    return this.initializeStorage();
  }
  async erase() {
    await this.#rootStore?.erase();
  }
  get endpointStores() {
    if (this.#rootStore === void 0) {
      throw new import_general.ImplementationError("Endpoint storage accessed prior to initialization");
    }
    return this.#rootStore;
  }
  async initializeStorage() {
    this.#rootStore = await (0, import_general.asyncNew)(import_EndpointStoreService.EndpointStoreFactory, {
      storage: this.factory.createContext("root")
    });
  }
  get factory() {
    return this.#factory;
  }
}
