// @matter/protocol/dist/cjs/certificate/CertificationDeclarationManager.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CertificationDeclarationManager_exports = {};
__export(CertificationDeclarationManager_exports, {
  CertificationDeclarationManager: () => CertificationDeclarationManager
});
module.exports = __toCommonJS(CertificationDeclarationManager_exports);
var import_general = require("#general");
var import_CertificateManager = require("./CertificateManager.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TestCMS_SignerPrivateKey = import_general.Bytes.fromHex("AEF3484116E9481EC57BE0472DF41BF499064E5024AD869ECA5E889802D48075");
const TestCMS_SignerSubjectKeyIdentifier = import_general.Bytes.fromHex("62FA823359ACFAA9963E1CFA140ADDF504F37160");
class CertificationDeclarationManager {
  static generate(vendorId, productId, provisional = false) {
    const certificationElements = import_CertificateManager.TlvCertificationDeclaration.encode({
      formatVersion: 1,
      vendorId,
      produceIdArray: [productId],
      deviceTypeId: 22,
      certificateId: "CSA00000SWC00000-00",
      securityLevel: 0,
      securityInformation: 0,
      versionNumber: 1,
      certificationType: provisional ? 1 : 0
      // 0 = Test, 1 = Provisional/In certification, 2 = official
    });
    return import_CertificateManager.CertificateManager.certificationDeclarationToAsn1(
      certificationElements,
      TestCMS_SignerSubjectKeyIdentifier,
      (0, import_general.PrivateKey)(TestCMS_SignerPrivateKey)
    );
  }
}
