// @matter/types/dist/cjs/tlv/TlvVoid.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvVoid_exports = {};
__export(TlvVoid_exports, {
  TlvVoid: () => TlvVoid,
  VoidSchema: () => VoidSchema
});
module.exports = __toCommonJS(TlvVoid_exports);
var import_general = require("#general");
var import_ValidationError = require("../common/ValidationError.js");
var import_TlvSchema = require("./TlvSchema.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class VoidSchema extends import_TlvSchema.TlvSchema {
  encodeTlvInternal(_writer, _value, _tag) {
  }
  decodeTlv(_encoded) {
  }
  decodeTlvInternalValue(_reader, _typeLength) {
    throw new import_general.InternalError("decodeTlvInternalValue should never be called");
  }
  validate(data) {
    if (data !== void 0) throw new import_ValidationError.ValidationDatatypeMismatchError(`Expected void, got ${typeof data}.`);
  }
}
const TlvVoid = new VoidSchema();
