// @matter/types/dist/esm/globals/MeasurementType.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var MeasurementType = /* @__PURE__ */ ((MeasurementType2) => {
  MeasurementType2[MeasurementType2["Unspecified"] = 0] = "Unspecified";
  MeasurementType2[MeasurementType2["Voltage"] = 1] = "Voltage";
  MeasurementType2[MeasurementType2["ActiveCurrent"] = 2] = "ActiveCurrent";
  MeasurementType2[MeasurementType2["ReactiveCurrent"] = 3] = "ReactiveCurrent";
  MeasurementType2[MeasurementType2["ApparentCurrent"] = 4] = "ApparentCurrent";
  MeasurementType2[MeasurementType2["ActivePower"] = 5] = "ActivePower";
  MeasurementType2[MeasurementType2["ReactivePower"] = 6] = "ReactivePower";
  MeasurementType2[MeasurementType2["ApparentPower"] = 7] = "ApparentPower";
  MeasurementType2[MeasurementType2["RmsVoltage"] = 8] = "RmsVoltage";
  MeasurementType2[MeasurementType2["RmsCurrent"] = 9] = "RmsCurrent";
  MeasurementType2[MeasurementType2["RmsPower"] = 10] = "RmsPower";
  MeasurementType2[MeasurementType2["Frequency"] = 11] = "Frequency";
  MeasurementType2[MeasurementType2["PowerFactor"] = 12] = "PowerFactor";
  MeasurementType2[MeasurementType2["NeutralCurrent"] = 13] = "NeutralCurrent";
  MeasurementType2[MeasurementType2["ElectricalEnergy"] = 14] = "ElectricalEnergy";
  return MeasurementType2;
})(MeasurementType || {});
export {
  MeasurementType
};
