// @matter/types/dist/esm/globals/Status.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var Status = /* @__PURE__ */ ((Status2) => {
  Status2[Status2["Success"] = 0] = "Success";
  Status2[Status2["Failure"] = 1] = "Failure";
  Status2[Status2["InvalidSubscription"] = 125] = "InvalidSubscription";
  Status2[Status2["UnsupportedAccess"] = 126] = "UnsupportedAccess";
  Status2[Status2["UnsupportedEndpoint"] = 127] = "UnsupportedEndpoint";
  Status2[Status2["InvalidAction"] = 128] = "InvalidAction";
  Status2[Status2["UnsupportedCommand"] = 129] = "UnsupportedCommand";
  Status2[Status2["InvalidCommand"] = 133] = "InvalidCommand";
  Status2[Status2["UnsupportedAttribute"] = 134] = "UnsupportedAttribute";
  Status2[Status2["ConstraintError"] = 135] = "ConstraintError";
  Status2[Status2["UnsupportedWrite"] = 136] = "UnsupportedWrite";
  Status2[Status2["ResourceExhausted"] = 137] = "ResourceExhausted";
  Status2[Status2["NotFound"] = 139] = "NotFound";
  Status2[Status2["UnreportableAttribute"] = 140] = "UnreportableAttribute";
  Status2[Status2["InvalidDataType"] = 141] = "InvalidDataType";
  Status2[Status2["UnsupportedRead"] = 143] = "UnsupportedRead";
  Status2[Status2["DataVersionMismatch"] = 146] = "DataVersionMismatch";
  Status2[Status2["Timeout"] = 148] = "Timeout";
  Status2[Status2["UnsupportedNode"] = 155] = "UnsupportedNode";
  Status2[Status2["Busy"] = 156] = "Busy";
  Status2[Status2["AccessRestricted"] = 157] = "AccessRestricted";
  Status2[Status2["UnsupportedCluster"] = 195] = "UnsupportedCluster";
  Status2[Status2["NoUpstreamSubscription"] = 197] = "NoUpstreamSubscription";
  Status2[Status2["NeedsTimedInteraction"] = 198] = "NeedsTimedInteraction";
  Status2[Status2["UnsupportedEvent"] = 199] = "UnsupportedEvent";
  Status2[Status2["PathsExhausted"] = 200] = "PathsExhausted";
  Status2[Status2["TimedRequestMismatch"] = 201] = "TimedRequestMismatch";
  Status2[Status2["FailsafeRequired"] = 202] = "FailsafeRequired";
  Status2[Status2["InvalidInState"] = 203] = "InvalidInState";
  Status2[Status2["NoCommandResponse"] = 204] = "NoCommandResponse";
  Status2[Status2["TermsAndConditionsChanged"] = 205] = "TermsAndConditionsChanged";
  Status2[Status2["MaintenanceRequired"] = 206] = "MaintenanceRequired";
  return Status2;
})(Status || {});
export {
  Status
};
