// @matter/node/dist/cjs/tags/AreaNamespaceTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AreaNamespaceTag_exports = {};
__export(AreaNamespaceTag_exports, {
  AreaNamespaceTag: () => AreaNamespaceTag
});
module.exports = __toCommonJS(AreaNamespaceTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const AreaNamespaceTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 16,
  tags: {
    Aisle: { id: 0, label: "Aisle" },
    Attic: { id: 1, label: "Attic" },
    BackDoor: { id: 2, label: "Back Door" },
    BackYard: { id: 3, label: "Back Yard" },
    Balcony: { id: 4, label: "Balcony" },
    Ballroom: { id: 5, label: "Ballroom" },
    /**
     * Also known as Restroom
     */
    Bathroom: { id: 6, label: "Bathroom" },
    Bedroom: { id: 7, label: "Bedroom" },
    Border: { id: 8, label: "Border" },
    /**
     * A small room typically used for storage
     */
    Boxroom: { id: 9, label: "Boxroom" },
    BreakfastRoom: { id: 10, label: "Breakfast Room" },
    Carport: { id: 11, label: "Carport" },
    Cellar: { id: 12, label: "Cellar" },
    Cloakroom: { id: 13, label: "Cloakroom" },
    Closet: { id: 14, label: "Closet" },
    Conservatory: { id: 15, label: "Conservatory" },
    Corridor: { id: 16, label: "Corridor" },
    CraftRoom: { id: 17, label: "Craft Room" },
    Cupboard: { id: 18, label: "Cupboard" },
    Deck: { id: 19, label: "Deck" },
    /**
     * A small, comfortable room for individual activities such as work or hobbies
     */
    Den: { id: 20, label: "Den" },
    Dining: { id: 21, label: "Dining" },
    DrawingRoom: { id: 22, label: "Drawing Room" },
    DressingRoom: { id: 23, label: "Dressing Room" },
    Driveway: { id: 24, label: "Driveway" },
    Elevator: { id: 25, label: "Elevator" },
    /**
     * A bathroom directly accessible from a bedroom
     */
    Ensuite: { id: 26, label: "Ensuite" },
    Entrance: { id: 27, label: "Entrance" },
    Entryway: { id: 28, label: "Entryway" },
    FamilyRoom: { id: 29, label: "Family Room" },
    Foyer: { id: 30, label: "Foyer" },
    FrontDoor: { id: 31, label: "Front Door" },
    FrontYard: { id: 32, label: "Front Yard" },
    GameRoom: { id: 33, label: "Game Room" },
    Garage: { id: 34, label: "Garage" },
    GarageDoor: { id: 35, label: "Garage Door" },
    Garden: { id: 36, label: "Garden" },
    GardenDoor: { id: 37, label: "Garden Door" },
    /**
     * Also known as Guest Restroom
     */
    GuestBathroom: { id: 38, label: "Guest Bathroom" },
    GuestBedroom: { id: 39, label: "Guest Bedroom" },
    GuestRoom: { id: 40, label: "Guest Room" },
    Gym: { id: 41, label: "Gym" },
    Hallway: { id: 42, label: "Hallway" },
    /**
     * A cozy room containing a fireplace or other point heat source
     */
    HearthRoom: { id: 43, label: "Hearth Room" },
    KidsRoom: { id: 44, label: "Kids Room" },
    KidsBedroom: { id: 45, label: "Kids Bedroom" },
    Kitchen: { id: 46, label: "Kitchen" },
    LaundryRoom: { id: 47, label: "Laundry Room" },
    Lawn: { id: 48, label: "Lawn" },
    Library: { id: 49, label: "Library" },
    LivingRoom: { id: 50, label: "Living Room" },
    Lounge: { id: 51, label: "Lounge" },
    MediaTvRoom: { id: 52, label: "Media/TV Room" },
    /**
     * A space used to remove soiled garments prior to entering the domicile proper
     */
    MudRoom: { id: 53, label: "Mud Room" },
    MusicRoom: { id: 54, label: "Music Room" },
    Nursery: { id: 55, label: "Nursery" },
    Office: { id: 56, label: "Office" },
    OutdoorKitchen: { id: 57, label: "Outdoor Kitchen" },
    Outside: { id: 58, label: "Outside" },
    /**
     * AKA a larder, a place where food is stored
     */
    Pantry: { id: 59, label: "Pantry" },
    ParkingLot: { id: 60, label: "Parking Lot" },
    Parlor: { id: 61, label: "Parlor" },
    Patio: { id: 62, label: "Patio" },
    PlayRoom: { id: 63, label: "Play Room" },
    PoolRoom: { id: 64, label: "Pool Room" },
    Porch: { id: 65, label: "Porch" },
    PrimaryBathroom: { id: 66, label: "Primary Bathroom" },
    PrimaryBedroom: { id: 67, label: "Primary Bedroom" },
    Ramp: { id: 68, label: "Ramp" },
    ReceptionRoom: { id: 69, label: "Reception Room" },
    RecreationRoom: { id: 70, label: "Recreation Room" },
    Roof: { id: 71, label: "Roof" },
    Sauna: { id: 72, label: "Sauna" },
    /**
     * A utility space for cleaning dishes and laundry
     */
    Scullery: { id: 73, label: "Scullery" },
    SewingRoom: { id: 74, label: "Sewing Room" },
    Shed: { id: 75, label: "Shed" },
    SideDoor: { id: 76, label: "Side Door" },
    SideYard: { id: 77, label: "Side Yard" },
    SittingRoom: { id: 78, label: "Sitting Room" },
    /**
     * An informal space meant to be 'cozy', 'snug', relaxed, meant to share with family or friends
     */
    Snug: { id: 79, label: "Snug" },
    Spa: { id: 80, label: "Spa" },
    Staircase: { id: 81, label: "Staircase" },
    SteamRoom: { id: 82, label: "Steam Room" },
    StorageRoom: { id: 83, label: "Storage Room" },
    Studio: { id: 84, label: "Studio" },
    Study: { id: 85, label: "Study" },
    SunRoom: { id: 86, label: "Sun Room" },
    SwimmingPool: { id: 87, label: "Swimming Pool" },
    Terrace: { id: 88, label: "Terrace" },
    /**
     * A room dedicated to a toilet; a water closet / WC
     */
    Toilet: { id: 89, label: "Toilet" },
    UtilityRoom: { id: 90, label: "Utility Room" },
    Ward: { id: 91, label: "Ward" },
    Workshop: { id: 92, label: "Workshop" }
  }
});
