// @matter/node/dist/cjs/tags/ClosureTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ClosureTag_exports = {};
__export(ClosureTag_exports, {
  ClosureTag: () => ClosureTag
});
module.exports = __toCommonJS(ClosureTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ClosureTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 1,
  tags: {
    /**
     * Move toward open position
     */
    Opening: { id: 0, label: "Opening" },
    /**
     * Move toward closed position
     */
    Closing: { id: 1, label: "Closing" },
    /**
     * Stop any movement
     */
    Stop: { id: 2, label: "Stop" }
  }
});
