// @matter/node/dist/cjs/tags/CompassLocationTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CompassLocationTag_exports = {};
__export(CompassLocationTag_exports, {
  CompassLocationTag: () => CompassLocationTag
});
module.exports = __toCommonJS(CompassLocationTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const CompassLocationTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 3,
  tags: {
    North: { id: 0, label: "North" },
    NorthEast: { id: 1, label: "North-East" },
    East: { id: 2, label: "East" },
    SouthEast: { id: 3, label: "South-East" },
    South: { id: 4, label: "South" },
    SouthWest: { id: 5, label: "South-West" },
    West: { id: 6, label: "West" },
    NorthWest: { id: 7, label: "North-West" }
  }
});
