// @matter/node/dist/cjs/tags/CompassDirectionTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CompassDirectionTag_exports = {};
__export(CompassDirectionTag_exports, {
  CompassDirectionTag: () => CompassDirectionTag
});
module.exports = __toCommonJS(CompassDirectionTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const CompassDirectionTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 2,
  tags: {
    Northward: { id: 0, label: "Northward" },
    NorthEastward: { id: 1, label: "North-Eastward" },
    Eastward: { id: 2, label: "Eastward" },
    SouthEastward: { id: 3, label: "South-Eastward" },
    Southward: { id: 4, label: "Southward" },
    SouthWestward: { id: 5, label: "South-Westward" },
    Westward: { id: 6, label: "Westward" },
    NorthWestward: { id: 7, label: "North-Westward" }
  }
});
