// @matter/node/dist/cjs/tags/DirectionTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DirectionTag_exports = {};
__export(DirectionTag_exports, {
  DirectionTag: () => DirectionTag
});
module.exports = __toCommonJS(DirectionTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const DirectionTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 4,
  tags: {
    Upward: { id: 0, label: "Upward" },
    Downward: { id: 1, label: "Downward" },
    Leftward: { id: 2, label: "Leftward" },
    Rightward: { id: 3, label: "Rightward" },
    Forward: { id: 4, label: "Forward" },
    Backward: { id: 5, label: "Backward" }
  }
});
