// @matter/node/dist/cjs/tags/ElectricalMeasurementTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ElectricalMeasurementTag_exports = {};
__export(ElectricalMeasurementTag_exports, {
  ElectricalMeasurementTag: () => ElectricalMeasurementTag
});
module.exports = __toCommonJS(ElectricalMeasurementTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const ElectricalMeasurementTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 10,
  tags: {
    /**
     * Indicates values measured for a DC load
     */
    Dc: { id: 0, label: "DC" },
    /**
     * Indicates values measured for a single-phase AC load, or values measured for the collective load on a
     * polyphase AC power supply
     */
    Ac: { id: 1, label: "AC" },
    /**
     * Indicates values measured for an AC load on phase 1 of a polyphase power supply
     */
    AcPhase1: { id: 2, label: "ACPhase1" },
    /**
     * Indicates values measured for an AC load on phase 2 of a polyphase power supply
     */
    AcPhase2: { id: 3, label: "ACPhase2" },
    /**
     * Indicates values measured for an AC load on phase 3 of a polyphase power supply
     */
    AcPhase3: { id: 4, label: "ACPhase3" }
  }
});
