// @matter/node/dist/cjs/tags/LandmarkNamespaceTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LandmarkNamespaceTag_exports = {};
__export(LandmarkNamespaceTag_exports, {
  LandmarkNamespaceTag: () => LandmarkNamespaceTag
});
module.exports = __toCommonJS(LandmarkNamespaceTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LandmarkNamespaceTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 17,
  tags: {
    AirConditioner: { id: 0, label: "Air Conditioner" },
    AirPurifier: { id: 1, label: "Air Purifier" },
    BackDoor: { id: 2, label: "Back Door" },
    BarStool: { id: 3, label: "Bar Stool" },
    BathMat: { id: 4, label: "Bath Mat" },
    Bathtub: { id: 5, label: "Bathtub" },
    Bed: { id: 6, label: "Bed" },
    Bookshelf: { id: 7, label: "Bookshelf" },
    Chair: { id: 8, label: "Chair" },
    ChristmasTree: { id: 9, label: "Christmas Tree" },
    CoatRack: { id: 10, label: "Coat Rack" },
    CoffeeTable: { id: 11, label: "Coffee Table" },
    CookingRange: { id: 12, label: "Cooking Range" },
    Couch: { id: 13, label: "Couch" },
    Countertop: { id: 14, label: "Countertop" },
    Cradle: { id: 15, label: "Cradle" },
    Crib: { id: 16, label: "Crib" },
    Desk: { id: 17, label: "Desk" },
    DiningTable: { id: 18, label: "Dining Table" },
    Dishwasher: { id: 19, label: "Dishwasher" },
    Door: { id: 20, label: "Door" },
    Dresser: { id: 21, label: "Dresser" },
    LaundryDryer: { id: 22, label: "Laundry Dryer" },
    Fan: { id: 23, label: "Fan" },
    Fireplace: { id: 24, label: "Fireplace" },
    Freezer: { id: 25, label: "Freezer" },
    FrontDoor: { id: 26, label: "Front Door" },
    HighChair: { id: 27, label: "High Chair" },
    KitchenIsland: { id: 28, label: "Kitchen Island" },
    Lamp: { id: 29, label: "Lamp" },
    LitterBox: { id: 30, label: "Litter Box" },
    Mirror: { id: 31, label: "Mirror" },
    Nightstand: { id: 32, label: "Nightstand" },
    Oven: { id: 33, label: "Oven" },
    PetBed: { id: 34, label: "Pet Bed" },
    PetBowl: { id: 35, label: "Pet Bowl" },
    PetCrate: { id: 36, label: "Pet Crate" },
    Refrigerator: { id: 37, label: "Refrigerator" },
    ScratchingPost: { id: 38, label: "Scratching Post" },
    ShoeRack: { id: 39, label: "Shoe Rack" },
    Shower: { id: 40, label: "Shower" },
    SideDoor: { id: 41, label: "Side Door" },
    Sink: { id: 42, label: "Sink" },
    Sofa: { id: 43, label: "Sofa" },
    Stove: { id: 44, label: "Stove" },
    Table: { id: 45, label: "Table" },
    Toilet: { id: 46, label: "Toilet" },
    TrashCan: { id: 47, label: "Trash Can" },
    LaundryWasher: { id: 48, label: "Laundry Washer" },
    Window: { id: 49, label: "Window" },
    WineCooler: { id: 50, label: "Wine Cooler" }
  }
});
