// @matter/node/dist/cjs/tags/LocationTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LocationTag_exports = {};
__export(LocationTag_exports, {
  LocationTag: () => LocationTag
});
module.exports = __toCommonJS(LocationTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const LocationTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 6,
  tags: {
    /**
     * Element is indoors or related to indoor equipment/conditions (e.g. the "indoor" temperature).
     */
    Indoor: { id: 0, label: "Indoor" },
    /**
     * Element is outdoors or related to outdoor equipment/conditions (e.g. the "outdoor" temperature).
     */
    Outdoor: { id: 1, label: "Outdoor" },
    /**
     * Element is located inside the equipment (e.g. a sensor "inside" a cabinet).
     */
    Inside: { id: 2, label: "Inside" },
    /**
     * Element is located outside the equipment (e.g. a sensor "outside" a cabinet)
     */
    Outside: { id: 3, label: "Outside" }
  }
});
