// @matter/node/dist/cjs/tags/NumberTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var NumberTag_exports = {};
__export(NumberTag_exports, {
  NumberTag: () => NumberTag
});
module.exports = __toCommonJS(NumberTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const NumberTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 7,
  tags: {
    Zero: { id: 0, label: "Zero" },
    One: { id: 1, label: "One" },
    Two: { id: 2, label: "Two" },
    Three: { id: 3, label: "Three" },
    Four: { id: 4, label: "Four" },
    Five: { id: 5, label: "Five" },
    Six: { id: 6, label: "Six" },
    Seven: { id: 7, label: "Seven" },
    Eight: { id: 8, label: "Eight" },
    Nine: { id: 9, label: "Nine" },
    Ten: { id: 10, label: "Ten" }
  }
});
