// @matter/node/dist/cjs/tags/PositionTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PositionTag_exports = {};
__export(PositionTag_exports, {
  PositionTag: () => PositionTag
});
module.exports = __toCommonJS(PositionTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PositionTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 8,
  tags: {
    Left: { id: 0, label: "Left" },
    Right: { id: 1, label: "Right" },
    Top: { id: 2, label: "Top" },
    Bottom: { id: 3, label: "Bottom" },
    Middle: { id: 4, label: "Middle" },
    /**
     * Numeric value provided in Label field
     */
    Row: { id: 5, label: "Row" },
    /**
     * Numeric value provided in Label field
     */
    Column: { id: 6, label: "Column" }
  }
});
