// @matter/node/dist/cjs/tags/PowerSourceTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PowerSourceTag_exports = {};
__export(PowerSourceTag_exports, {
  PowerSourceTag: () => PowerSourceTag
});
module.exports = __toCommonJS(PowerSourceTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const PowerSourceTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 15,
  tags: {
    /**
     * The Power Source cluster is related to power provided from an unknown source
     */
    Unknown: { id: 0, label: "Unknown" },
    /**
     * The Power Source cluster is related to power provided from the electrical grid
     *
     * Power Source clusters with this tag shall implement the WIRED feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.1
     */
    Grid: { id: 1, label: "Grid" },
    /**
     * The Power Source cluster is related to power provided from a solar panel array
     *
     * Power Source clusters with this tag shall implement the WIRED feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.2
     */
    Solar: { id: 2, label: "Solar" },
    /**
     * The Power Source cluster is related to power provided from a battery
     *
     * Power Source clusters with this tag shall implement the BAT feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.3
     */
    Battery: { id: 3, label: "Battery" },
    /**
     * The Power Source cluster is related to power provided from an electric vehicle
     *
     * Power Source clusters with this tag shall implement the BAT feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.4
     */
    Ev: { id: 4, label: "EV" }
  }
});
