// @matter/node/dist/cjs/tags/SwitchesTag.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var SwitchesTag_exports = {};
__export(SwitchesTag_exports, {
  SwitchesTag: () => SwitchesTag
});
module.exports = __toCommonJS(SwitchesTag_exports);
var import_SemanticNamespace = require("../endpoint/type/SemanticNamespace.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const SwitchesTag = (0, import_SemanticNamespace.SemanticNamespace)({
  id: 67,
  tags: {
    On: { id: 0, label: "On" },
    Off: { id: 1, label: "Off" },
    Toggle: { id: 2, label: "Toggle" },
    /**
     * e.g. dim up (light)
     */
    Up: { id: 3, label: "Up" },
    /**
     * e.g. dim down (light)
     */
    Down: { id: 4, label: "Down" },
    /**
     * e.g. select next scene
     */
    Next: { id: 5, label: "Next" },
    /**
     * e.g. select previous scene
     */
    Previous: { id: 6, label: "Previous" },
    EnterOkSelect: { id: 7, label: "Enter/OK/Select" },
    /**
     * Textual description provided in Label field
     *
     * When this value is used, the Label field in the same Semantic Tag structure shall be filled with a textual
     * description of the function indicated on the button, such as a label or icon printed on the button, e.g.
     * "dining".
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 18.1
     */
    Custom: { id: 8, label: "Custom" }
  }
});
