// @matter/types/dist/esm/cluster/ClusterType.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ClusterId } from "../datatype/ClusterId.js";
import {
  GlobalAttributes
} from "./Cluster.js";
function ClusterType(options) {
  const cluster = {
    id: ClusterId(options.id),
    name: options.name,
    revision: options.revision,
    features: options.features ?? {},
    supportedFeatures: options.supportedFeatures ?? {},
    attributes: {
      ...options.attributes,
      ...GlobalAttributes(options.features ?? {})
    },
    commands: options.commands ?? {},
    events: options.events ?? {},
    unknown: false
  };
  if (options.base) {
    cluster.base = options.base;
  }
  if (options.extensions) {
    cluster.extensions = options.extensions;
  }
  return cluster;
}
((ClusterType2) => {
  ClusterType2.Unknown = ClusterType2({
    id: 0,
    revision: 0,
    name: "Unknown",
    attributes: {},
    commands: {},
    events: {}
  });
})(ClusterType || (ClusterType = {}));
export {
  ClusterType
};
