// @matter/model/dist/cjs/aspects/Aspect.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Aspect_exports = {};
__export(Aspect_exports, {
  Aspect: () => Aspect
});
module.exports = __toCommonJS(Aspect_exports);
var import_general = require("@matter/general");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class Aspect {
  definition;
  get valid() {
    return !this.errors;
  }
  constructor(definition) {
    this.definition = definition;
  }
  get empty() {
    for (const [k, v] of Object.entries(this)) {
      if (k !== "definition" && v !== void 0) {
        return false;
      }
    }
    return true;
  }
  /**
   * Test for logical equivalence.
   */
  equals(other) {
    if (!(other instanceof this.constructor)) {
      return false;
    }
    return (0, import_general.isDeepEqual)(this.valueOf(), other.valueOf());
  }
  valueOf() {
    return this.toString();
  }
  error(code, message) {
    if (!this.errors) {
      this.errors = [];
    }
    this.errors.push({
      code,
      source: `${this.constructor.name} ${(0, import_general.serialize)(this.definition)}`,
      message
    });
  }
  extend(other) {
    const descriptors = [
      ...Object.entries(Object.getOwnPropertyDescriptors(this)),
      ...Object.entries(Object.getOwnPropertyDescriptors(other))
    ];
    const definition = {};
    for (const [name, descriptor] of descriptors) {
      if (name === "definition" || name === "errors" || descriptor.value === void 0) {
        continue;
      }
      definition[name] = descriptor.value;
    }
    const constructor = this.constructor;
    return new constructor(definition);
  }
  freeze() {
    Object.freeze(this);
  }
}
