// @matter/general/dist/cjs/util/PromiseQueue.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PromiseQueue_exports = {};
__export(PromiseQueue_exports, {
  PromiseQueue: () => PromiseQueue
});
module.exports = __toCommonJS(PromiseQueue_exports);
var import_Time = require("#time/Time.js");
var import_Logger = require("../log/Logger.js");
var import_Promises = require("./Promises.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const logger = import_Logger.Logger.get("PromiseQueue");
class PromiseQueue {
  #delay;
  #queue = new Array();
  #concurrency;
  #runningCount = 0;
  constructor(concurrency = 1, delay = 0) {
    this.#concurrency = concurrency;
    this.#delay = delay;
  }
  /**
   * Add a promise to the queue. It returns a promise that can be awaited.
   */
  add(executor) {
    const { promise, resolver, rejecter } = (0, import_Promises.createPromise)();
    logger.debug("Add promise to queue on place", this.#queue.length + 1);
    this.#queue.push({
      func: () => executor().then(resolver).catch(rejecter),
      rejecter
    });
    this.#run();
    return promise;
  }
  /**
   * Run the next promise in the queue.
   */
  #run() {
    if (this.#runningCount >= this.#concurrency || this.#queue.length === 0) {
      return;
    }
    logger.debug(
      "Processing promise from queue ... Current queue length:",
      this.#queue.length,
      "Already running:",
      this.#runningCount
    );
    const { func } = this.#queue.shift() ?? {};
    if (func !== void 0) {
      this.#runningCount++;
      func().catch((error) => {
        throw error;
      }).finally(() => {
        logger.debug("Promise processed ... Still running:", this.#runningCount - 1);
        if (this.#delay > 0) {
          import_Time.Time.sleep("Queue delay", this.#delay).then(() => this.#runNext()).catch((error) => {
            throw error;
          });
        } else {
          this.#runNext();
        }
      });
    }
  }
  #runNext() {
    this.#runningCount--;
    this.#run();
  }
  /**
   * Clear the queue.
   */
  clear(reject) {
    if (reject) {
      for (const { rejecter } of this.#queue) {
        rejecter();
      }
    }
    this.#queue.length = 0;
  }
  /** Get the number of promises in the queue. */
  get count() {
    return this.#queue.length;
  }
  /**
   * Close the queue and remove all outstanding promises (but do not reject them).
   */
  close() {
    this.clear(false);
  }
}
