// @matter/types/dist/cjs/clusters/fixed-label.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fixed_label_exports = {};
__export(fixed_label_exports, {
  FixedLabel: () => FixedLabel,
  FixedLabelCluster: () => FixedLabelCluster
});
module.exports = __toCommonJS(fixed_label_exports);
var import_MutableCluster = require("../cluster/mutation/MutableCluster.js");
var import_Cluster = require("../cluster/Cluster.js");
var import_TlvArray = require("../tlv/TlvArray.js");
var import_label = require("./label.js");
var import_ClusterRegistry = require("../cluster/ClusterRegistry.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
var FixedLabel;
((FixedLabel2) => {
  FixedLabel2.ClusterInstance = (0, import_MutableCluster.MutableCluster)({
    id: 64,
    name: "FixedLabel",
    revision: 1,
    attributes: {
      /**
       * @see {@link MatterSpecification.v14.Core} ยง 9.8.4
       */
      labelList: (0, import_Cluster.Attribute)(0, (0, import_TlvArray.TlvArray)(import_label.Label.TlvLabelStruct), { persistent: true, default: [] })
    }
  });
  FixedLabel2.Cluster = FixedLabel2.ClusterInstance;
  FixedLabel2.Complete = FixedLabel2.Cluster;
})(FixedLabel || (FixedLabel = {}));
const FixedLabelCluster = FixedLabel.Cluster;
import_ClusterRegistry.ClusterRegistry.register(FixedLabel.Complete);
