// @matter/node/dist/esm/behavior/system/commissioning/CommissioningClient.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Behavior } from "#behavior/Behavior.js";
import { ImplementationError, NotImplementedError, Time } from "#general";
import { DatatypeModel, FieldElement } from "#model";
import { IdentityService } from "#node/server/IdentityService.js";
import {
  ControllerCommissioner,
  FabricAuthority,
  FabricManager
} from "#protocol";
import { ControllerBehavior } from "../controller/ControllerBehavior.js";
import { RemoteDescriptor } from "./RemoteDescriptor.js";
class CommissioningClient extends Behavior {
  static early = true;
  static id = "commissioning";
  initialize(options) {
    const descriptor = options?.descriptor;
    if (descriptor) {
      this.descriptor = descriptor;
    }
    if (this.state.discoveredAt === void 0) {
      this.state.discoveredAt = Time.nowMs();
    }
    this.reactTo(this.endpoint.lifecycle.partsReady, this.#initializeNode);
  }
  async commission(options) {
    if (typeof options !== "object") {
      options = { passcode: options };
    }
    let { passcode } = options;
    if (typeof passcode !== "number" || Number.isNaN(passcode)) {
      passcode = Number.parseInt(passcode);
      if (Number.isNaN(passcode)) {
        throw new ImplementationError(`You must provide the numeric passcode to commission a node`);
      }
    }
    const node = this.endpoint;
    if (this.state.peerAddress !== void 0) {
      throw new ImplementationError(`${node} is already commissioned`);
    }
    await this.endpoint.owner?.act((agent) => agent.load(ControllerBehavior));
    const fabricAuthority = options.fabricAuthority || this.env.get(FabricAuthority);
    let { fabric } = options;
    if (fabric === void 0) {
      if (this.context.fabric === void 0) {
        fabric = await fabricAuthority.defaultFabric();
      } else {
        fabric = node.env.get(FabricManager).for(this.context.fabric);
      }
    }
    if (!fabricAuthority.hasControlOf(fabric)) {
      throw new ImplementationError(
        `Cannot commission ${node} fabric ${fabric.fabricIndex} because we do not control this fabric`
      );
    }
    const addresses = this.state.addresses;
    if (!addresses?.length) {
      throw new ImplementationError(`Cannot commission ${node} because the node has not been located`);
    }
    const commissioner = this.endpoint.env.get(ControllerCommissioner);
    const identityService = this.endpoint.env.get(IdentityService);
    const address = identityService.assignNodeAddress(node, fabric.fabricIndex, options.nodeId);
    const commissioningOptions = {
      addresses,
      fabric,
      nodeId: address.nodeId,
      passcode,
      discoveryData: this.descriptor
    };
    if (this.finalizeCommissioning !== CommissioningClient.prototype.finalizeCommissioning) {
      commissioningOptions.finalizeCommissioning = this.finalizeCommissioning.bind(this);
    }
    try {
      await commissioner.commission(commissioningOptions);
      this.state.peerAddress = address;
    } finally {
      if (this.state.peerAddress !== address) {
        identityService.releaseNodeAddress(address);
      }
    }
    return node;
  }
  /**
   * Override to implement CASE commissioning yourself.
   *
   * If you override, matter.js commissions to the point where over PASE is complete.  You must then complete
   * commissioning yourself by connecting to the device and invokeint the "CommissioningComplete" command.
   */
  async finalizeCommissioning(_address, _discoveryData) {
    throw new NotImplementedError();
  }
  get descriptor() {
    return RemoteDescriptor.fromLongForm(this.state);
  }
  set descriptor(descriptor) {
    RemoteDescriptor.toLongForm(descriptor, this.state);
  }
  #initializeNode() {
    const endpoint = this.endpoint;
    endpoint.lifecycle.initialized.emit(this.state.peerAddress !== void 0);
  }
  /**
   * Define logical schema.  This enables runtime validation, make fields persistent and makes subfields editable.
   */
  static schema = new DatatypeModel({
    name: "CommissioningState",
    type: "struct",
    children: [
      FieldElement({
        name: "peerAddress",
        type: "struct",
        quality: "N",
        children: [
          FieldElement({ name: "fabricIndex", type: "fabric-id" }),
          FieldElement({ name: "nodeId", type: "node-id" })
        ]
      }),
      FieldElement({
        name: "addresses",
        type: "list",
        quality: "N",
        children: [
          FieldElement({
            name: "entry",
            type: "struct",
            children: [
              FieldElement({ name: "type", type: "string" }),
              FieldElement({ name: "ip", type: "string" }),
              FieldElement({ name: "port", type: "uint16" }),
              FieldElement({ name: "peripheralAddress", type: "string" })
            ]
          })
        ]
      }),
      FieldElement({ name: "discoveredAt", type: "systime-ms", quality: "N", conformance: "M" }),
      FieldElement({ name: "onlineAt", type: "systime-ms" }),
      FieldElement({ name: "offlineAt", type: "systime-ms" }),
      FieldElement({ name: "ttl", type: "uint32", quality: "N" }),
      FieldElement({ name: "deviceIdentifier", type: "string", quality: "N" }),
      FieldElement({ name: "discriminator", type: "uint16", quality: "N" }),
      FieldElement({ name: "commissioningMode", type: "uint8", quality: "N" }),
      FieldElement({ name: "vendorId", type: "vendor-id", quality: "N" }),
      FieldElement({ name: "productId", type: "uint16", quality: "N" }),
      FieldElement({ name: "deviceType", type: "uint16", quality: "N" }),
      FieldElement({ name: "deviceName", type: "string", quality: "N" }),
      FieldElement({ name: "rotatingIdentifier", type: "string", quality: "N" }),
      FieldElement({ name: "pairingHint", type: "uint32", quality: "N" }),
      FieldElement({ name: "pairingInstructions", type: "string", quality: "N" }),
      FieldElement({
        name: "sessionParameters",
        type: "struct",
        quality: "N",
        children: [
          FieldElement({ name: "idleIntervalMs", type: "uint32", constraint: "max 3600000" }),
          FieldElement({ name: "activeIntervalMs", type: "uint32", constraint: "max 3600000" }),
          FieldElement({ name: "activeThresholdMs", type: "uint16" })
        ]
      }),
      FieldElement({ name: "tcpSupport", type: "uint8", quality: "N" }),
      FieldElement({ name: "longIdleOperatingMode", type: "bool", quality: "N" })
    ]
  });
}
((CommissioningClient2) => {
  class State {
    /**
     * Fabric index and node ID for paired peers.  If this is undefined the node is uncommissioned.
     */
    peerAddress;
    /**
     * Known network addresses for the device.  If this is undefined the node has not been located on any network
     * interface.
     */
    addresses;
    /**
     * Time at which the device was discovered.
     */
    discoveredAt = Time.nowMs();
    /**
     * Time at which we discovered the device's current operational addresses.
     */
    onlineAt;
    /**
     * Time at which we concluded the device's current operational address is unreachable.
     */
    offlineAt;
    /**
     * The TTL of the discovery record if applicable.
     */
    ttl;
    /**
     * The canonical global ID of the device.
     */
    deviceIdentifier;
    /**
     * The device's long discriminator.
     */
    discriminator;
    /**
     * The last know commissioning mode of the device.
     */
    commissioningMode;
    /**
     * Vendor.
     */
    vendorId;
    /**
     * Product.
     */
    productId;
    /**
     * Advertised device type.
     */
    deviceType;
    /**
     * The advertised device name specified by the user.
     */
    deviceName;
    /**
     * An optional manufacturer-specific unique rotating ID for uniquely identifying the device.
     */
    rotatingIdentifier;
    /**
     * A bitmap indicating how to transition the device to commissioning mode from its current state.
     */
    pairingHint;
    /**
     * Textual pairing instructions associated with pairing hint.
     */
    pairingInstructions;
    /**
     * The remote node's session parameters.
     */
    sessionParameters;
    /**
     * TCP support bitmap.
     */
    tcpSupport;
    /**
     * Indicates whether node is ICD with a slow (15 s+) polling interval.
     */
    longIdleTimeOperatingMode;
  }
  CommissioningClient2.State = State;
})(CommissioningClient || (CommissioningClient = {}));
export {
  CommissioningClient
};
