// @matter/types/dist/cjs/protocol/types/TlvAttributePath.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvAttributePath_exports = {};
__export(TlvAttributePath_exports, {
  TlvAttributePath: () => TlvAttributePath
});
module.exports = __toCommonJS(TlvAttributePath_exports);
var import_AttributeId = require("../../datatype/AttributeId.js");
var import_ClusterId = require("../../datatype/ClusterId.js");
var import_EndpointNumber = require("../../datatype/EndpointNumber.js");
var import_NodeId = require("../../datatype/NodeId.js");
var import_TlvBoolean = require("../../tlv/TlvBoolean.js");
var import_TlvNullable = require("../../tlv/TlvNullable.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_WildcardPathFlagsBitmap = require("./WildcardPathFlagsBitmap.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvAttributePath = (0, import_TlvObject.TlvTaggedList)({
  // AttributePathIB
  enableTagCompression: (0, import_TlvObject.TlvOptionalField)(0, import_TlvBoolean.TlvBoolean),
  nodeId: (0, import_TlvObject.TlvOptionalField)(1, import_NodeId.TlvNodeId),
  endpointId: (0, import_TlvObject.TlvOptionalField)(2, import_EndpointNumber.TlvEndpointNumber),
  clusterId: (0, import_TlvObject.TlvOptionalField)(3, import_ClusterId.TlvClusterId),
  attributeId: (0, import_TlvObject.TlvOptionalField)(4, import_AttributeId.TlvAttributeId),
  listIndex: (0, import_TlvObject.TlvOptionalField)(5, (0, import_TlvNullable.TlvNullable)(import_TlvNumber.TlvUInt16)),
  wildcardPathFlags: (0, import_TlvObject.TlvOptionalField)(6, (0, import_TlvNumber.TlvBitmap)(import_TlvNumber.TlvUInt32, import_WildcardPathFlagsBitmap.WildcardPathFlagsBitmap))
});
