// @matter/types/dist/cjs/protocol/types/TlvEventData.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TlvEventData_exports = {};
__export(TlvEventData_exports, {
  TlvEventData: () => TlvEventData
});
module.exports = __toCommonJS(TlvEventData_exports);
var import_EventNumber = require("../../datatype/EventNumber.js");
var import_TlvAny = require("../../tlv/TlvAny.js");
var import_TlvNumber = require("../../tlv/TlvNumber.js");
var import_TlvObject = require("../../tlv/TlvObject.js");
var import_TlvEventPath = require("./TlvEventPath.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
const TlvEventData = (0, import_TlvObject.TlvObject)({
  // EventDataIB
  path: (0, import_TlvObject.TlvField)(0, import_TlvEventPath.TlvEventPath),
  eventNumber: (0, import_TlvObject.TlvField)(1, import_EventNumber.TlvEventNumber),
  priority: (0, import_TlvObject.TlvField)(2, (0, import_TlvNumber.TlvEnum)()),
  epochTimestamp: (0, import_TlvObject.TlvOptionalField)(3, import_TlvNumber.TlvPosixMs),
  systemTimestamp: (0, import_TlvObject.TlvOptionalField)(4, import_TlvNumber.TlvSysTimeMS),
  deltaEpochTimestamp: (0, import_TlvObject.TlvOptionalField)(5, import_TlvNumber.TlvPosixMs),
  deltaSystemTimestamp: (0, import_TlvObject.TlvOptionalField)(6, import_TlvNumber.TlvSysTimeMS),
  data: (0, import_TlvObject.TlvOptionalField)(7, import_TlvAny.TlvAny)
});
