// @matter/node/dist/cjs/behavior/context/ActionTracer.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ActionTracer_exports = {};
__export(ActionTracer_exports, {
  ActionTracer: () => ActionTracer
});
module.exports = __toCommonJS(ActionTracer_exports);
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class ActionTracer {
}
((ActionTracer2) => {
  let ActionType;
  ((ActionType2) => {
    ActionType2["Initialize"] = "initialize";
    ActionType2["Read"] = "read";
    ActionType2["Write"] = "write";
    ActionType2["Invoke"] = "invoke";
    ActionType2["Destroy"] = "destroy";
  })(ActionType = ActionTracer2.ActionType || (ActionTracer2.ActionType = {}));
})(ActionTracer || (ActionTracer = {}));
