// @matter/general/dist/esm/time/Time.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Boot } from "#util/Boot.js";
import { CancelablePromise } from "#util/Cancelable.js";
import { ImplementationError } from "../MatterError.js";
import { Diagnostic } from "../log/Diagnostic.js";
import { DiagnosticSource } from "../log/DiagnosticSource.js";
const registry = /* @__PURE__ */ new Set();
class Time {
  static get;
  static startup = {
    systemMs: 0,
    processMs: 0
  };
  now() {
    return /* @__PURE__ */ new Date();
  }
  static now = () => Time.get().now();
  nowMs() {
    return this.now().getTime();
  }
  static nowMs = () => Time.get().nowMs();
  /**
   * Create a timer that will call callback after durationMs has passed.
   */
  getTimer(name, durationMs, callback) {
    return new StandardTimer(name, durationMs, callback, false);
  }
  static getTimer = (name, durationMs, callback) => Time.get().getTimer(name, durationMs, callback);
  /**
   * Create a timer that will periodically call callback at intervalMs intervals.
   */
  getPeriodicTimer(name, intervalMs, callback) {
    return new StandardTimer(name, intervalMs, callback, true);
  }
  static getPeriodicTimer = (name, intervalMs, callback) => Time.get().getPeriodicTimer(name, intervalMs, callback);
  /**
   * Create a promise that resolves after a specific interval or when canceled, whichever comes first.
   */
  sleep(name, durationMs) {
    let timer;
    let resolver;
    return new CancelablePromise(
      (resolve) => {
        resolver = resolve;
        timer = Time.getTimer(name, durationMs, resolve);
        timer.start();
      },
      () => {
        timer.stop();
        resolver();
      }
    );
  }
  static sleep(name, durationMs) {
    return Time.get().sleep(name, durationMs);
  }
  static register(timer) {
    timer.elapsed = Diagnostic.elapsed();
    registry.add(timer);
  }
  static unregister(timer) {
    registry.delete(timer);
  }
  static get timers() {
    return registry;
  }
}
const time = new Time();
Time.startup.systemMs = Time.startup.processMs = time.nowMs();
class StandardTimer {
  constructor(name, intervalMs, callback, isPeriodic) {
    this.name = name;
    this.intervalMs = intervalMs;
    this.callback = callback;
    this.isPeriodic = isPeriodic;
    if (intervalMs < 0 || intervalMs > 2147483647) {
      throw new ImplementationError(
        `Invalid intervalMs: ${intervalMs}. The value must be between 0 and 32-bit maximum value (2147483647)`
      );
    }
  }
  #timerId;
  #utility = false;
  isRunning = false;
  get systemId() {
    return Number(this.#timerId);
  }
  get utility() {
    return this.#utility;
  }
  set utility(utility) {
    if (utility === this.#utility) {
      return;
    }
    if (this.#timerId != void 0) {
      const timerId = this.#timerId;
      if (utility) {
        timerId.unref?.();
      } else {
        timerId.ref?.();
      }
    }
    this.#utility = utility;
  }
  start() {
    if (this.isRunning) this.stop();
    Time.register(this);
    this.isRunning = true;
    this.#timerId = (this.isPeriodic ? setInterval : setTimeout)(() => {
      if (!this.isPeriodic) {
        Time.unregister(this);
        this.isRunning = false;
      }
      this.callback();
    }, this.intervalMs);
    return this;
  }
  stop() {
    (this.isPeriodic ? clearInterval : clearTimeout)(this.#timerId);
    Time.unregister(this);
    this.isRunning = false;
    return this;
  }
}
DiagnosticSource.add({
  get [Diagnostic.value]() {
    return Diagnostic.node("\u23F1", "Timers", {
      children: [...registry].map((timer) => [
        timer.name,
        Diagnostic.dict({
          periodic: timer.isPeriodic,
          interval: Diagnostic.interval(timer.intervalMs),
          system: timer.systemId,
          elapsed: timer.elapsed
        })
      ])
    });
  }
});
Boot.init(() => {
  Time.get = () => time;
  if (typeof MatterHooks !== "undefined") {
    MatterHooks?.timeSetup?.(Time);
  }
});
export {
  StandardTimer,
  Time
};
