// @matter/node/dist/cjs/node/ServerNode.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ServerNode_exports = {};
__export(ServerNode_exports, {
  ServerNode: () => ServerNode
});
module.exports = __toCommonJS(ServerNode_exports);
var import_CommissioningServer = require("#behavior/system/commissioning/CommissioningServer.js");
var import_ControllerBehavior = require("#behavior/system/controller/ControllerBehavior.js");
var import_EventsBehavior = require("#behavior/system/events/EventsBehavior.js");
var import_NetworkServer = require("#behavior/system/network/NetworkServer.js");
var import_ServerNetworkRuntime = require("#behavior/system/network/ServerNetworkRuntime.js");
var import_ProductDescriptionServer = require("#behavior/system/product-description/ProductDescriptionServer.js");
var import_SessionsBehavior = require("#behavior/system/sessions/SessionsBehavior.js");
var import_subscription = require("#behavior/system/subscription/index.js");
var import_general = require("#general");
var import_protocol = require("#protocol");
var import_root = require("../endpoints/root.js");
var import_Node = require("./Node.js");
var import_ClientNodes = require("./client/ClientNodes.js");
var import_ServerEnvironment = require("./server/ServerEnvironment.js");
var import_ServerNodeStore = require("./storage/ServerNodeStore.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class FactoryResetError extends import_general.MatterError {
  constructor(message, cause) {
    super(message);
    this.cause = (0, import_general.errorOf)(cause);
  }
}
class ServerNode extends import_Node.Node {
  #nodes;
  #interaction;
  constructor(definition, options) {
    super(import_Node.Node.nodeConfigFor(ServerNode.RootEndpoint, definition, options));
    this.env.set(ServerNode, this);
    import_general.DiagnosticSource.add(this);
    this.construction.start();
  }
  static async create(definition, options) {
    return await (0, import_general.asyncNew)(this, definition, options);
  }
  createRuntime() {
    return new import_ServerNetworkRuntime.ServerNetworkRuntime(this);
  }
  async [import_general.Construction.destruct]() {
    await super[import_general.Construction.destruct]();
    await import_ServerEnvironment.ServerEnvironment.close(this);
  }
  async prepareRuntimeShutdown() {
    const sessions = this.env.get(import_protocol.SessionManager);
    await sessions.close();
  }
  /**
   * Perform a factory reset of the node.
   */
  async erase() {
    await this.lifecycle.mutex.produce(this.eraseWithMutex.bind(this));
  }
  async eraseWithMutex() {
    try {
      await this.construction;
      const isOnline = this.lifecycle.isOnline;
      if (isOnline) {
        await this.cancelWithMutex();
      }
      this.statusUpdate("resetting to factory defaults");
      await this.resetWithMutex();
      await this.resetStorage();
      this.construction.start();
      if (isOnline) {
        await this.startWithMutex();
      } else {
        await this.construction.ready;
      }
    } catch (e) {
      this.construction.crash();
      throw new FactoryResetError(`Error during factory reset of ${this}`, e);
    }
  }
  /**
   * Access other nodes on the fabric.
   */
  get nodes() {
    if (!this.#nodes) {
      this.#nodes = new import_ClientNodes.ClientNodes(this);
      this.#nodes.initialize();
    }
    return this.#nodes;
  }
  get interaction() {
    if (this.#interaction === void 0) {
      this.#interaction = new import_protocol.ServerInteraction(this.protocol);
    }
    return this.#interaction;
  }
  async advertiseNow() {
    await this.act(`advertiseNow<${this}>`, (agent) => agent.get(import_NetworkServer.NetworkServer).advertiseNow());
  }
  async initialize() {
    await import_ServerEnvironment.ServerEnvironment.initialize(this);
    await super.initialize();
  }
  /**
   * By default on factory reset we erase all stored data.
   *
   * If this is inappropriate for your application you may override to alter the behavior.   Matter requires that all
   * "security- and privacy-related data and key material" is removed on factory reset.
   *
   * @see {@link MatterSpecification.v12.Core} ยง 13.4
   */
  async resetStorage() {
    await this.env.get(import_protocol.SessionManager).clear();
    await this.env.get(import_protocol.FabricManager).clear();
    await this.env.get(import_protocol.OccurrenceManager).clear();
    await this.env.get(import_ServerNodeStore.ServerNodeStore).erase();
  }
  /**
   * Normal endpoints must have an owner to complete construction but server nodes have no such precondition for
   * construction.
   */
  assertConstructable() {
  }
}
((ServerNode2) => {
  ServerNode2.RootEndpoint = import_root.RootEndpoint.with(
    import_CommissioningServer.CommissioningServer,
    import_NetworkServer.NetworkServer,
    import_ProductDescriptionServer.ProductDescriptionServer,
    import_subscription.SubscriptionBehavior,
    import_SessionsBehavior.SessionsBehavior,
    import_EventsBehavior.EventsBehavior,
    import_ControllerBehavior.ControllerBehavior
  );
})(ServerNode || (ServerNode = {}));
