// @matter/node/dist/cjs/behavior/system/parts/PartsBehavior.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PartsBehavior_exports = {};
__export(PartsBehavior_exports, {
  PartsBehavior: () => PartsBehavior
});
module.exports = __toCommonJS(PartsBehavior_exports);
var import_Behavior = require("../../Behavior.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class PartsBehavior extends import_Behavior.Behavior {
  static id = "parts";
  add(child) {
    this.endpoint.parts.add(child);
  }
  delete(child) {
    return this.endpoint.parts.delete(child);
  }
  clear() {
    this.endpoint.parts.clear();
  }
  has(child) {
    return this.endpoint.parts.has(child);
  }
  indexOf(child) {
    return this.endpoint.parts.indexOf(child);
  }
  *[Symbol.iterator]() {
    for (const part of this.endpoint.parts) {
      yield this.context.agentFor(part);
    }
  }
}
