// @matter/protocol/dist/cjs/mdns/MdnsInstanceBroadcaster.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var MdnsInstanceBroadcaster_exports = {};
__export(MdnsInstanceBroadcaster_exports, {
  MdnsInstanceBroadcaster: () => MdnsInstanceBroadcaster
});
module.exports = __toCommonJS(MdnsInstanceBroadcaster_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MdnsInstanceBroadcaster {
  #instancePort;
  #mdnsBroadcaster;
  #onclose;
  #isClosed = false;
  constructor(instancePort, mdnsBroadcaster, onclose) {
    this.#instancePort = instancePort;
    this.#mdnsBroadcaster = mdnsBroadcaster;
    this.#onclose = onclose;
  }
  async close() {
    if (this.#isClosed) {
      return;
    }
    this.#isClosed = true;
    await this.#mdnsBroadcaster.expireAllAnnouncements(this.#instancePort);
    this.#onclose();
  }
  async setCommissionMode(mode, deviceData) {
    this.#assertOpen();
    await this.#mdnsBroadcaster.setCommissionMode(this.#instancePort, mode, deviceData);
  }
  /** Set the Broadcaster Data to announce a device for operative discovery (aka "already paired") */
  async setFabrics(fabrics, expireCommissioningAnnouncements = false) {
    this.#assertOpen();
    if (expireCommissioningAnnouncements) {
      await this.#mdnsBroadcaster.expireCommissioningAnnouncement(this.#instancePort);
    }
    await this.#mdnsBroadcaster.setFabrics(this.#instancePort, fabrics);
  }
  /** Set the Broadcaster data to announce a Commissioner (aka Commissioner discovery) */
  async setCommissionerInfo(commissionerData) {
    this.#assertOpen();
    await this.#mdnsBroadcaster.setCommissionerInfo(this.#instancePort, commissionerData);
  }
  async announce() {
    this.#assertOpen();
    await this.#mdnsBroadcaster.announce(this.#instancePort);
  }
  async expireFabricAnnouncement() {
    this.#assertOpen();
    await this.#mdnsBroadcaster.expireFabricAnnouncement(this.#instancePort);
  }
  async expireCommissioningAnnouncement() {
    this.#assertOpen();
    await this.#mdnsBroadcaster.expireCommissioningAnnouncement(this.#instancePort);
  }
  async expireAllAnnouncements() {
    this.#assertOpen();
    await this.#mdnsBroadcaster.expireAllAnnouncements(this.#instancePort);
  }
  #assertOpen() {
    if (this.#isClosed) {
      throw new import_general.ImplementationError("Illegal operation on closed MdnsInstanceBroadcaster");
    }
  }
}
