// @matter/general/dist/cjs/transaction/errors.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var errors_exports = {};
__export(errors_exports, {
  FinalizationError: () => FinalizationError,
  SynchronousTransactionConflictError: () => SynchronousTransactionConflictError,
  TransactionDeadlockError: () => TransactionDeadlockError,
  TransactionDestroyedError: () => TransactionDestroyedError,
  TransactionFlowError: () => TransactionFlowError,
  UnsettledStateError: () => UnsettledStateError
});
module.exports = __toCommonJS(errors_exports);
var import_MatterError = require("#MatterError.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class TransactionFlowError extends import_MatterError.MatterError {
}
class SynchronousTransactionConflictError extends import_MatterError.MatterError {
}
class TransactionDeadlockError extends import_MatterError.MatterError {
}
class FinalizationError extends import_MatterError.MatterError {
}
class TransactionDestroyedError extends import_MatterError.MatterError {
}
class UnsettledStateError extends FinalizationError {
}
