// @matter/protocol/dist/esm/certificate/CertificationDeclarationManager.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { Bytes, PrivateKey } from "#general";
import { CertificateManager, TlvCertificationDeclaration } from "./CertificateManager.js";
const TestCMS_SignerPrivateKey = Bytes.fromHex("AEF3484116E9481EC57BE0472DF41BF499064E5024AD869ECA5E889802D48075");
const TestCMS_SignerSubjectKeyIdentifier = Bytes.fromHex("62FA823359ACFAA9963E1CFA140ADDF504F37160");
class CertificationDeclarationManager {
  static generate(vendorId, productId, provisional = false) {
    const certificationElements = TlvCertificationDeclaration.encode({
      formatVersion: 1,
      vendorId,
      produceIdArray: [productId],
      deviceTypeId: 22,
      certificateId: "CSA00000SWC00000-00",
      securityLevel: 0,
      securityInformation: 0,
      versionNumber: 1,
      certificationType: provisional ? 1 : 0
      // 0 = Test, 1 = Provisional/In certification, 2 = official
    });
    return CertificateManager.certificationDeclarationToAsn1(
      certificationElements,
      TestCMS_SignerSubjectKeyIdentifier,
      PrivateKey(TestCMS_SignerPrivateKey)
    );
  }
}
export {
  CertificationDeclarationManager
};
