// @matter/model/dist/esm/elements/EventElement.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { ElementTag } from "../common/index.js";
import { ValueElement } from "./ValueElement.js";
function EventElement(definition, ...children) {
  return ValueElement(EventElement.Tag, definition, children);
}
((EventElement2) => {
  EventElement2.Tag = ElementTag.Event;
  let Priority;
  ((Priority2) => {
    Priority2["Debug"] = "debug";
    Priority2["Info"] = "info";
    Priority2["Critical"] = "critical";
  })(Priority = EventElement2.Priority || (EventElement2.Priority = {}));
  let PriorityId;
  ((PriorityId2) => {
    PriorityId2[PriorityId2["debug"] = 0] = "debug";
    PriorityId2[PriorityId2["info"] = 1] = "info";
    PriorityId2[PriorityId2["critical"] = 2] = "critical";
  })(PriorityId = EventElement2.PriorityId || (EventElement2.PriorityId = {}));
})(EventElement || (EventElement = {}));
const Debug = "debug" /* Debug */;
const Info = "info" /* Info */;
const Critical = "critical" /* Critical */;
export {
  Critical,
  Debug,
  EventElement,
  Info
};
