// @matter/model/dist/cjs/elements/EventElement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EventElement_exports = {};
__export(EventElement_exports, {
  Critical: () => Critical,
  Debug: () => Debug,
  EventElement: () => EventElement,
  Info: () => Info
});
module.exports = __toCommonJS(EventElement_exports);
var import_common = require("../common/index.js");
var import_ValueElement = require("./ValueElement.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function EventElement(definition, ...children) {
  return (0, import_ValueElement.ValueElement)(EventElement.Tag, definition, children);
}
((EventElement2) => {
  EventElement2.Tag = import_common.ElementTag.Event;
  let Priority;
  ((Priority2) => {
    Priority2["Debug"] = "debug";
    Priority2["Info"] = "info";
    Priority2["Critical"] = "critical";
  })(Priority = EventElement2.Priority || (EventElement2.Priority = {}));
  let PriorityId;
  ((PriorityId2) => {
    PriorityId2[PriorityId2["debug"] = 0] = "debug";
    PriorityId2[PriorityId2["info"] = 1] = "info";
    PriorityId2[PriorityId2["critical"] = 2] = "critical";
  })(PriorityId = EventElement2.PriorityId || (EventElement2.PriorityId = {}));
})(EventElement || (EventElement = {}));
const Debug = "debug" /* Debug */;
const Info = "info" /* Info */;
const Critical = "critical" /* Critical */;
