// @matter/model/dist/cjs/elements/RequirementElement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var RequirementElement_exports = {};
__export(RequirementElement_exports, {
  RequirementElement: () => RequirementElement
});
module.exports = __toCommonJS(RequirementElement_exports);
var import_common = require("../common/index.js");
var import_BaseElement = require("./BaseElement.js");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
function RequirementElement(definition, ...children) {
  return {
    ...(0, import_BaseElement.BaseElement)(RequirementElement.Tag, definition, children),
    tag: RequirementElement.Tag
  };
}
((RequirementElement2) => {
  RequirementElement2.Tag = import_common.ElementTag.Requirement;
  let ElementType;
  ((ElementType2) => {
    ElementType2["ClientCluster"] = "clientCluster";
    ElementType2["ServerCluster"] = "serverCluster";
    ElementType2["Feature"] = "feature";
    ElementType2["Attribute"] = "attribute";
    ElementType2["Command"] = "command";
    ElementType2["Event"] = "event";
    ElementType2["CommandField"] = "commandField";
  })(ElementType = RequirementElement2.ElementType || (RequirementElement2.ElementType = {}));
})(RequirementElement || (RequirementElement = {}));
