// @matter/node/dist/esm/tags/ElectricalMeasurementTag.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SemanticNamespace } from "../endpoint/type/SemanticNamespace.js";
const ElectricalMeasurementTag = SemanticNamespace({
  id: 10,
  tags: {
    /**
     * Indicates values measured for a DC load
     */
    Dc: { id: 0, label: "DC" },
    /**
     * Indicates values measured for a single-phase AC load, or values measured for the collective load on a
     * polyphase AC power supply
     */
    Ac: { id: 1, label: "AC" },
    /**
     * Indicates values measured for an AC load on phase 1 of a polyphase power supply
     */
    AcPhase1: { id: 2, label: "ACPhase1" },
    /**
     * Indicates values measured for an AC load on phase 2 of a polyphase power supply
     */
    AcPhase2: { id: 3, label: "ACPhase2" },
    /**
     * Indicates values measured for an AC load on phase 3 of a polyphase power supply
     */
    AcPhase3: { id: 4, label: "ACPhase3" }
  }
});
export {
  ElectricalMeasurementTag
};
