// @matter/node/dist/esm/tags/PowerSourceTag.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SemanticNamespace } from "../endpoint/type/SemanticNamespace.js";
const PowerSourceTag = SemanticNamespace({
  id: 15,
  tags: {
    /**
     * The Power Source cluster is related to power provided from an unknown source
     */
    Unknown: { id: 0, label: "Unknown" },
    /**
     * The Power Source cluster is related to power provided from the electrical grid
     *
     * Power Source clusters with this tag shall implement the WIRED feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.1
     */
    Grid: { id: 1, label: "Grid" },
    /**
     * The Power Source cluster is related to power provided from a solar panel array
     *
     * Power Source clusters with this tag shall implement the WIRED feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.2
     */
    Solar: { id: 2, label: "Solar" },
    /**
     * The Power Source cluster is related to power provided from a battery
     *
     * Power Source clusters with this tag shall implement the BAT feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.3
     */
    Battery: { id: 3, label: "Battery" },
    /**
     * The Power Source cluster is related to power provided from an electric vehicle
     *
     * Power Source clusters with this tag shall implement the BAT feature.
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 15.4
     */
    Ev: { id: 4, label: "EV" }
  }
});
export {
  PowerSourceTag
};
