// @matter/node/dist/esm/tags/SwitchesTag.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { SemanticNamespace } from "../endpoint/type/SemanticNamespace.js";
const SwitchesTag = SemanticNamespace({
  id: 67,
  tags: {
    On: { id: 0, label: "On" },
    Off: { id: 1, label: "Off" },
    Toggle: { id: 2, label: "Toggle" },
    /**
     * e.g. dim up (light)
     */
    Up: { id: 3, label: "Up" },
    /**
     * e.g. dim down (light)
     */
    Down: { id: 4, label: "Down" },
    /**
     * e.g. select next scene
     */
    Next: { id: 5, label: "Next" },
    /**
     * e.g. select previous scene
     */
    Previous: { id: 6, label: "Previous" },
    EnterOkSelect: { id: 7, label: "Enter/OK/Select" },
    /**
     * Textual description provided in Label field
     *
     * When this value is used, the Label field in the same Semantic Tag structure shall be filled with a textual
     * description of the function indicated on the button, such as a label or icon printed on the button, e.g.
     * "dining".
     *
     * @see {@link MatterSpecification.v14.Namespace} ยง 18.1
     */
    Custom: { id: 8, label: "Custom" }
  }
});
export {
  SwitchesTag
};
