// @matter/protocol/dist/esm/ble/BleConsts.js
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MAX_UDP_MESSAGE_SIZE } from "#general";
const BLE_MATTER_SERVICE_UUID = "fff6";
const BLE_MATTER_C1_CHARACTERISTIC_UUID = "18EE2EF5-263D-4559-959F-4F9C429F9D11";
const BLE_MATTER_C2_CHARACTERISTIC_UUID = "18EE2EF5-263D-4559-959F-4F9C429F9D12";
const BLE_MATTER_C3_CHARACTERISTIC_UUID = "64630238-8772-45F2-B87D-748A83218F04";
const BLE_MINIMUM_ATT_MTU = 20;
const BLE_MAXIMUM_BTP_MTU = 244;
const BTP_MAXIMUM_WINDOW_SIZE = 255;
const BTP_CONN_RSP_TIMEOUT_MS = 5e3;
const BTP_ACK_TIMEOUT_MS = 15e3;
const BTP_SEND_ACK_TIMEOUT_MS = BTP_ACK_TIMEOUT_MS / 3;
const BTP_CONN_IDLE_TIMEOUT = 3e4;
const BLE_MAX_MATTER_PAYLOAD_SIZE = MAX_UDP_MESSAGE_SIZE;
export {
  BLE_MATTER_C1_CHARACTERISTIC_UUID,
  BLE_MATTER_C2_CHARACTERISTIC_UUID,
  BLE_MATTER_C3_CHARACTERISTIC_UUID,
  BLE_MATTER_SERVICE_UUID,
  BLE_MAXIMUM_BTP_MTU,
  BLE_MAX_MATTER_PAYLOAD_SIZE,
  BLE_MINIMUM_ATT_MTU,
  BTP_ACK_TIMEOUT_MS,
  BTP_CONN_IDLE_TIMEOUT,
  BTP_CONN_RSP_TIMEOUT_MS,
  BTP_MAXIMUM_WINDOW_SIZE,
  BTP_SEND_ACK_TIMEOUT_MS
};
